{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "70192ac1-c745-42f1-bf8f-ba806e015f45",
   "metadata": {
    "collapsed": true,
    "execution": {
     "iopub.execute_input": "2025-04-18T08:10:24.292107Z",
     "iopub.status.busy": "2025-04-18T08:10:24.291658Z",
     "iopub.status.idle": "2025-04-18T08:11:43.768995Z",
     "shell.execute_reply": "2025-04-18T08:11:43.768114Z",
     "shell.execute_reply.started": "2025-04-18T08:10:24.292077Z"
    },
    "jupyter": {
     "outputs_hidden": true
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2025-04-18 08:10:24--  https://www.encodeproject.org/files/ENCFF000MBM/@@download/ENCFF000MBM.fastq.gz\n",
      "Resolving www.encodeproject.org (www.encodeproject.org)... 34.211.244.144\n",
      "Connecting to www.encodeproject.org (www.encodeproject.org)|34.211.244.144|:443... connected.\n",
      "HTTP request sent, awaiting response... 307 Temporary Redirect\n",
      "Location: https://encode-public.s3.amazonaws.com/2011/08/31/57768564-b4d3-4dc9-b824-db071328a28e/ENCFF000MBM.fastq.gz?response-content-disposition=attachment%3B%20filename%3DENCFF000MBM.fastq.gz&AWSAccessKeyId=ASIATGZNGCNX56TUAY6D&Signature=ZvPlYdGBbe4JhUGWtcaQ93mFf4I%3D&x-amz-security-token=IQoJb3JpZ2luX2VjEOj%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEaCXVzLXdlc3QtMiJIMEYCIQCzWPmnA85J4tF1WwmoUFZqzVEsQpQqd3RaVwFJTILrygIhAJH1KXhnsCA02FExCLY%2FoCOOAkzNJT8EmUh7sT7Qb7hPKrMFCHEQABoMMjIwNzQ4NzE0ODYzIgzyUw9tMBaf%2FVoG%2FTEqkAWq9mvR%2FeIQOgeQfY7pTwufMqHYunaQ%2BkDlH91Hci0qd72yrUqgWEoJANoqhADYXw4%2BUND%2F4KCldPKNBo4jjdOKUPzXlRw2hVFajl8YlhtiPhXJAG%2FnjKSIlQw6VhFmr0brb1c8%2BSvV%2FzqZWLoSQQ5u5ysC9DbQD%2Fcq1yJUIr0uHMK%2Bxr6KjdjWKN6TboSrAOPHBQaWfBcZiNUvNdmjbdWoGmE73RIqyQwAe7pWWDkmuzz9w1Pg2c21YU%2FeDQ0ITBakLw4ayZLEweJk3fT96h0gqn4BNxX1vKubTUhRvWsXiK%2BLrEV45fnmnlpAJWbHXzrKauVGn7IsTXovakdOvsyZlMpkVc8IIcVQL2fRRt2zh21AABKzjBHRehDeOGaC4uwJb%2BV%2FNCKjS7nav0TkPYaaYZVs7FC%2BGFwU%2BmeI%2BEaC6OeF1LRjpc7VOV2nHv8vhBWItf%2BzmvB2UOQhDCjvcGrtTVqpm5lsVrHJsW6VLG6FimGGXitkZ5Tl932pPQmIAG5JB%2B5r4yMks0jk7U6gz%2FfYb6PQoanKIMMJt6f%2FQhcgQZ01tOp8osx9t16Tu5Xq0M3B9fpgFZil79zkpxzQ%2F5uies%2FsU7GEJzYzgWbkuv7aREwAL1ZBNv%2BBHJr%2B%2BlusrizaXBu0mbuJEqnhq6LSdv9w%2BwvLwz7Fxot8irFLN0wuhIqiWiNjtrvvP9GWRXE%2Bz5zJlfChfDGOYFt3wbvK2hq55mmL4r%2Bm4IfMY5X8LCxWEf%2B0Qvj%2FUXIBTjAMIG8dsoT3hRpyBoBj3SIYtqLD0pDVdPEQ3PqSpTTbIm0Cbw55kqhh9bw4t6JDf7JIxZRFt3g9jhgmyr8H9PQhZp2hRpCOgoHeM0u9nqfr4gkziAsEBzCajYjABjqwAWD9VMR%2BJPW6IYrzeSpBAL6xSNPOEgl3goR%2Fo%2By22L%2BJ43c2ZjpGb5kQao8RXXWbf%2FXLe6fbGoz0BLLO4khfxqK7wmRr7EpOhuq83ZuZaSHZWlPDux2vy7jRNVsFZ1xWMeRrAnoeexEVrFDNF9NRjc4o1kAvbRJy4ISeiW%2BD4rJ3ca%2BR11tCH5P3x8wrz7hUp7oXayJvPkuyal3ShXG6gmsn5YFSP05znH%2BdRmZxzpeA&Expires=1745093424 [following]\n",
      "--2025-04-18 08:10:24--  https://encode-public.s3.amazonaws.com/2011/08/31/57768564-b4d3-4dc9-b824-db071328a28e/ENCFF000MBM.fastq.gz?response-content-disposition=attachment%3B%20filename%3DENCFF000MBM.fastq.gz&AWSAccessKeyId=ASIATGZNGCNX56TUAY6D&Signature=ZvPlYdGBbe4JhUGWtcaQ93mFf4I%3D&x-amz-security-token=IQoJb3JpZ2luX2VjEOj%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEaCXVzLXdlc3QtMiJIMEYCIQCzWPmnA85J4tF1WwmoUFZqzVEsQpQqd3RaVwFJTILrygIhAJH1KXhnsCA02FExCLY%2FoCOOAkzNJT8EmUh7sT7Qb7hPKrMFCHEQABoMMjIwNzQ4NzE0ODYzIgzyUw9tMBaf%2FVoG%2FTEqkAWq9mvR%2FeIQOgeQfY7pTwufMqHYunaQ%2BkDlH91Hci0qd72yrUqgWEoJANoqhADYXw4%2BUND%2F4KCldPKNBo4jjdOKUPzXlRw2hVFajl8YlhtiPhXJAG%2FnjKSIlQw6VhFmr0brb1c8%2BSvV%2FzqZWLoSQQ5u5ysC9DbQD%2Fcq1yJUIr0uHMK%2Bxr6KjdjWKN6TboSrAOPHBQaWfBcZiNUvNdmjbdWoGmE73RIqyQwAe7pWWDkmuzz9w1Pg2c21YU%2FeDQ0ITBakLw4ayZLEweJk3fT96h0gqn4BNxX1vKubTUhRvWsXiK%2BLrEV45fnmnlpAJWbHXzrKauVGn7IsTXovakdOvsyZlMpkVc8IIcVQL2fRRt2zh21AABKzjBHRehDeOGaC4uwJb%2BV%2FNCKjS7nav0TkPYaaYZVs7FC%2BGFwU%2BmeI%2BEaC6OeF1LRjpc7VOV2nHv8vhBWItf%2BzmvB2UOQhDCjvcGrtTVqpm5lsVrHJsW6VLG6FimGGXitkZ5Tl932pPQmIAG5JB%2B5r4yMks0jk7U6gz%2FfYb6PQoanKIMMJt6f%2FQhcgQZ01tOp8osx9t16Tu5Xq0M3B9fpgFZil79zkpxzQ%2F5uies%2FsU7GEJzYzgWbkuv7aREwAL1ZBNv%2BBHJr%2B%2BlusrizaXBu0mbuJEqnhq6LSdv9w%2BwvLwz7Fxot8irFLN0wuhIqiWiNjtrvvP9GWRXE%2Bz5zJlfChfDGOYFt3wbvK2hq55mmL4r%2Bm4IfMY5X8LCxWEf%2B0Qvj%2FUXIBTjAMIG8dsoT3hRpyBoBj3SIYtqLD0pDVdPEQ3PqSpTTbIm0Cbw55kqhh9bw4t6JDf7JIxZRFt3g9jhgmyr8H9PQhZp2hRpCOgoHeM0u9nqfr4gkziAsEBzCajYjABjqwAWD9VMR%2BJPW6IYrzeSpBAL6xSNPOEgl3goR%2Fo%2By22L%2BJ43c2ZjpGb5kQao8RXXWbf%2FXLe6fbGoz0BLLO4khfxqK7wmRr7EpOhuq83ZuZaSHZWlPDux2vy7jRNVsFZ1xWMeRrAnoeexEVrFDNF9NRjc4o1kAvbRJy4ISeiW%2BD4rJ3ca%2BR11tCH5P3x8wrz7hUp7oXayJvPkuyal3ShXG6gmsn5YFSP05znH%2BdRmZxzpeA&Expires=1745093424\n",
      "Resolving encode-public.s3.amazonaws.com (encode-public.s3.amazonaws.com)... 52.92.210.153, 52.218.183.41, 52.92.153.225, ...\n",
      "Connecting to encode-public.s3.amazonaws.com (encode-public.s3.amazonaws.com)|52.92.210.153|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 674224855 (643M) [binary/octet-stream]\n",
      "Saving to: ‘ENCFF000MBM.fastq.gz’\n",
      "\n",
      "ENCFF000MBM.fastq.g 100%[===================>] 642.99M  31.4MB/s    in 18s     \n",
      "\n",
      "2025-04-18 08:10:43 (34.9 MB/s) - ‘ENCFF000MBM.fastq.gz’ saved [674224855/674224855]\n",
      "\n",
      "--2025-04-18 08:10:43--  https://www.encodeproject.org/files/ENCFF000MBV/@@download/ENCFF000MBV.fastq.gz\n",
      "Resolving www.encodeproject.org (www.encodeproject.org)... 34.211.244.144\n",
      "Connecting to www.encodeproject.org (www.encodeproject.org)|34.211.244.144|:443... connected.\n",
      "HTTP request sent, awaiting response... 307 Temporary Redirect\n",
      "Location: https://encode-public.s3.amazonaws.com/2011/08/31/d3e082d8-f80c-4a3c-b751-84380a906b0f/ENCFF000MBV.fastq.gz?response-content-disposition=attachment%3B%20filename%3DENCFF000MBV.fastq.gz&AWSAccessKeyId=ASIATGZNGCNX56TUAY6D&Signature=3DCdBsVuglkvUXh1OJhleX6ecCo%3D&x-amz-security-token=IQoJb3JpZ2luX2VjEOj%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEaCXVzLXdlc3QtMiJIMEYCIQCzWPmnA85J4tF1WwmoUFZqzVEsQpQqd3RaVwFJTILrygIhAJH1KXhnsCA02FExCLY%2FoCOOAkzNJT8EmUh7sT7Qb7hPKrMFCHEQABoMMjIwNzQ4NzE0ODYzIgzyUw9tMBaf%2FVoG%2FTEqkAWq9mvR%2FeIQOgeQfY7pTwufMqHYunaQ%2BkDlH91Hci0qd72yrUqgWEoJANoqhADYXw4%2BUND%2F4KCldPKNBo4jjdOKUPzXlRw2hVFajl8YlhtiPhXJAG%2FnjKSIlQw6VhFmr0brb1c8%2BSvV%2FzqZWLoSQQ5u5ysC9DbQD%2Fcq1yJUIr0uHMK%2Bxr6KjdjWKN6TboSrAOPHBQaWfBcZiNUvNdmjbdWoGmE73RIqyQwAe7pWWDkmuzz9w1Pg2c21YU%2FeDQ0ITBakLw4ayZLEweJk3fT96h0gqn4BNxX1vKubTUhRvWsXiK%2BLrEV45fnmnlpAJWbHXzrKauVGn7IsTXovakdOvsyZlMpkVc8IIcVQL2fRRt2zh21AABKzjBHRehDeOGaC4uwJb%2BV%2FNCKjS7nav0TkPYaaYZVs7FC%2BGFwU%2BmeI%2BEaC6OeF1LRjpc7VOV2nHv8vhBWItf%2BzmvB2UOQhDCjvcGrtTVqpm5lsVrHJsW6VLG6FimGGXitkZ5Tl932pPQmIAG5JB%2B5r4yMks0jk7U6gz%2FfYb6PQoanKIMMJt6f%2FQhcgQZ01tOp8osx9t16Tu5Xq0M3B9fpgFZil79zkpxzQ%2F5uies%2FsU7GEJzYzgWbkuv7aREwAL1ZBNv%2BBHJr%2B%2BlusrizaXBu0mbuJEqnhq6LSdv9w%2BwvLwz7Fxot8irFLN0wuhIqiWiNjtrvvP9GWRXE%2Bz5zJlfChfDGOYFt3wbvK2hq55mmL4r%2Bm4IfMY5X8LCxWEf%2B0Qvj%2FUXIBTjAMIG8dsoT3hRpyBoBj3SIYtqLD0pDVdPEQ3PqSpTTbIm0Cbw55kqhh9bw4t6JDf7JIxZRFt3g9jhgmyr8H9PQhZp2hRpCOgoHeM0u9nqfr4gkziAsEBzCajYjABjqwAWD9VMR%2BJPW6IYrzeSpBAL6xSNPOEgl3goR%2Fo%2By22L%2BJ43c2ZjpGb5kQao8RXXWbf%2FXLe6fbGoz0BLLO4khfxqK7wmRr7EpOhuq83ZuZaSHZWlPDux2vy7jRNVsFZ1xWMeRrAnoeexEVrFDNF9NRjc4o1kAvbRJy4ISeiW%2BD4rJ3ca%2BR11tCH5P3x8wrz7hUp7oXayJvPkuyal3ShXG6gmsn5YFSP05znH%2BdRmZxzpeA&Expires=1745093443 [following]\n",
      "--2025-04-18 08:10:43--  https://encode-public.s3.amazonaws.com/2011/08/31/d3e082d8-f80c-4a3c-b751-84380a906b0f/ENCFF000MBV.fastq.gz?response-content-disposition=attachment%3B%20filename%3DENCFF000MBV.fastq.gz&AWSAccessKeyId=ASIATGZNGCNX56TUAY6D&Signature=3DCdBsVuglkvUXh1OJhleX6ecCo%3D&x-amz-security-token=IQoJb3JpZ2luX2VjEOj%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEaCXVzLXdlc3QtMiJIMEYCIQCzWPmnA85J4tF1WwmoUFZqzVEsQpQqd3RaVwFJTILrygIhAJH1KXhnsCA02FExCLY%2FoCOOAkzNJT8EmUh7sT7Qb7hPKrMFCHEQABoMMjIwNzQ4NzE0ODYzIgzyUw9tMBaf%2FVoG%2FTEqkAWq9mvR%2FeIQOgeQfY7pTwufMqHYunaQ%2BkDlH91Hci0qd72yrUqgWEoJANoqhADYXw4%2BUND%2F4KCldPKNBo4jjdOKUPzXlRw2hVFajl8YlhtiPhXJAG%2FnjKSIlQw6VhFmr0brb1c8%2BSvV%2FzqZWLoSQQ5u5ysC9DbQD%2Fcq1yJUIr0uHMK%2Bxr6KjdjWKN6TboSrAOPHBQaWfBcZiNUvNdmjbdWoGmE73RIqyQwAe7pWWDkmuzz9w1Pg2c21YU%2FeDQ0ITBakLw4ayZLEweJk3fT96h0gqn4BNxX1vKubTUhRvWsXiK%2BLrEV45fnmnlpAJWbHXzrKauVGn7IsTXovakdOvsyZlMpkVc8IIcVQL2fRRt2zh21AABKzjBHRehDeOGaC4uwJb%2BV%2FNCKjS7nav0TkPYaaYZVs7FC%2BGFwU%2BmeI%2BEaC6OeF1LRjpc7VOV2nHv8vhBWItf%2BzmvB2UOQhDCjvcGrtTVqpm5lsVrHJsW6VLG6FimGGXitkZ5Tl932pPQmIAG5JB%2B5r4yMks0jk7U6gz%2FfYb6PQoanKIMMJt6f%2FQhcgQZ01tOp8osx9t16Tu5Xq0M3B9fpgFZil79zkpxzQ%2F5uies%2FsU7GEJzYzgWbkuv7aREwAL1ZBNv%2BBHJr%2B%2BlusrizaXBu0mbuJEqnhq6LSdv9w%2BwvLwz7Fxot8irFLN0wuhIqiWiNjtrvvP9GWRXE%2Bz5zJlfChfDGOYFt3wbvK2hq55mmL4r%2Bm4IfMY5X8LCxWEf%2B0Qvj%2FUXIBTjAMIG8dsoT3hRpyBoBj3SIYtqLD0pDVdPEQ3PqSpTTbIm0Cbw55kqhh9bw4t6JDf7JIxZRFt3g9jhgmyr8H9PQhZp2hRpCOgoHeM0u9nqfr4gkziAsEBzCajYjABjqwAWD9VMR%2BJPW6IYrzeSpBAL6xSNPOEgl3goR%2Fo%2By22L%2BJ43c2ZjpGb5kQao8RXXWbf%2FXLe6fbGoz0BLLO4khfxqK7wmRr7EpOhuq83ZuZaSHZWlPDux2vy7jRNVsFZ1xWMeRrAnoeexEVrFDNF9NRjc4o1kAvbRJy4ISeiW%2BD4rJ3ca%2BR11tCH5P3x8wrz7hUp7oXayJvPkuyal3ShXG6gmsn5YFSP05znH%2BdRmZxzpeA&Expires=1745093443\n",
      "Resolving encode-public.s3.amazonaws.com (encode-public.s3.amazonaws.com)... 52.218.182.209, 52.92.165.129, 52.218.234.75, ...\n",
      "Connecting to encode-public.s3.amazonaws.com (encode-public.s3.amazonaws.com)|52.218.182.209|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 1766152791 (1.6G) [binary/octet-stream]\n",
      "Saving to: ‘ENCFF000MBV.fastq.gz’\n",
      "\n",
      "ENCFF000MBV.fastq.g 100%[===================>]   1.64G  28.6MB/s    in 59s     \n",
      "\n",
      "2025-04-18 08:11:43 (28.4 MB/s) - ‘ENCFF000MBV.fastq.gz’ saved [1766152791/1766152791]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!wget -c https://www.encodeproject.org/files/ENCFF000MBM/@@download/ENCFF000MBM.fastq.gz -O ENCFF000MBM.fastq.gz\n",
    "!wget -c https://www.encodeproject.org/files/ENCFF000MBV/@@download/ENCFF000MBV.fastq.gz -O ENCFF000MBV.fastq.gz\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "7194eb2f-ec72-4112-88b5-ef7aa412eee7",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-04-24T13:49:36.066348Z",
     "iopub.status.busy": "2025-04-24T13:49:36.065994Z",
     "iopub.status.idle": "2025-04-24T13:50:11.512096Z",
     "shell.execute_reply": "2025-04-24T13:50:11.511550Z",
     "shell.execute_reply.started": "2025-04-24T13:49:36.066327Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Retrieving notices: ...working... done\n",
      "Channels:\n",
      " - bioconda\n",
      " - conda-forge\n",
      "Platform: linux-64\n",
      "Collecting package metadata (repodata.json): done\n",
      "Solving environment: done\n",
      "\n",
      "\n",
      "==> WARNING: A newer version of conda exists. <==\n",
      "    current version: 24.9.2\n",
      "    latest version: 25.3.1\n",
      "\n",
      "Please update conda by running\n",
      "\n",
      "    $ conda update -n base -c conda-forge conda\n",
      "\n",
      "\n",
      "\n",
      "## Package Plan ##\n",
      "\n",
      "  environment location: /home/sagemaker-user/.conda/envs/rrbs-env\n",
      "\n",
      "  added / updated specs:\n",
      "    - bismark\n",
      "    - bowtie2\n",
      "    - cutadapt\n",
      "    - fastqc\n",
      "    - samtools\n",
      "    - trim-galore\n",
      "\n",
      "\n",
      "The following packages will be downloaded:\n",
      "\n",
      "    package                    |            build\n",
      "    ---------------------------|-----------------\n",
      "    ca-certificates-2025.1.31  |       hbd8a1cb_1         155 KB  conda-forge\n",
      "    freetype-2.13.3            |       ha770c72_1         168 KB  conda-forge\n",
      "    lerc-4.0.0                 |       h0aef613_1         258 KB  conda-forge\n",
      "    libdeflate-1.23            |       h86f0d12_0          71 KB  conda-forge\n",
      "    libfreetype-2.13.3         |       ha770c72_1           8 KB  conda-forge\n",
      "    libfreetype6-2.13.3        |       h48d6fc4_1         371 KB  conda-forge\n",
      "    libjpeg-turbo-3.1.0        |       hb9d3cd8_0         614 KB  conda-forge\n",
      "    libtiff-4.7.0              |       hd9ff511_4         419 KB  conda-forge\n",
      "    python_abi-3.12            |          7_cp312           7 KB  conda-forge\n",
      "    setuptools-79.0.1          |     pyhff2d567_0         769 KB  conda-forge\n",
      "    ------------------------------------------------------------\n",
      "                                           Total:         2.8 MB\n",
      "\n",
      "The following NEW packages will be INSTALLED:\n",
      "\n",
      "  _libgcc_mutex      conda-forge/linux-64::_libgcc_mutex-0.1-conda_forge \n",
      "  _openmp_mutex      conda-forge/linux-64::_openmp_mutex-4.5-2_gnu \n",
      "  alsa-lib           conda-forge/linux-64::alsa-lib-1.2.14-hb9d3cd8_0 \n",
      "  bismark            bioconda/noarch::bismark-0.24.2-hdfd78af_0 \n",
      "  bowtie2            bioconda/linux-64::bowtie2-2.5.4-he96a11b_5 \n",
      "  bzip2              conda-forge/linux-64::bzip2-1.0.8-h4bc722e_7 \n",
      "  c-ares             conda-forge/linux-64::c-ares-1.34.5-hb9d3cd8_0 \n",
      "  ca-certificates    conda-forge/noarch::ca-certificates-2025.1.31-hbd8a1cb_1 \n",
      "  cairo              conda-forge/linux-64::cairo-1.18.4-h3394656_0 \n",
      "  cffi               conda-forge/linux-64::cffi-1.17.1-py312h06ac9bb_0 \n",
      "  coreutils          bioconda/linux-64::coreutils-8.25-1 \n",
      "  cutadapt           bioconda/linux-64::cutadapt-5.0-py312h0fa9677_0 \n",
      "  dnaio              bioconda/linux-64::dnaio-1.2.2-py312hf67a6ed_0 \n",
      "  fastqc             bioconda/noarch::fastqc-0.12.1-hdfd78af_0 \n",
      "  font-ttf-dejavu-s~ conda-forge/noarch::font-ttf-dejavu-sans-mono-2.37-hab24e00_0 \n",
      "  font-ttf-inconsol~ conda-forge/noarch::font-ttf-inconsolata-3.000-h77eed37_0 \n",
      "  font-ttf-source-c~ conda-forge/noarch::font-ttf-source-code-pro-2.038-h77eed37_0 \n",
      "  font-ttf-ubuntu    conda-forge/noarch::font-ttf-ubuntu-0.83-h77eed37_3 \n",
      "  fontconfig         conda-forge/linux-64::fontconfig-2.15.0-h7e30c49_1 \n",
      "  fonts-conda-ecosy~ conda-forge/noarch::fonts-conda-ecosystem-1-0 \n",
      "  fonts-conda-forge  conda-forge/noarch::fonts-conda-forge-1-0 \n",
      "  freetype           conda-forge/linux-64::freetype-2.13.3-ha770c72_1 \n",
      "  giflib             conda-forge/linux-64::giflib-5.2.2-hd590300_0 \n",
      "  graphite2          conda-forge/linux-64::graphite2-1.3.13-h59595ed_1003 \n",
      "  harfbuzz           conda-forge/linux-64::harfbuzz-11.1.0-h3beb420_0 \n",
      "  hisat2             bioconda/linux-64::hisat2-2.2.1-h503566f_8 \n",
      "  htslib             bioconda/linux-64::htslib-1.21-h566b1c6_1 \n",
      "  icu                conda-forge/linux-64::icu-75.1-he02047a_0 \n",
      "  isa-l              conda-forge/linux-64::isa-l-2.31.1-hb9d3cd8_1 \n",
      "  k8                 bioconda/linux-64::k8-1.2-h6618dc6_3 \n",
      "  kernel-headers_li~ conda-forge/noarch::kernel-headers_linux-64-3.10.0-he073ed8_18 \n",
      "  keyutils           conda-forge/linux-64::keyutils-1.6.1-h166bdaf_0 \n",
      "  krb5               conda-forge/linux-64::krb5-1.21.3-h659f571_0 \n",
      "  lcms2              conda-forge/linux-64::lcms2-2.17-h717163a_0 \n",
      "  ld_impl_linux-64   conda-forge/linux-64::ld_impl_linux-64-2.43-h712a8e2_4 \n",
      "  lerc               conda-forge/linux-64::lerc-4.0.0-h0aef613_1 \n",
      "  libcups            conda-forge/linux-64::libcups-2.3.3-h4637d8d_4 \n",
      "  libcurl            conda-forge/linux-64::libcurl-8.13.0-h332b0f4_0 \n",
      "  libdeflate         conda-forge/linux-64::libdeflate-1.23-h86f0d12_0 \n",
      "  libedit            conda-forge/linux-64::libedit-3.1.20250104-pl5321h7949ede_0 \n",
      "  libev              conda-forge/linux-64::libev-4.33-hd590300_2 \n",
      "  libexpat           conda-forge/linux-64::libexpat-2.7.0-h5888daf_0 \n",
      "  libffi             conda-forge/linux-64::libffi-3.4.6-h2dba641_1 \n",
      "  libfreetype        conda-forge/linux-64::libfreetype-2.13.3-ha770c72_1 \n",
      "  libfreetype6       conda-forge/linux-64::libfreetype6-2.13.3-h48d6fc4_1 \n",
      "  libgcc             conda-forge/linux-64::libgcc-14.2.0-h767d61c_2 \n",
      "  libgcc-ng          conda-forge/linux-64::libgcc-ng-14.2.0-h69a702a_2 \n",
      "  libglib            conda-forge/linux-64::libglib-2.84.1-h2ff4ddf_0 \n",
      "  libgomp            conda-forge/linux-64::libgomp-14.2.0-h767d61c_2 \n",
      "  libiconv           conda-forge/linux-64::libiconv-1.18-h4ce23a2_1 \n",
      "  libjpeg-turbo      conda-forge/linux-64::libjpeg-turbo-3.1.0-hb9d3cd8_0 \n",
      "  liblzma            conda-forge/linux-64::liblzma-5.8.1-hb9d3cd8_0 \n",
      "  libnghttp2         conda-forge/linux-64::libnghttp2-1.64.0-h161d5f1_0 \n",
      "  libnsl             conda-forge/linux-64::libnsl-2.0.1-hd590300_0 \n",
      "  libpng             conda-forge/linux-64::libpng-1.6.47-h943b412_0 \n",
      "  libsqlite          conda-forge/linux-64::libsqlite-3.49.1-hee588c1_2 \n",
      "  libssh2            conda-forge/linux-64::libssh2-1.11.1-hf672d98_0 \n",
      "  libstdcxx          conda-forge/linux-64::libstdcxx-14.2.0-h8f9b012_2 \n",
      "  libstdcxx-ng       conda-forge/linux-64::libstdcxx-ng-14.2.0-h4852527_2 \n",
      "  libtiff            conda-forge/linux-64::libtiff-4.7.0-hd9ff511_4 \n",
      "  libuuid            conda-forge/linux-64::libuuid-2.38.1-h0b41bf4_0 \n",
      "  libwebp-base       conda-forge/linux-64::libwebp-base-1.5.0-h851e524_0 \n",
      "  libxcb             conda-forge/linux-64::libxcb-1.17.0-h8a09558_0 \n",
      "  libxcrypt          conda-forge/linux-64::libxcrypt-4.4.36-hd590300_1 \n",
      "  libzlib            conda-forge/linux-64::libzlib-1.3.1-hb9d3cd8_2 \n",
      "  minimap2           bioconda/linux-64::minimap2-2.29-h577a1d6_0 \n",
      "  ncurses            conda-forge/linux-64::ncurses-6.5-h2d0b736_3 \n",
      "  openjdk            conda-forge/linux-64::openjdk-23.0.2-h53dfc1b_2 \n",
      "  openssl            conda-forge/linux-64::openssl-3.5.0-h7b32b05_0 \n",
      "  pbzip2             conda-forge/linux-64::pbzip2-1.1.13-h1fcc475_2 \n",
      "  pcre2              conda-forge/linux-64::pcre2-10.44-hba22ea6_2 \n",
      "  perl               conda-forge/linux-64::perl-5.32.1-7_hd590300_perl5 \n",
      "  pigz               conda-forge/linux-64::pigz-2.8-hadc24fc_1 \n",
      "  pip                conda-forge/noarch::pip-25.0.1-pyh8b19718_0 \n",
      "  pixman             conda-forge/linux-64::pixman-0.44.2-h29eaf8c_0 \n",
      "  pthread-stubs      conda-forge/linux-64::pthread-stubs-0.4-hb9d3cd8_1002 \n",
      "  pycparser          conda-forge/noarch::pycparser-2.22-pyh29332c3_1 \n",
      "  python             conda-forge/linux-64::python-3.12.10-h9e4cc4f_0_cpython \n",
      "  python-isal        conda-forge/linux-64::python-isal-1.7.2-py312h66e93f0_0 \n",
      "  python-zlib-ng     conda-forge/linux-64::python-zlib-ng-0.5.1-py312h27ca174_0 \n",
      "  python_abi         conda-forge/noarch::python_abi-3.12-7_cp312 \n",
      "  readline           conda-forge/linux-64::readline-8.2-h8c095d6_2 \n",
      "  samtools           bioconda/linux-64::samtools-1.21-h96c455f_1 \n",
      "  setuptools         conda-forge/noarch::setuptools-79.0.1-pyhff2d567_0 \n",
      "  sysroot_linux-64   conda-forge/noarch::sysroot_linux-64-2.17-h0157908_18 \n",
      "  tk                 conda-forge/linux-64::tk-8.6.13-noxft_h4845f30_101 \n",
      "  trim-galore        bioconda/noarch::trim-galore-0.6.10-hdfd78af_1 \n",
      "  tzdata             conda-forge/noarch::tzdata-2025b-h78e105d_0 \n",
      "  wheel              conda-forge/noarch::wheel-0.45.1-pyhd8ed1ab_1 \n",
      "  xopen              conda-forge/noarch::xopen-2.0.2-pyh707e725_2 \n",
      "  xorg-libice        conda-forge/linux-64::xorg-libice-1.1.2-hb9d3cd8_0 \n",
      "  xorg-libsm         conda-forge/linux-64::xorg-libsm-1.2.6-he73a12e_0 \n",
      "  xorg-libx11        conda-forge/linux-64::xorg-libx11-1.8.12-h4f16b4b_0 \n",
      "  xorg-libxau        conda-forge/linux-64::xorg-libxau-1.0.12-hb9d3cd8_0 \n",
      "  xorg-libxdmcp      conda-forge/linux-64::xorg-libxdmcp-1.1.5-hb9d3cd8_0 \n",
      "  xorg-libxext       conda-forge/linux-64::xorg-libxext-1.3.6-hb9d3cd8_0 \n",
      "  xorg-libxfixes     conda-forge/linux-64::xorg-libxfixes-6.0.1-hb9d3cd8_0 \n",
      "  xorg-libxi         conda-forge/linux-64::xorg-libxi-1.8.2-hb9d3cd8_0 \n",
      "  xorg-libxrandr     conda-forge/linux-64::xorg-libxrandr-1.5.4-hb9d3cd8_0 \n",
      "  xorg-libxrender    conda-forge/linux-64::xorg-libxrender-0.9.12-hb9d3cd8_0 \n",
      "  xorg-libxt         conda-forge/linux-64::xorg-libxt-1.3.1-hb9d3cd8_0 \n",
      "  xorg-libxtst       conda-forge/linux-64::xorg-libxtst-1.2.5-hb9d3cd8_3 \n",
      "  zlib-ng            conda-forge/linux-64::zlib-ng-2.2.4-h7955e40_0 \n",
      "  zstandard          conda-forge/linux-64::zstandard-0.23.0-py312h66e93f0_1 \n",
      "  zstd               conda-forge/linux-64::zstd-1.5.7-hb8e6e7a_2 \n",
      "\n",
      "\n",
      "\n",
      "Downloading and Extracting Packages:\n",
      "setuptools-79.0.1    | 769 KB    |                                       |   0% \n",
      "libjpeg-turbo-3.1.0  | 614 KB    |                                       |   0% \u001b[A\n",
      "\n",
      "libtiff-4.7.0        | 419 KB    |                                       |   0% \u001b[A\u001b[A\n",
      "\n",
      "\n",
      "libfreetype6-2.13.3  | 371 KB    |                                       |   0% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "lerc-4.0.0           | 258 KB    |                                       |   0% \u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "freetype-2.13.3      | 168 KB    |                                       |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "ca-certificates-2025 | 155 KB    |                                       |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "libdeflate-1.23      | 71 KB     |                                       |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "libfreetype-2.13.3   | 8 KB      |                                       |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "python_abi-3.12      | 7 KB      |                                       |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "lerc-4.0.0           | 258 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "libfreetype6-2.13.3  | 371 KB    | #######9                              |  22% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "libtiff-4.7.0        | 419 KB    | ##################################### | 100% \u001b[A\u001b[A\n",
      "\n",
      "libtiff-4.7.0        | 419 KB    | ##################################### | 100% \u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "libdeflate-1.23      | 71 KB     | ########3                             |  23% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "ca-certificates-2025 | 155 KB    | ###8                                  |  10% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "python_abi-3.12      | 7 KB      | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "libfreetype-2.13.3   | 8 KB      | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "libjpeg-turbo-3.1.0  | 614 KB    | ##################################### | 100% \u001b[A\n",
      "libjpeg-turbo-3.1.0  | 614 KB    | ##################################### | 100% \u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "freetype-2.13.3      | 168 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "freetype-2.13.3      | 168 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "libfreetype6-2.13.3  | 371 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "ca-certificates-2025 | 155 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "libdeflate-1.23      | 71 KB     | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "python_abi-3.12      | 7 KB      | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "                                                                                \u001b[A\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "Preparing transaction: done\n",
      "Verifying transaction: done\n",
      "Executing transaction: done\n",
      "#\n",
      "# To activate this environment, use\n",
      "#\n",
      "#     $ conda activate rrbs-env\n",
      "#\n",
      "# To deactivate an active environment, use\n",
      "#\n",
      "#     $ conda deactivate\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!conda create -y -n rrbs-env -c bioconda fastqc trim-galore cutadapt bowtie2 bismark samtools\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "9579a045-f4c9-496c-a73a-d2e2f89cfa8f",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-04-21T07:21:56.528955Z",
     "iopub.status.busy": "2025-04-21T07:21:56.528361Z",
     "iopub.status.idle": "2025-04-21T07:21:57.732248Z",
     "shell.execute_reply": "2025-04-21T07:21:57.731684Z",
     "shell.execute_reply.started": "2025-04-21T07:21:56.528930Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FastQC v0.12.1\n",
      "5.0\n",
      "\n",
      "                        Quality-/Adapter-/RRBS-/Speciality-Trimming\n",
      "                                [powered by Cutadapt]\n",
      "                                  version 0.6.10\n",
      "\n",
      "                               Last update: 02 02 2023\n",
      "\n",
      "/home/sagemaker-user/.conda/envs/rrbs-env/bin/bowtie2-align-s version 2.5.4\n",
      "64-bit\n",
      "Built on fv-az849-261\n",
      "Tue Dec 10 19:14:18 UTC 2024\n",
      "Compiler: gcc version 13.3.0 (conda-forge gcc 13.3.0-1) \n",
      "Options: -O3 -msse2 -funroll-loops -g3 -fvisibility-inlines-hidden -fmessage-length=0 -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /home/sagemaker-user/.conda/envs/rrbs-env/include -fdebug-prefix-map=/opt/conda/conda-bld/bowtie2_1733857893168/work=/usr/local/src/conda/bowtie2-2.5.4 -fdebug-prefix-map=/home/sagemaker-user/.conda/envs/rrbs-env=/usr/local/src/conda-prefix -O3\n",
      "Sizeof {int, long, long long, void*, size_t, off_t}: {4, 8, 8, 8, 8, 8}\n",
      "\n",
      "\n",
      "          Bismark - Bisulfite Mapper and Methylation Caller.\n",
      "\n",
      "                       Bismark Version: v0.24.2\n",
      "        Copyright 2010-23 Felix Krueger, Altos Bioinformatics  \n",
      "                https://github.com/FelixKrueger/Bismark\n",
      "\n",
      "\n",
      "samtools 1.21\n",
      "Using htslib 1.21\n",
      "Copyright (C) 2024 Genome Research Ltd.\n",
      "\n",
      "Samtools compilation details:\n",
      "    Features:       build=configure curses=yes \n",
      "    CC:             /opt/conda/conda-bld/samtools_1734249775320/_build_env/bin/x86_64-conda-linux-gnu-cc\n",
      "    CPPFLAGS:       -DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -isystem /home/sagemaker-user/.conda/envs/rrbs-env/include\n",
      "    CFLAGS:         -Wall -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /home/sagemaker-user/.conda/envs/rrbs-env/include -fdebug-prefix-map=/opt/conda/conda-bld/samtools_1734249775320/work=/usr/local/src/conda/samtools-1.21 -fdebug-prefix-map=/home/sagemaker-user/.conda/envs/rrbs-env=/usr/local/src/conda-prefix\n",
      "    LDFLAGS:        -Wl,-O2 -Wl,--sort-common -Wl,--as-needed -Wl,-z,relro -Wl,-z,now -Wl,--disable-new-dtags -Wl,--gc-sections -Wl,--allow-shlib-undefined -Wl,-rpath,/home/sagemaker-user/.conda/envs/rrbs-env/lib -Wl,-rpath-link,/home/sagemaker-user/.conda/envs/rrbs-env/lib -L/home/sagemaker-user/.conda/envs/rrbs-env/lib\n",
      "    HTSDIR:         \n",
      "    LIBS:           \n",
      "    CURSES_LIB:     -ltinfow -lncursesw\n",
      "\n",
      "HTSlib compilation details:\n",
      "    Features:       build=configure libcurl=yes S3=yes GCS=yes libdeflate=yes lzma=yes bzip2=yes plugins=yes plugin-path=/home/sagemaker-user/.conda/envs/rrbs-env/libexec/htslib htscodecs=1.6.1\n",
      "    CC:             /opt/conda/conda-bld/htslib_1734145038589/_build_env/bin/x86_64-conda-linux-gnu-cc\n",
      "    CPPFLAGS:       -DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -isystem /home/sagemaker-user/.conda/envs/rrbs-env/include\n",
      "    CFLAGS:         -Wall -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /home/sagemaker-user/.conda/envs/rrbs-env/include -fdebug-prefix-map=/opt/conda/conda-bld/htslib_1734145038589/work=/usr/local/src/conda/htslib-1.21 -fdebug-prefix-map=/home/sagemaker-user/.conda/envs/rrbs-env=/usr/local/src/conda-prefix -fvisibility=hidden\n",
      "    LDFLAGS:        -Wl,-O2 -Wl,--sort-common -Wl,--as-needed -Wl,-z,relro -Wl,-z,now -Wl,--disable-new-dtags -Wl,--gc-sections -Wl,--allow-shlib-undefined -Wl,-rpath,/home/sagemaker-user/.conda/envs/rrbs-env/lib -Wl,-rpath-link,/home/sagemaker-user/.conda/envs/rrbs-env/lib -L/home/sagemaker-user/.conda/envs/rrbs-env/lib -fvisibility=hidden -rdynamic\n",
      "\n",
      "HTSlib URL scheme handlers present:\n",
      "    built-in:\t file, preload, data\n",
      "    S3 Multipart Upload:\t s3w+https, s3w+http, s3w\n",
      "    Amazon S3:\t s3+https, s3, s3+http\n",
      "    libcurl:\t gophers, smtp, wss, smb, rtsp, tftp, pop3, smbs, imaps, pop3s, ws, ftps, https, ftp, gopher, sftp, imap, http, smtps, scp, dict, mqtt, telnet\n",
      "    Google Cloud Storage:\t gs+http, gs+https, gs\n",
      "    crypt4gh-needed:\t crypt4gh\n",
      "    mem:\t mem\n"
     ]
    }
   ],
   "source": [
    "# FastQC (Java tool)\n",
    "!/home/sagemaker-user/.conda/envs/rrbs-env/opt/fastqc-0.12.1/fastqc --version\n",
    "\n",
    "# Cutadapt (Python, needed by Trim Galore)\n",
    "!/home/sagemaker-user/.conda/envs/rrbs-env/bin/cutadapt --version\n",
    "\n",
    "# Trim Galore (Perl wrapper)\n",
    "!/home/sagemaker-user/.conda/envs/rrbs-env/bin/trim_galore --version\n",
    "\n",
    "# Bowtie2 (Aligner)\n",
    "!/home/sagemaker-user/.conda/envs/rrbs-env/bin/bowtie2 --version\n",
    "\n",
    "# Bismark (Bisulfite aligner)\n",
    "!/home/sagemaker-user/.conda/envs/rrbs-env/bin/bismark --version\n",
    "\n",
    "# Samtools (for BAM handling)\n",
    "!/home/sagemaker-user/.conda/envs/rrbs-env/bin/samtools --version\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "39dd561f-a14f-49c7-8bf9-b16b8fe6205a",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-04-18T09:47:32.966545Z",
     "iopub.status.busy": "2025-04-18T09:47:32.965950Z",
     "iopub.status.idle": "2025-04-18T10:03:38.062004Z",
     "shell.execute_reply": "2025-04-18T10:03:38.061006Z",
     "shell.execute_reply.started": "2025-04-18T09:47:32.966520Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Path to Cutadapt set as: '/home/sagemaker-user/.conda/envs/rrbs-env/bin/cutadapt' (user defined)\n",
      "Cutadapt seems to be working fine (tested command '/home/sagemaker-user/.conda/envs/rrbs-env/bin/cutadapt --version')\n",
      "Cutadapt version: 5.0\n",
      "Could not detect version of Python used by Cutadapt from the first line of Cutadapt (but found this: >>>#!/bin/sh<<<)\n",
      "Letting the (modified) Cutadapt deal with the Python version instead\n",
      "Proceeding with 'gzip' for compression. PLEASE NOTE: Using multi-cores for trimming with 'gzip' only has only very limited effect! (see here: https://github.com/FelixKrueger/TrimGalore/issues/16#issuecomment-458557103)\n",
      "To increase performance, please install 'pigz' and run again\n",
      "\n",
      "Proceeding with 'gzip' for decompression\n",
      "To decrease CPU usage of decompression, please install 'igzip' and run again\n",
      "\n",
      "No quality encoding type selected. Assuming that the data provided uses Sanger encoded Phred scores (default)\n",
      "\n",
      "\n",
      "gzip: stdout: Broken pipe\n",
      "\n",
      "\n",
      "AUTO-DETECTING ADAPTER TYPE\n",
      "===========================\n",
      "Attempting to auto-detect adapter type from the first 1 million sequences of the first file (>> ENCFF000MBM.fastq.gz <<)\n",
      "\n",
      "Found perfect matches for the following adapter sequences:\n",
      "Adapter type\tCount\tSequence\tSequences analysed\tPercentage\n",
      "Illumina\t86\tAGATCGGAAGAGC\t1000000\t0.01\n",
      "Nextera\t0\tCTGTCTCTTATA\t1000000\t0.00\n",
      "smallRNA\t0\tTGGAATTCTCGG\t1000000\t0.00\n",
      "Using Illumina adapter for trimming (count: 86). Second best hit was Nextera (count: 0)\n",
      "\n",
      "\n",
      "gzip: stdout: Broken pipe\n",
      "Writing report to 'ENCFF000MBM.fastq.gz_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: ENCFF000MBM.fastq.gz\n",
      "Trimming mode: single-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 5.0\n",
      "Python version: could not detect\n",
      "Number of cores used for trimming: 8\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; auto-detected)\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length before a sequence gets removed: 20 bp\n",
      "File was specified to be an MspI-digested RRBS sample. Read 1 sequences with adapter contamination will be trimmed a further 2 bp from their 3' end, and Read 2 sequences will be trimmed by 2 bp from their 5' end to remove potential methylation-biased bases from the end-repair reaction\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 5.0). Setting -j 8\n",
      "  >>> Now performing adaptive quality trimming with a Phred-score cutoff of: 20 <<<\n",
      "\n",
      "This is cutadapt 5.0 with Python 3.12.10\n",
      "Command line parameters: -j 8 -e 0.1 -q 20 -a X ENCFF000MBM.fastq.gz\n",
      "Processing single-end reads on 8 cores ...\n",
      "[    8=------] 00:00:10     9,648,302 reads @   1.1 µs/read;  55.65 M reads/minute10000000 sequences processed\n",
      "[         8=-] 00:00:20    19,205,896 reads @   1.0 µs/read;  58.03 M reads/minute20000000 sequences processed\n",
      "Done           00:00:31    28,827,700 reads @   1.1 µs/read;  55.73 M reads/minute\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              28,827,700\n",
      "Reads with adapters:                         0 (0.0%)\n",
      "Reads written (passing filters):    28,827,700 (100.0%)\n",
      "\n",
      "Total basepairs processed: 1,037,797,200 bp\n",
      "Quality-trimmed:                       0 bp (0.0%)\n",
      "Total written (filtered):  1,037,797,200 bp (100.0%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: X; Type: regular 3'; Length: 1; Trimmed: 0 times\n",
      "\n",
      "  >>> Quality trimming completed <<<\n",
      "28827700 sequences processed in total\n",
      "\n",
      "Writing final adapter and quality trimmed output to ENCFF000MBM_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing adapter trimming for the adapter sequence: 'AGATCGGAAGAGC' from file ENCFF000MBM.fastq.gz_qual_trimmed.fastq <<< \n",
      "10000000 sequences processed\n",
      "20000000 sequences processed\n",
      "This is cutadapt 5.0 with Python 3.12.10\n",
      "Command line parameters: -j 8 -e 0.1 -O 1 -a AGATCGGAAGAGC ENCFF000MBM.fastq.gz_qual_trimmed.fastq\n",
      "Processing single-end reads on 8 cores ...\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              28,827,700\n",
      "Reads with adapters:                 7,237,361 (25.1%)\n",
      "Reads written (passing filters):    28,827,700 (100.0%)\n",
      "\n",
      "Total basepairs processed: 1,037,797,200 bp\n",
      "Total written (filtered):  1,026,149,290 bp (98.9%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 7237361 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 19.6%\n",
      "  C: 0.7%\n",
      "  G: 43.1%\n",
      "  T: 36.5%\n",
      "  none/other: 0.0%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t3608421\t7206925.0\t0\t3608421\n",
      "2\t3147688\t1801731.2\t0\t3147688\n",
      "3\t315857\t450432.8\t0\t315857\n",
      "4\t135462\t112608.2\t0\t135462\n",
      "5\t9663\t28152.1\t0\t9663\n",
      "6\t6611\t7038.0\t0\t6611\n",
      "7\t3035\t1759.5\t0\t3035\n",
      "8\t2397\t439.9\t0\t2397\n",
      "9\t1711\t110.0\t0\t1468 243\n",
      "10\t1232\t27.5\t1\t850 382\n",
      "11\t1269\t6.9\t1\t896 373\n",
      "12\t636\t1.7\t1\t447 189\n",
      "13\t326\t0.4\t1\t315 11\n",
      "14\t245\t0.4\t1\t228 17\n",
      "15\t204\t0.4\t1\t199 5\n",
      "16\t158\t0.4\t1\t153 5\n",
      "17\t220\t0.4\t1\t209 11\n",
      "18\t118\t0.4\t1\t114 4\n",
      "19\t130\t0.4\t1\t124 6\n",
      "20\t154\t0.4\t1\t144 10\n",
      "21\t97\t0.4\t1\t88 9\n",
      "22\t172\t0.4\t1\t163 9\n",
      "23\t102\t0.4\t1\t98 4\n",
      "24\t123\t0.4\t1\t114 9\n",
      "25\t95\t0.4\t1\t92 3\n",
      "26\t110\t0.4\t1\t109 1\n",
      "27\t149\t0.4\t1\t143 6\n",
      "28\t90\t0.4\t1\t85 5\n",
      "29\t97\t0.4\t1\t93 4\n",
      "30\t83\t0.4\t1\t82 1\n",
      "31\t94\t0.4\t1\t91 3\n",
      "32\t72\t0.4\t1\t66 6\n",
      "33\t101\t0.4\t1\t97 4\n",
      "34\t78\t0.4\t1\t74 4\n",
      "35\t87\t0.4\t1\t75 12\n",
      "36\t274\t0.4\t1\t167 107\n",
      "Successfully deleted temporary file ENCFF000MBM.fastq.gz_qual_trimmed.fastq\n",
      "\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: ENCFF000MBM.fastq.gz\n",
      "=============================================\n",
      "28827700 sequences processed in total\n",
      "Sequences were truncated to a varying degree because of deteriorating qualities (Phred score quality cutoff: 20):\t0 (0.0%)\n",
      "Sequences removed because they became shorter than the length cutoff of 20 bp:\t2808 (0.0%)\n",
      "RRBS reads trimmed by additional 2 bp when adapter contamination was detected:\t7237000 (25.1%)\n",
      "\n",
      "Path to Cutadapt set as: '/home/sagemaker-user/.conda/envs/rrbs-env/bin/cutadapt' (user defined)\n",
      "Cutadapt seems to be working fine (tested command '/home/sagemaker-user/.conda/envs/rrbs-env/bin/cutadapt --version')\n",
      "Cutadapt version: 5.0\n",
      "Could not detect version of Python used by Cutadapt from the first line of Cutadapt (but found this: >>>#!/bin/sh<<<)\n",
      "Letting the (modified) Cutadapt deal with the Python version instead\n",
      "Proceeding with 'gzip' for compression. PLEASE NOTE: Using multi-cores for trimming with 'gzip' only has only very limited effect! (see here: https://github.com/FelixKrueger/TrimGalore/issues/16#issuecomment-458557103)\n",
      "To increase performance, please install 'pigz' and run again\n",
      "\n",
      "Proceeding with 'gzip' for decompression\n",
      "To decrease CPU usage of decompression, please install 'igzip' and run again\n",
      "\n",
      "No quality encoding type selected. Assuming that the data provided uses Sanger encoded Phred scores (default)\n",
      "\n",
      "\n",
      "gzip: stdout: Broken pipe\n",
      "\n",
      "\n",
      "AUTO-DETECTING ADAPTER TYPE\n",
      "===========================\n",
      "Attempting to auto-detect adapter type from the first 1 million sequences of the first file (>> ENCFF000MBV.fastq.gz <<)\n",
      "\n",
      "Found perfect matches for the following adapter sequences:\n",
      "Adapter type\tCount\tSequence\tSequences analysed\tPercentage\n",
      "Illumina\t62\tAGATCGGAAGAGC\t1000000\t0.01\n",
      "smallRNA\t0\tTGGAATTCTCGG\t1000000\t0.00\n",
      "Nextera\t0\tCTGTCTCTTATA\t1000000\t0.00\n",
      "Using Illumina adapter for trimming (count: 62). Second best hit was smallRNA (count: 0)\n",
      "\n",
      "\n",
      "gzip: stdout: Broken pipe\n",
      "Writing report to 'ENCFF000MBV.fastq.gz_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: ENCFF000MBV.fastq.gz\n",
      "Trimming mode: single-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 5.0\n",
      "Python version: could not detect\n",
      "Number of cores used for trimming: 8\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; auto-detected)\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length before a sequence gets removed: 20 bp\n",
      "File was specified to be an MspI-digested RRBS sample. Read 1 sequences with adapter contamination will be trimmed a further 2 bp from their 3' end, and Read 2 sequences will be trimmed by 2 bp from their 5' end to remove potential methylation-biased bases from the end-repair reaction\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 5.0). Setting -j 8\n",
      "  >>> Now performing adaptive quality trimming with a Phred-score cutoff of: 20 <<<\n",
      "\n",
      "This is cutadapt 5.0 with Python 3.12.10\n",
      "Command line parameters: -j 8 -e 0.1 -q 20 -a X ENCFF000MBV.fastq.gz\n",
      "Processing single-end reads on 8 cores ...\n",
      "[    8=------] 00:00:10     9,350,070 reads @   1.2 µs/read;  51.43 M reads/minute10000000 sequences processed\n",
      "[          8<] 00:00:22    19,884,300 reads @   1.0 µs/read;  59.64 M reads/minute20000000 sequences processed\n",
      "[------>8    ] 00:00:32    29,593,958 reads @   1.2 µs/read;  50.52 M reads/minute30000000 sequences processed\n",
      "[->8         ] 00:00:43    39,193,778 reads @   1.1 µs/read;  54.02 M reads/minute40000000 sequences processed\n",
      "[   8=-------] 00:00:54    49,769,330 reads @   1.1 µs/read;  56.77 M reads/minute50000000 sequences processed\n",
      "Done           00:00:59    53,885,966 reads @   1.1 µs/read;  54.35 M reads/minute\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              53,885,966\n",
      "Reads with adapters:                         0 (0.0%)\n",
      "Reads written (passing filters):    53,885,966 (100.0%)\n",
      "\n",
      "Total basepairs processed: 1,939,894,776 bp\n",
      "Quality-trimmed:                       0 bp (0.0%)\n",
      "Total written (filtered):  1,939,894,776 bp (100.0%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: X; Type: regular 3'; Length: 1; Trimmed: 0 times\n",
      "\n",
      "  >>> Quality trimming completed <<<\n",
      "53885966 sequences processed in total\n",
      "\n",
      "Writing final adapter and quality trimmed output to ENCFF000MBV_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing adapter trimming for the adapter sequence: 'AGATCGGAAGAGC' from file ENCFF000MBV.fastq.gz_qual_trimmed.fastq <<< \n",
      "10000000 sequences processed\n",
      "20000000 sequences processed\n",
      "30000000 sequences processed\n",
      "40000000 sequences processed\n",
      "50000000 sequences processed\n",
      "This is cutadapt 5.0 with Python 3.12.10\n",
      "Command line parameters: -j 8 -e 0.1 -O 1 -a AGATCGGAAGAGC ENCFF000MBV.fastq.gz_qual_trimmed.fastq\n",
      "Processing single-end reads on 8 cores ...\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              53,885,966\n",
      "Reads with adapters:                12,518,758 (23.2%)\n",
      "Reads written (passing filters):    53,885,966 (100.0%)\n",
      "\n",
      "Total basepairs processed: 1,939,894,776 bp\n",
      "Total written (filtered):  1,919,906,660 bp (99.0%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 12518758 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 22.0%\n",
      "  C: 1.0%\n",
      "  G: 39.1%\n",
      "  T: 37.9%\n",
      "  none/other: 0.1%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t6316097\t13471491.5\t0\t6316097\n",
      "2\t5425925\t3367872.9\t0\t5425925\n",
      "3\t509762\t841968.2\t0\t509762\n",
      "4\t232596\t210492.1\t0\t232596\n",
      "5\t9357\t52623.0\t0\t9357\n",
      "6\t7198\t13155.8\t0\t7198\n",
      "7\t3094\t3288.9\t0\t3094\n",
      "8\t3145\t822.2\t0\t3145\n",
      "9\t1616\t205.6\t0\t1413 203\n",
      "10\t1332\t51.4\t1\t925 407\n",
      "11\t1253\t12.8\t1\t806 447\n",
      "12\t576\t3.2\t1\t464 112\n",
      "13\t350\t0.8\t1\t306 44\n",
      "14\t287\t0.8\t1\t250 37\n",
      "15\t249\t0.8\t1\t218 31\n",
      "16\t253\t0.8\t1\t221 32\n",
      "17\t248\t0.8\t1\t225 23\n",
      "18\t235\t0.8\t1\t216 19\n",
      "19\t222\t0.8\t1\t204 18\n",
      "20\t232\t0.8\t1\t209 23\n",
      "21\t177\t0.8\t1\t162 15\n",
      "22\t219\t0.8\t1\t196 23\n",
      "23\t164\t0.8\t1\t147 17\n",
      "24\t152\t0.8\t1\t134 18\n",
      "25\t162\t0.8\t1\t141 21\n",
      "26\t145\t0.8\t1\t127 18\n",
      "27\t1053\t0.8\t1\t878 175\n",
      "28\t137\t0.8\t1\t111 26\n",
      "29\t376\t0.8\t1\t314 62\n",
      "30\t143\t0.8\t1\t125 18\n",
      "31\t111\t0.8\t1\t93 18\n",
      "32\t98\t0.8\t1\t77 21\n",
      "33\t132\t0.8\t1\t108 24\n",
      "34\t149\t0.8\t1\t131 18\n",
      "35\t209\t0.8\t1\t160 49\n",
      "36\t1304\t0.8\t1\t482 822\n",
      "Successfully deleted temporary file ENCFF000MBV.fastq.gz_qual_trimmed.fastq\n",
      "\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: ENCFF000MBV.fastq.gz\n",
      "=============================================\n",
      "53885966 sequences processed in total\n",
      "Sequences were truncated to a varying degree because of deteriorating qualities (Phred score quality cutoff: 20):\t0 (0.0%)\n",
      "Sequences removed because they became shorter than the length cutoff of 20 bp:\t6170 (0.0%)\n",
      "RRBS reads trimmed by additional 2 bp when adapter contamination was detected:\t12517245 (23.2%)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!/home/sagemaker-user/.conda/envs/rrbs-env/bin/trim_galore \\\n",
    "  --rrbs \\\n",
    "  --quality 20 \\\n",
    "  --cores 8 \\\n",
    "  --path_to_cutadapt /home/sagemaker-user/.conda/envs/rrbs-env/bin/cutadapt \\\n",
    "  ENCFF000MBM.fastq.gz\n",
    "\n",
    "!/home/sagemaker-user/.conda/envs/rrbs-env/bin/trim_galore \\\n",
    "  --rrbs \\\n",
    "  --quality 20 \\\n",
    "  --cores 8 \\\n",
    "  --path_to_cutadapt /home/sagemaker-user/.conda/envs/rrbs-env/bin/cutadapt \\\n",
    "  ENCFF000MBV.fastq.gz\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7205adbf-aff4-419f-85ec-9b4aa6c296d7",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-04-18T11:59:39.012609Z",
     "iopub.status.busy": "2025-04-18T11:59:39.011997Z",
     "iopub.status.idle": "2025-04-18T11:59:39.168790Z",
     "shell.execute_reply": "2025-04-18T11:59:39.167765Z",
     "shell.execute_reply.started": "2025-04-18T11:59:39.012570Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENCFF000MBM.fastq.gz_trimming_report.txt\n",
      "ENCFF000MBV.fastq.gz_trimming_report.txt\n"
     ]
    }
   ],
   "source": [
    "!ls *.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "57a10cc4-a0b3-420b-a9d8-5ff0b699f8b9",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-04-18T11:59:56.464607Z",
     "iopub.status.busy": "2025-04-18T11:59:56.464189Z",
     "iopub.status.idle": "2025-04-18T11:59:56.794814Z",
     "shell.execute_reply": "2025-04-18T11:59:56.793595Z",
     "shell.execute_reply.started": "2025-04-18T11:59:56.464579Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: ENCFF000MBM.fastq.gz\n",
      "Trimming mode: single-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 5.0\n",
      "Python version: could not detect\n",
      "Number of cores used for trimming: 8\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Using Illumina adapter for trimming (count: 86). Second best hit was Nextera (count: 0)\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; auto-detected)\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length before a sequence gets removed: 20 bp\n",
      "File was specified to be an MspI-digested RRBS sample. Read 1 sequences with adapter contamination will be trimmed a further 2 bp from their 3' end, and Read 2 sequences will be trimmed by 2 bp from their 5' end to remove potential methylation-biased bases from the end-repair reaction\n",
      "Output file will be GZIP compressed\n",
      "\n",
      "\n",
      "This is cutadapt 5.0 with Python 3.12.10\n",
      "Command line parameters: -j 8 -e 0.1 -O 1 -a AGATCGGAAGAGC ENCFF000MBM.fastq.gz_qual_trimmed.fastq\n",
      "Processing single-end reads on 8 cores ...\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              28,827,700\n",
      "Reads with adapters:                 7,237,361 (25.1%)\n",
      "Reads written (passing filters):    28,827,700 (100.0%)\n",
      "\n",
      "Total basepairs processed: 1,037,797,200 bp\n",
      "Total written (filtered):  1,026,149,290 bp (98.9%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 7237361 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 19.6%\n",
      "  C: 0.7%\n",
      "  G: 43.1%\n",
      "  T: 36.5%\n",
      "  none/other: 0.0%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t3608421\t7206925.0\t0\t3608421\n",
      "2\t3147688\t1801731.2\t0\t3147688\n",
      "3\t315857\t450432.8\t0\t315857\n",
      "4\t135462\t112608.2\t0\t135462\n",
      "5\t9663\t28152.1\t0\t9663\n",
      "6\t6611\t7038.0\t0\t6611\n",
      "7\t3035\t1759.5\t0\t3035\n",
      "8\t2397\t439.9\t0\t2397\n",
      "9\t1711\t110.0\t0\t1468 243\n",
      "10\t1232\t27.5\t1\t850 382\n",
      "11\t1269\t6.9\t1\t896 373\n",
      "12\t636\t1.7\t1\t447 189\n",
      "13\t326\t0.4\t1\t315 11\n",
      "14\t245\t0.4\t1\t228 17\n",
      "15\t204\t0.4\t1\t199 5\n",
      "16\t158\t0.4\t1\t153 5\n",
      "17\t220\t0.4\t1\t209 11\n",
      "18\t118\t0.4\t1\t114 4\n",
      "19\t130\t0.4\t1\t124 6\n",
      "20\t154\t0.4\t1\t144 10\n",
      "21\t97\t0.4\t1\t88 9\n",
      "22\t172\t0.4\t1\t163 9\n",
      "23\t102\t0.4\t1\t98 4\n",
      "24\t123\t0.4\t1\t114 9\n",
      "25\t95\t0.4\t1\t92 3\n",
      "26\t110\t0.4\t1\t109 1\n",
      "27\t149\t0.4\t1\t143 6\n",
      "28\t90\t0.4\t1\t85 5\n",
      "29\t97\t0.4\t1\t93 4\n",
      "30\t83\t0.4\t1\t82 1\n",
      "31\t94\t0.4\t1\t91 3\n",
      "32\t72\t0.4\t1\t66 6\n",
      "33\t101\t0.4\t1\t97 4\n",
      "34\t78\t0.4\t1\t74 4\n",
      "35\t87\t0.4\t1\t75 12\n",
      "36\t274\t0.4\t1\t167 107\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: ENCFF000MBM.fastq.gz\n",
      "=============================================\n",
      "28827700 sequences processed in total\n",
      "Sequences were truncated to a varying degree because of deteriorating qualities (Phred score quality cutoff: 20):\t0 (0.0%)\n",
      "Sequences removed because they became shorter than the length cutoff of 20 bp:\t2808 (0.0%)\n",
      "RRBS reads trimmed by additional 2 bp when adapter contamination was detected:\t7237000 (25.1%)\n",
      "\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: ENCFF000MBV.fastq.gz\n",
      "Trimming mode: single-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 5.0\n",
      "Python version: could not detect\n",
      "Number of cores used for trimming: 8\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Using Illumina adapter for trimming (count: 62). Second best hit was smallRNA (count: 0)\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; auto-detected)\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length before a sequence gets removed: 20 bp\n",
      "File was specified to be an MspI-digested RRBS sample. Read 1 sequences with adapter contamination will be trimmed a further 2 bp from their 3' end, and Read 2 sequences will be trimmed by 2 bp from their 5' end to remove potential methylation-biased bases from the end-repair reaction\n",
      "Output file will be GZIP compressed\n",
      "\n",
      "\n",
      "This is cutadapt 5.0 with Python 3.12.10\n",
      "Command line parameters: -j 8 -e 0.1 -O 1 -a AGATCGGAAGAGC ENCFF000MBV.fastq.gz_qual_trimmed.fastq\n",
      "Processing single-end reads on 8 cores ...\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              53,885,966\n",
      "Reads with adapters:                12,518,758 (23.2%)\n",
      "Reads written (passing filters):    53,885,966 (100.0%)\n",
      "\n",
      "Total basepairs processed: 1,939,894,776 bp\n",
      "Total written (filtered):  1,919,906,660 bp (99.0%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 12518758 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 22.0%\n",
      "  C: 1.0%\n",
      "  G: 39.1%\n",
      "  T: 37.9%\n",
      "  none/other: 0.1%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t6316097\t13471491.5\t0\t6316097\n",
      "2\t5425925\t3367872.9\t0\t5425925\n",
      "3\t509762\t841968.2\t0\t509762\n",
      "4\t232596\t210492.1\t0\t232596\n",
      "5\t9357\t52623.0\t0\t9357\n",
      "6\t7198\t13155.8\t0\t7198\n",
      "7\t3094\t3288.9\t0\t3094\n",
      "8\t3145\t822.2\t0\t3145\n",
      "9\t1616\t205.6\t0\t1413 203\n",
      "10\t1332\t51.4\t1\t925 407\n",
      "11\t1253\t12.8\t1\t806 447\n",
      "12\t576\t3.2\t1\t464 112\n",
      "13\t350\t0.8\t1\t306 44\n",
      "14\t287\t0.8\t1\t250 37\n",
      "15\t249\t0.8\t1\t218 31\n",
      "16\t253\t0.8\t1\t221 32\n",
      "17\t248\t0.8\t1\t225 23\n",
      "18\t235\t0.8\t1\t216 19\n",
      "19\t222\t0.8\t1\t204 18\n",
      "20\t232\t0.8\t1\t209 23\n",
      "21\t177\t0.8\t1\t162 15\n",
      "22\t219\t0.8\t1\t196 23\n",
      "23\t164\t0.8\t1\t147 17\n",
      "24\t152\t0.8\t1\t134 18\n",
      "25\t162\t0.8\t1\t141 21\n",
      "26\t145\t0.8\t1\t127 18\n",
      "27\t1053\t0.8\t1\t878 175\n",
      "28\t137\t0.8\t1\t111 26\n",
      "29\t376\t0.8\t1\t314 62\n",
      "30\t143\t0.8\t1\t125 18\n",
      "31\t111\t0.8\t1\t93 18\n",
      "32\t98\t0.8\t1\t77 21\n",
      "33\t132\t0.8\t1\t108 24\n",
      "34\t149\t0.8\t1\t131 18\n",
      "35\t209\t0.8\t1\t160 49\n",
      "36\t1304\t0.8\t1\t482 822\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: ENCFF000MBV.fastq.gz\n",
      "=============================================\n",
      "53885966 sequences processed in total\n",
      "Sequences were truncated to a varying degree because of deteriorating qualities (Phred score quality cutoff: 20):\t0 (0.0%)\n",
      "Sequences removed because they became shorter than the length cutoff of 20 bp:\t6170 (0.0%)\n",
      "RRBS reads trimmed by additional 2 bp when adapter contamination was detected:\t12517245 (23.2%)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!cat ENCFF000MBM.fastq.gz_trimming_report.txt\n",
    "!cat ENCFF000MBV.fastq.gz_trimming_report.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e28b9676-6870-426a-a894-35be593e809e",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-04-20T17:06:25.686652Z",
     "iopub.status.busy": "2025-04-20T17:06:25.686450Z",
     "iopub.status.idle": "2025-04-20T17:06:25.689895Z",
     "shell.execute_reply": "2025-04-20T17:06:25.689280Z",
     "shell.execute_reply.started": "2025-04-20T17:06:25.686632Z"
    }
   },
   "outputs": [],
   "source": [
    "import os\n",
    "os.environ[\"PATH\"] = \"/home/sagemaker-user/.conda/envs/rrbs-env/bin:\" + os.environ[\"PATH\"]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f93322af-3cda-40f5-8990-3cc92f10ebb7",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-04-18T12:03:02.812314Z",
     "iopub.status.busy": "2025-04-18T12:03:02.811595Z",
     "iopub.status.idle": "2025-04-18T12:04:06.147712Z",
     "shell.execute_reply": "2025-04-18T12:04:06.146720Z",
     "shell.execute_reply.started": "2025-04-18T12:03:02.812287Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing bisulfite genomes out into a single MFA (multi FastA) file\n",
      "\n",
      "Bisulfite Genome Indexer version v0.24.2 (last modified: 19 May 2022)\n",
      "\n",
      "Step I - Prepare genome folders - completed\n",
      "\n",
      "\n",
      "\n",
      "Total number of conversions performed:\n",
      "C->T:\t579913581\n",
      "G->A:\t580204968\n",
      "\n",
      "Step II - Genome bisulfite conversions - completed\n",
      "\n",
      "\n",
      "Bismark Genome Preparation - Step III: Launching the Bowtie 2 indexer\n",
      "Please be aware that this process can - depending on genome size - take several hours!\n",
      "Can't exec \"bowtie2-build\": No such file or directory at /home/sagemaker-user/.conda/envs/rrbs-env/bin/bismark_genome_preparation line 280, <IN> line 46.\n",
      "Parent process: Failed to build index\n"
     ]
    }
   ],
   "source": [
    "!mkdir -p genome_hg19\n",
    "!cp hg19_main.fa genome_hg19/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "868dff97-569f-4346-a855-6c95e6b62404",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-04-18T12:07:50.408957Z",
     "iopub.status.busy": "2025-04-18T12:07:50.408274Z",
     "iopub.status.idle": "2025-04-18T15:22:24.471555Z",
     "shell.execute_reply": "2025-04-18T15:22:24.470363Z",
     "shell.execute_reply.started": "2025-04-18T12:07:50.408925Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing bisulfite genomes out into a single MFA (multi FastA) file\n",
      "\n",
      "Bisulfite Genome Indexer version v0.24.2 (last modified: 19 May 2022)\n",
      "\n",
      "A directory called /home/sagemaker-user/src/genome_hg19/Bisulfite_Genome/ already exists. Already existing converted sequences and/or already existing Bowtie 2, HISAT2 or Minimap2) indices will be overwritten!\n",
      "\n",
      "\n",
      "Step I - Prepare genome folders - completed\n",
      "\n",
      "\n",
      "\n",
      "Total number of conversions performed:\n",
      "C->T:\t579913581\n",
      "G->A:\t580204968\n",
      "\n",
      "Step II - Genome bisulfite conversions - completed\n",
      "\n",
      "\n",
      "Bismark Genome Preparation - Step III: Launching the Bowtie 2 indexer\n",
      "Please be aware that this process can - depending on genome size - take several hours!\n",
      "Settings:\n",
      "  Output files: \"BS_CT.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genome_mfa.CT_conversion.fa\n",
      "Building a SMALL index\n",
      "Reading reference sizes\n",
      "Settings:\n",
      "  Output files: \"BS_GA.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genome_mfa.GA_conversion.fa\n",
      "Building a SMALL index\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:15\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time reading reference sizes: 00:00:15\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:16\n",
      "bmax according to bmaxDivN setting: 708918391\n",
      "Using parameters --bmax 531688794 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Time to join reference sequences: 00:00:16\n",
      "bmax according to bmaxDivN setting: 708918391\n",
      "Using parameters --bmax 531688794 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 531688794 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Passed!  Constructing with these parameters: --bmax 531688794 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:01:58\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  V-Sorting samples time: 00:01:58\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:34\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Ranking v-sort output time: 00:00:34\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:50\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:51\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 1, merged 8; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 2, merged 7; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 1, merged 0; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 1, merged 0; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 1, merged 1; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 4.05096e+08 (target: 531688793)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 7\n",
      "  Reserving size (531688794) for bucket 1\n",
      "  Calculating Z arrays for bucket 1\n",
      "  Entering block accumulator loop for bucket 1:\n",
      "  bucket 1: 10%\n",
      "  bucket 1: 20%\n",
      "  bucket 1: 30%\n",
      "  bucket 1: 40%\n",
      "  bucket 1: 50%\n",
      "  bucket 1: 60%\n",
      "  bucket 1: 70%\n",
      "  bucket 1: 80%\n",
      "  bucket 1: 90%\n",
      "  bucket 1: 100%\n",
      "  Sorting block of length 434155799 for bucket 1\n",
      "  (Using difference cover)\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 4.05096e+08 (target: 531688793)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 7\n",
      "  Reserving size (531688794) for bucket 1\n",
      "  Calculating Z arrays for bucket 1\n",
      "  Entering block accumulator loop for bucket 1:\n",
      "  bucket 1: 10%\n",
      "  bucket 1: 20%\n",
      "  bucket 1: 30%\n",
      "  bucket 1: 40%\n",
      "  bucket 1: 50%\n",
      "  bucket 1: 60%\n",
      "  bucket 1: 70%\n",
      "  bucket 1: 80%\n",
      "  bucket 1: 90%\n",
      "  bucket 1: 100%\n",
      "  Sorting block of length 413661440 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:07:35\n",
      "Returning block of 434155800 for bucket 1\n",
      "Getting block 2 of 7\n",
      "  Reserving size (531688794) for bucket 2\n",
      "  Calculating Z arrays for bucket 2\n",
      "  Entering block accumulator loop for bucket 2:\n",
      "  bucket 2: 10%\n",
      "  bucket 2: 20%\n",
      "  bucket 2: 30%\n",
      "  bucket 2: 40%\n",
      "  bucket 2: 50%\n",
      "  Sorting block time: 00:09:06\n",
      "Returning block of 413661441 for bucket 1\n",
      "  bucket 2: 60%\n",
      "  bucket 2: 70%\n",
      "  bucket 2: 80%\n",
      "  bucket 2: 90%\n",
      "  bucket 2: 100%\n",
      "  Sorting block of length 497026000 for bucket 2\n",
      "  (Using difference cover)\n",
      "Getting block 2 of 7\n",
      "  Reserving size (531688794) for bucket 2\n",
      "  Calculating Z arrays for bucket 2\n",
      "  Entering block accumulator loop for bucket 2:\n",
      "  bucket 2: 10%\n",
      "  bucket 2: 20%\n",
      "  bucket 2: 30%\n",
      "  bucket 2: 40%\n",
      "  bucket 2: 50%\n",
      "  bucket 2: 60%\n",
      "  bucket 2: 70%\n",
      "  bucket 2: 80%\n",
      "  bucket 2: 90%\n",
      "  bucket 2: 100%\n",
      "  Sorting block of length 302891130 for bucket 2\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:06:38\n",
      "Returning block of 302891131 for bucket 2\n",
      "  Sorting block time: 00:09:00\n",
      "Returning block of 497026001 for bucket 2\n",
      "Getting block 3 of 7\n",
      "  Reserving size (531688794) for bucket 3\n",
      "  Calculating Z arrays for bucket 3\n",
      "  Entering block accumulator loop for bucket 3:\n",
      "  bucket 3: 10%\n",
      "  bucket 3: 20%\n",
      "  bucket 3: 30%\n",
      "  bucket 3: 40%\n",
      "  bucket 3: 50%\n",
      "  bucket 3: 60%\n",
      "  bucket 3: 70%\n",
      "  bucket 3: 80%\n",
      "  bucket 3: 90%\n",
      "Getting block 3 of 7\n",
      "  Reserving size (531688794) for bucket 3\n",
      "  Calculating Z arrays for bucket 3\n",
      "  Entering block accumulator loop for bucket 3:\n",
      "  bucket 3: 100%\n",
      "  Sorting block of length 338552273 for bucket 3\n",
      "  (Using difference cover)\n",
      "  bucket 3: 10%\n",
      "  bucket 3: 20%\n",
      "  bucket 3: 30%\n",
      "  bucket 3: 40%\n",
      "  bucket 3: 50%\n",
      "  bucket 3: 60%\n",
      "  bucket 3: 70%\n",
      "  bucket 3: 80%\n",
      "  bucket 3: 90%\n",
      "  bucket 3: 100%\n",
      "  Sorting block of length 380883709 for bucket 3\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:07:44\n",
      "Returning block of 338552274 for bucket 3\n",
      "  Sorting block time: 00:06:50\n",
      "Returning block of 380883710 for bucket 3\n",
      "Getting block 4 of 7\n",
      "  Reserving size (531688794) for bucket 4\n",
      "  Calculating Z arrays for bucket 4\n",
      "  Entering block accumulator loop for bucket 4:\n",
      "  bucket 4: 10%\n",
      "  bucket 4: 20%\n",
      "  bucket 4: 30%\n",
      "Getting block 4 of 7\n",
      "  Reserving size (531688794) for bucket 4\n",
      "  Calculating Z arrays for bucket 4\n",
      "  Entering block accumulator loop for bucket 4:\n",
      "  bucket 4: 40%\n",
      "  bucket 4: 10%\n",
      "  bucket 4: 50%\n",
      "  bucket 4: 20%\n",
      "  bucket 4: 60%\n",
      "  bucket 4: 30%\n",
      "  bucket 4: 70%\n",
      "  bucket 4: 40%\n",
      "  bucket 4: 80%\n",
      "  bucket 4: 50%\n",
      "  bucket 4: 90%\n",
      "  bucket 4: 60%\n",
      "  bucket 4: 100%\n",
      "  Sorting block of length 423248845 for bucket 4\n",
      "  (Using difference cover)\n",
      "  bucket 4: 70%\n",
      "  bucket 4: 80%\n",
      "  bucket 4: 90%\n",
      "  bucket 4: 100%\n",
      "  Sorting block of length 320071772 for bucket 4\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:05:37\n",
      "Returning block of 320071773 for bucket 4\n",
      "Getting block 5 of 7\n",
      "  Reserving size (531688794) for bucket 5\n",
      "  Calculating Z arrays for bucket 5\n",
      "  Entering block accumulator loop for bucket 5:\n",
      "  bucket 5: 10%\n",
      "  bucket 5: 20%\n",
      "  bucket 5: 30%\n",
      "  bucket 5: 40%\n",
      "  bucket 5: 50%\n",
      "  bucket 5: 60%\n",
      "  bucket 5: 70%\n",
      "  bucket 5: 80%\n",
      "  bucket 5: 90%\n",
      "  bucket 5: 100%\n",
      "  Sorting block of length 438201836 for bucket 5\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:10:04\n",
      "Returning block of 423248846 for bucket 4\n",
      "Getting block 5 of 7\n",
      "  Reserving size (531688794) for bucket 5\n",
      "  Calculating Z arrays for bucket 5\n",
      "  Entering block accumulator loop for bucket 5:\n",
      "  bucket 5: 10%\n",
      "  bucket 5: 20%\n",
      "  bucket 5: 30%\n",
      "  bucket 5: 40%\n",
      "  bucket 5: 50%\n",
      "  bucket 5: 60%\n",
      "  bucket 5: 70%\n",
      "  bucket 5: 80%\n",
      "  bucket 5: 90%\n",
      "  bucket 5: 100%\n",
      "  Sorting block of length 524804600 for bucket 5\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:08:37\n",
      "Returning block of 438201837 for bucket 5\n",
      "Getting block 6 of 7\n",
      "  Reserving size (531688794) for bucket 6\n",
      "  Calculating Z arrays for bucket 6\n",
      "  Entering block accumulator loop for bucket 6:\n",
      "  bucket 6: 10%\n",
      "  bucket 6: 20%\n",
      "  bucket 6: 30%\n",
      "  bucket 6: 40%\n",
      "  bucket 6: 50%\n",
      "  bucket 6: 60%\n",
      "  bucket 6: 70%\n",
      "  bucket 6: 80%\n",
      "  bucket 6: 90%\n",
      "  bucket 6: 100%\n",
      "  Sorting block of length 298794954 for bucket 6\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:05:27\n",
      "Returning block of 298794955 for bucket 6\n",
      "  Sorting block time: 00:13:21\n",
      "Returning block of 524804601 for bucket 5\n",
      "Getting block 7 of 7\n",
      "  Reserving size (531688794) for bucket 7\n",
      "  Calculating Z arrays for bucket 7\n",
      "  Entering block accumulator loop for bucket 7:\n",
      "  bucket 7: 10%\n",
      "  bucket 7: 20%\n",
      "  bucket 7: 30%\n",
      "  bucket 7: 40%\n",
      "  bucket 7: 50%\n",
      "  bucket 7: 60%\n",
      "  bucket 7: 70%\n",
      "  bucket 7: 80%\n",
      "  bucket 7: 90%\n",
      "  bucket 7: 100%\n",
      "  Sorting block of length 466539489 for bucket 7\n",
      "  (Using difference cover)\n",
      "Getting block 6 of 7\n",
      "  Reserving size (531688794) for bucket 6\n",
      "  Calculating Z arrays for bucket 6\n",
      "  Entering block accumulator loop for bucket 6:\n",
      "  bucket 6: 10%\n",
      "  bucket 6: 20%\n",
      "  bucket 6: 30%\n",
      "  bucket 6: 40%\n",
      "  bucket 6: 50%\n",
      "  bucket 6: 60%\n",
      "  bucket 6: 70%\n",
      "  bucket 6: 80%\n",
      "  bucket 6: 90%\n",
      "  bucket 6: 100%\n",
      "  Sorting block of length 483606073 for bucket 6\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:09:12\n",
      "Returning block of 466539490 for bucket 7\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 1417400275\n",
      "fchr[G]: 1997313856\n",
      "fchr[T]: 1997313856\n",
      "fchr[$]: 2835673565\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 949423420 bytes to primary EBWT file: BS_GA.1.bt2.tmp\n",
      "Wrote 708918396 bytes to secondary EBWT file: BS_GA.2.bt2.tmp\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 2835673565\n",
      "    bwtLen: 2835673566\n",
      "    sz: 708918392\n",
      "    bwtSz: 708918392\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 177229598\n",
      "    offsSz: 708918392\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 14769134\n",
      "    numLines: 14769134\n",
      "    ebwtTotLen: 945224576\n",
      "    ebwtTotSz: 945224576\n",
      "    color: 0\n",
      "    reverse: 0\n",
      "Total time for call to driver() for forward index: 01:22:01\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:14\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:17\n",
      "  Time to reverse reference sequence: 00:00:03\n",
      "bmax according to bmaxDivN setting: 708918391\n",
      "Using parameters --bmax 531688794 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 531688794 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:02:03\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Sorting block time: 00:12:14\n",
      "Returning block of 483606074 for bucket 6\n",
      "  Ranking v-sort output time: 00:00:33\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:52\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Getting block 7 of 7\n",
      "  Reserving size (531688794) for bucket 7\n",
      "  Calculating Z arrays for bucket 7\n",
      "  Entering block accumulator loop for bucket 7:\n",
      "  bucket 7: 10%\n",
      "  bucket 7: 20%\n",
      "  bucket 7: 30%\n",
      "  bucket 7: 40%\n",
      "  bucket 7: 50%\n",
      "  bucket 7: 60%\n",
      "  bucket 7: 70%\n",
      "  bucket 7: 80%\n",
      "  bucket 7: 90%\n",
      "  bucket 7: 100%\n",
      "  Sorting block of length 348909198 for bucket 7\n",
      "  (Using difference cover)\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 1, merged 6; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 3.54459e+08 (target: 531688793)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 8\n",
      "  Reserving size (531688794) for bucket 1\n",
      "  Calculating Z arrays for bucket 1\n",
      "  Entering block accumulator loop for bucket 1:\n",
      "  bucket 1: 10%\n",
      "  bucket 1: 20%\n",
      "  bucket 1: 30%\n",
      "  bucket 1: 40%\n",
      "  bucket 1: 50%\n",
      "  bucket 1: 60%\n",
      "  bucket 1: 70%\n",
      "  bucket 1: 80%\n",
      "  bucket 1: 90%\n",
      "  bucket 1: 100%\n",
      "  Sorting block of length 379344346 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:08:44\n",
      "Returning block of 348909199 for bucket 7\n",
      "  Sorting block time: 00:06:51\n",
      "Returning block of 379344347 for bucket 1\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 837195307\n",
      "fchr[G]: 837195307\n",
      "fchr[T]: 1417400275\n",
      "fchr[$]: 2835673565\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 949423420 bytes to primary EBWT file: BS_CT.1.bt2.tmp\n",
      "Wrote 708918396 bytes to secondary EBWT file: BS_CT.2.bt2.tmp\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 2835673565\n",
      "    bwtLen: 2835673566\n",
      "    sz: 708918392\n",
      "    bwtSz: 708918392\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 177229598\n",
      "    offsSz: 708918392\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 14769134\n",
      "    numLines: 14769134\n",
      "    ebwtTotLen: 945224576\n",
      "    ebwtTotSz: 945224576\n",
      "    color: 0\n",
      "    reverse: 0\n",
      "Total time for call to driver() for forward index: 01:38:20\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:13\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:16\n",
      "  Time to reverse reference sequence: 00:00:02\n",
      "bmax according to bmaxDivN setting: 708918391\n",
      "Using parameters --bmax 531688794 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 531688794 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "Getting block 2 of 8\n",
      "  Reserving size (531688794) for bucket 2\n",
      "  Calculating Z arrays for bucket 2\n",
      "  Entering block accumulator loop for bucket 2:\n",
      "  bucket 2: 10%\n",
      "  bucket 2: 20%\n",
      "  bucket 2: 30%\n",
      "  bucket 2: 40%\n",
      "  bucket 2: 50%\n",
      "  bucket 2: 60%\n",
      "  bucket 2: 70%\n",
      "  bucket 2: 80%\n",
      "  bucket 2: 90%\n",
      "  bucket 2: 100%\n",
      "  Sorting block of length 241732036 for bucket 2\n",
      "  (Using difference cover)\n",
      "  V-Sorting samples time: 00:01:58\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:33\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:51\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 4.05096e+08 (target: 531688793)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 7\n",
      "  Reserving size (531688794) for bucket 1\n",
      "  Calculating Z arrays for bucket 1\n",
      "  Entering block accumulator loop for bucket 1:\n",
      "  bucket 1: 10%\n",
      "  bucket 1: 20%\n",
      "  Sorting block time: 00:04:20\n",
      "Returning block of 241732037 for bucket 2\n",
      "  bucket 1: 30%\n",
      "  bucket 1: 40%\n",
      "  bucket 1: 50%\n",
      "  bucket 1: 60%\n",
      "  bucket 1: 70%\n",
      "  bucket 1: 80%\n",
      "  bucket 1: 90%\n",
      "  bucket 1: 100%\n",
      "  Sorting block of length 204720703 for bucket 1\n",
      "  (Using difference cover)\n",
      "Getting block 3 of 8\n",
      "  Reserving size (531688794) for bucket 3\n",
      "  Calculating Z arrays for bucket 3\n",
      "  Entering block accumulator loop for bucket 3:\n",
      "  bucket 3: 10%\n",
      "  bucket 3: 20%\n",
      "  bucket 3: 30%\n",
      "  bucket 3: 40%\n",
      "  bucket 3: 50%\n",
      "  bucket 3: 60%\n",
      "  bucket 3: 70%\n",
      "  bucket 3: 80%\n",
      "  bucket 3: 90%\n",
      "  bucket 3: 100%\n",
      "  Sorting block of length 329254955 for bucket 3\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:04:22\n",
      "Returning block of 204720704 for bucket 1\n",
      "Getting block 2 of 7\n",
      "  Reserving size (531688794) for bucket 2\n",
      "  Calculating Z arrays for bucket 2\n",
      "  Entering block accumulator loop for bucket 2:\n",
      "  bucket 2: 10%\n",
      "  bucket 2: 20%\n",
      "  bucket 2: 30%\n",
      "  bucket 2: 40%\n",
      "  bucket 2: 50%\n",
      "  bucket 2: 60%\n",
      "  bucket 2: 70%\n",
      "  bucket 2: 80%\n",
      "  bucket 2: 90%\n",
      "  bucket 2: 100%\n",
      "  Sorting block of length 528184092 for bucket 2\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:06:02\n",
      "Returning block of 329254956 for bucket 3\n",
      "Getting block 4 of 8\n",
      "  Reserving size (531688794) for bucket 4\n",
      "  Calculating Z arrays for bucket 4\n",
      "  Entering block accumulator loop for bucket 4:\n",
      "  bucket 4: 10%\n",
      "  bucket 4: 20%\n",
      "  bucket 4: 30%\n",
      "  bucket 4: 40%\n",
      "  bucket 4: 50%\n",
      "  bucket 4: 60%\n",
      "  bucket 4: 70%\n",
      "  bucket 4: 80%\n",
      "  bucket 4: 90%\n",
      "  bucket 4: 100%\n",
      "  Sorting block of length 473153477 for bucket 4\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:13:41\n",
      "Returning block of 528184093 for bucket 2\n",
      "  Sorting block time: 00:09:55\n",
      "Returning block of 473153478 for bucket 4\n",
      "Getting block 5 of 8\n",
      "  Reserving size (531688794) for bucket 5\n",
      "  Calculating Z arrays for bucket 5\n",
      "  Entering block accumulator loop for bucket 5:\n",
      "Getting block 3 of 7\n",
      "  Reserving size (531688794) for bucket 3\n",
      "  Calculating Z arrays for bucket 3\n",
      "  Entering block accumulator loop for bucket 3:\n",
      "  bucket 5: 10%\n",
      "  bucket 3: 10%\n",
      "  bucket 5: 20%\n",
      "  bucket 3: 20%\n",
      "  bucket 5: 30%\n",
      "  bucket 3: 30%\n",
      "  bucket 5: 40%\n",
      "  bucket 3: 40%\n",
      "  bucket 5: 50%\n",
      "  bucket 3: 50%\n",
      "  bucket 5: 60%\n",
      "  bucket 3: 60%\n",
      "  bucket 5: 70%\n",
      "  bucket 3: 70%\n",
      "  bucket 3: 80%\n",
      "  bucket 5: 80%\n",
      "  bucket 3: 90%\n",
      "  bucket 5: 90%\n",
      "  bucket 3: 100%\n",
      "  Sorting block of length 355209560 for bucket 3\n",
      "  (Using difference cover)\n",
      "  bucket 5: 100%\n",
      "  Sorting block of length 507042134 for bucket 5\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:08:40\n",
      "Returning block of 355209561 for bucket 3\n",
      "  Sorting block time: 00:10:02\n",
      "Returning block of 507042135 for bucket 5\n",
      "Getting block 4 of 7\n",
      "  Reserving size (531688794) for bucket 4\n",
      "  Calculating Z arrays for bucket 4\n",
      "  Entering block accumulator loop for bucket 4:\n",
      "  bucket 4: 10%\n",
      "  bucket 4: 20%\n",
      "  bucket 4: 30%\n",
      "  bucket 4: 40%\n",
      "  bucket 4: 50%\n",
      "  bucket 4: 60%\n",
      "  bucket 4: 70%\n",
      "  bucket 4: 80%\n",
      "  bucket 4: 90%\n",
      "  bucket 4: 100%\n",
      "  Sorting block of length 499679517 for bucket 4\n",
      "  (Using difference cover)\n",
      "Getting block 6 of 8\n",
      "  Reserving size (531688794) for bucket 6\n",
      "  Calculating Z arrays for bucket 6\n",
      "  Entering block accumulator loop for bucket 6:\n",
      "  bucket 6: 10%\n",
      "  bucket 6: 20%\n",
      "  bucket 6: 30%\n",
      "  bucket 6: 40%\n",
      "  bucket 6: 50%\n",
      "  bucket 6: 60%\n",
      "  bucket 6: 70%\n",
      "  bucket 6: 80%\n",
      "  bucket 6: 90%\n",
      "  bucket 6: 100%\n",
      "  Sorting block of length 184538363 for bucket 6\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:03:13\n",
      "Returning block of 184538364 for bucket 6\n",
      "Getting block 7 of 8\n",
      "  Reserving size (531688794) for bucket 7\n",
      "  Calculating Z arrays for bucket 7\n",
      "  Entering block accumulator loop for bucket 7:\n",
      "  bucket 7: 10%\n",
      "  bucket 7: 20%\n",
      "  bucket 7: 30%\n",
      "  bucket 7: 40%\n",
      "  bucket 7: 50%\n",
      "  bucket 7: 60%\n",
      "  bucket 7: 70%\n",
      "  bucket 7: 80%\n",
      "  bucket 7: 90%\n",
      "  bucket 7: 100%\n",
      "  Sorting block of length 349844215 for bucket 7\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:12:09\n",
      "Returning block of 499679518 for bucket 4\n",
      "  Sorting block time: 00:06:24\n",
      "Returning block of 349844216 for bucket 7\n",
      "Getting block 5 of 7\n",
      "  Reserving size (531688794) for bucket 5\n",
      "  Calculating Z arrays for bucket 5\n",
      "  Entering block accumulator loop for bucket 5:\n",
      "  bucket 5: 10%\n",
      "  bucket 5: 20%\n",
      "  bucket 5: 30%\n",
      "  bucket 5: 40%\n",
      "  bucket 5: 50%\n",
      "  bucket 5: 60%\n",
      "  bucket 5: 70%\n",
      "Getting block 8 of 8\n",
      "  Reserving size (531688794) for bucket 8\n",
      "  Calculating Z arrays for bucket 8\n",
      "  Entering block accumulator loop for bucket 8:\n",
      "  bucket 8: 10%\n",
      "  bucket 5: 80%\n",
      "  bucket 8: 20%\n",
      "  bucket 8: 30%\n",
      "  bucket 5: 90%\n",
      "  bucket 8: 40%\n",
      "  bucket 5: 100%\n",
      "  Sorting block of length 495690421 for bucket 5\n",
      "  (Using difference cover)\n",
      "  bucket 8: 50%\n",
      "  bucket 8: 60%\n",
      "  bucket 8: 70%\n",
      "  bucket 8: 80%\n",
      "  bucket 8: 90%\n",
      "  bucket 8: 100%\n",
      "  Sorting block of length 370764032 for bucket 8\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:06:45\n",
      "Returning block of 370764033 for bucket 8\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 1417400275\n",
      "fchr[G]: 1997313856\n",
      "fchr[T]: 1997313856\n",
      "fchr[$]: 2835673565\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 949423420 bytes to primary EBWT file: BS_GA.rev.1.bt2.tmp\n",
      "Wrote 708918396 bytes to secondary EBWT file: BS_GA.rev.2.bt2.tmp\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 2835673565\n",
      "    bwtLen: 2835673566\n",
      "    sz: 708918392\n",
      "    bwtSz: 708918392\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 177229598\n",
      "    offsSz: 708918392\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 14769134\n",
      "    numLines: 14769134\n",
      "    ebwtTotLen: 945224576\n",
      "    ebwtTotSz: 945224576\n",
      "    color: 0\n",
      "    reverse: 1\n",
      "Total time for backward call to driver() for mirror index: 01:22:22\n",
      "Renaming BS_GA.3.bt2.tmp to BS_GA.3.bt2\n",
      "Renaming BS_GA.4.bt2.tmp to BS_GA.4.bt2\n",
      "Renaming BS_GA.1.bt2.tmp to BS_GA.1.bt2\n",
      "Renaming BS_GA.2.bt2.tmp to BS_GA.2.bt2\n",
      "Renaming BS_GA.rev.1.bt2.tmp to BS_GA.rev.1.bt2\n",
      "Renaming BS_GA.rev.2.bt2.tmp to BS_GA.rev.2.bt2\n",
      "  Sorting block time: 00:11:52\n",
      "Returning block of 495690422 for bucket 5\n",
      "Getting block 6 of 7\n",
      "  Reserving size (531688794) for bucket 6\n",
      "  Calculating Z arrays for bucket 6\n",
      "  Entering block accumulator loop for bucket 6:\n",
      "  bucket 6: 10%\n",
      "  bucket 6: 20%\n",
      "  bucket 6: 30%\n",
      "  bucket 6: 40%\n",
      "  bucket 6: 50%\n",
      "  bucket 6: 60%\n",
      "  bucket 6: 70%\n",
      "  bucket 6: 80%\n",
      "  bucket 6: 90%\n",
      "  bucket 6: 100%\n",
      "  Sorting block of length 516265447 for bucket 6\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:12:44\n",
      "Returning block of 516265448 for bucket 6\n",
      "Getting block 7 of 7\n",
      "  Reserving size (531688794) for bucket 7\n",
      "  Calculating Z arrays for bucket 7\n",
      "  Entering block accumulator loop for bucket 7:\n",
      "  bucket 7: 10%\n",
      "  bucket 7: 20%\n",
      "  bucket 7: 30%\n",
      "  bucket 7: 40%\n",
      "  bucket 7: 50%\n",
      "  bucket 7: 60%\n",
      "  bucket 7: 70%\n",
      "  bucket 7: 80%\n",
      "  bucket 7: 90%\n",
      "  bucket 7: 100%\n",
      "  Sorting block of length 235923819 for bucket 7\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:05:52\n",
      "Returning block of 235923820 for bucket 7\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 837195307\n",
      "fchr[G]: 837195307\n",
      "fchr[T]: 1417400275\n",
      "fchr[$]: 2835673565\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 949423420 bytes to primary EBWT file: BS_CT.rev.1.bt2.tmp\n",
      "Wrote 708918396 bytes to secondary EBWT file: BS_CT.rev.2.bt2.tmp\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 2835673565\n",
      "    bwtLen: 2835673566\n",
      "    sz: 708918392\n",
      "    bwtSz: 708918392\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 177229598\n",
      "    offsSz: 708918392\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 14769134\n",
      "    numLines: 14769134\n",
      "    ebwtTotLen: 945224576\n",
      "    ebwtTotSz: 945224576\n",
      "    color: 0\n",
      "    reverse: 1\n",
      "Total time for backward call to driver() for mirror index: 01:35:32\n",
      "Renaming BS_CT.3.bt2.tmp to BS_CT.3.bt2\n",
      "Renaming BS_CT.4.bt2.tmp to BS_CT.4.bt2\n",
      "Renaming BS_CT.1.bt2.tmp to BS_CT.1.bt2\n",
      "Renaming BS_CT.2.bt2.tmp to BS_CT.2.bt2\n",
      "Renaming BS_CT.rev.1.bt2.tmp to BS_CT.rev.1.bt2\n",
      "Renaming BS_CT.rev.2.bt2.tmp to BS_CT.rev.2.bt2\n",
      "\n",
      "=========================================\n",
      "\n",
      "Parallel genome indexing complete. Enjoy!\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!/home/sagemaker-user/.conda/envs/rrbs-env/bin/bismark_genome_preparation --bowtie2 genome_hg19/\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e606b8c8-f534-4055-9789-4aa2d49b798f",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-04-20T08:05:54.326054Z",
     "iopub.status.busy": "2025-04-20T08:05:54.325654Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bowtie 2 seems to be working fine (tested command 'bowtie2 --version' [2.5.4])\n",
      "Output format is BAM (default)\n",
      "Alignments will be written out in BAM format. Samtools found here: '/home/sagemaker-user/.conda/envs/rrbs-env/bin/samtools'\n",
      "Reference genome folder provided is genome_hg19/\t(absolute path is '/home/sagemaker-user/src/genome_hg19/)'\n",
      "FastQ format assumed (by default)\n",
      "Attention: early reports suggested that high values of -p  to have diminishing returns. Please test different values using a small subset of data for your hardware setting.\n",
      "Each Bowtie 2 instance is going to be run with 32 threads. Please monitor performance closely and tune down if necessary!\n",
      "\n",
      "Input files to be analysed (in current folder '/home/sagemaker-user/src'):\n",
      "ENCFF000MBM_trimmed.fq.gz\n",
      "Library is assumed to be strand-specific (directional), alignments to strands complementary to the original top or bottom strands will be ignored (i.e. not performed!)\n",
      "Setting parallelization to single-threaded (default)\n",
      "\n",
      "Summary of all aligner options:\t-q --score-min L,0,-0.2 -p 32 --reorder --ignore-quals\n",
      "Current working directory is: /home/sagemaker-user/src\n",
      "\n",
      "Now reading in and storing sequence information of the genome specified in: /home/sagemaker-user/src/genome_hg19/\n",
      "\n",
      "chr chr1 (249250621 bp)\n",
      "chr chr2 (243199373 bp)\n",
      "chr chr3 (198022430 bp)\n",
      "chr chr4 (191154276 bp)\n",
      "chr chr5 (180915260 bp)\n",
      "chr chr6 (171115067 bp)\n",
      "chr chr7 (159138663 bp)\n",
      "chr chrX (155270560 bp)\n",
      "chr chr8 (146364022 bp)\n",
      "chr chr9 (141213431 bp)\n",
      "chr chr10 (135534747 bp)\n",
      "chr chr11 (135006516 bp)\n",
      "chr chr12 (133851895 bp)\n",
      "chr chr13 (115169878 bp)\n",
      "chr chr14 (107349540 bp)\n",
      "chr chr15 (102531392 bp)\n",
      "chr chr16 (90354753 bp)\n",
      "chr chr17 (81195210 bp)\n",
      "chr chr18 (78077248 bp)\n",
      "chr chr20 (63025520 bp)\n",
      "chr chr19 (59128983 bp)\n",
      "chr chr22 (51304566 bp)\n",
      "chr chr21 (48129895 bp)\n",
      "\n",
      "Single-core mode: setting pid to 1\n",
      "\n",
      "Single-end alignments will be performed\n",
      "=======================================\n",
      "\n",
      "Input file is in FastQ format\n",
      "Writing a C -> T converted version of the input file ENCFF000MBM_trimmed.fq.gz to ENCFF000MBM_trimmed.fq.gz_C_to_T.fastq\n",
      "\n",
      "Created C -> T converted version of the FastQ file ENCFF000MBM_trimmed.fq.gz (28824892 sequences in total)\n",
      "\n",
      "Input file is ENCFF000MBM_trimmed.fq.gz_C_to_T.fastq (FastQ)\n",
      "\n",
      "Now running 2 instances of Bowtie 2 against the bisulfite genome of /home/sagemaker-user/src/genome_hg19/ with the specified options: -q --score-min L,0,-0.2 -p 32 --reorder --ignore-quals\n",
      "\n",
      "Now starting the Bowtie 2 aligner for CTreadCTgenome (reading in sequences from ENCFF000MBM_trimmed.fq.gz_C_to_T.fastq with options -q --score-min L,0,-0.2 -p 32 --reorder --ignore-quals --norc)\n",
      "Using Bowtie 2 index: /home/sagemaker-user/src/genome_hg19/Bisulfite_Genome/CT_conversion/BS_CT\n",
      "\n",
      "Found first alignment:\tHWUSI-EAS1758_7:1:1:996:16820:0:1:0\t4\t*\t0\t0\t*\t*\t0\t0\tNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN\tBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB\tYT:Z:UU\tYF:Z:NS\n",
      "Now starting the Bowtie 2 aligner for CTreadGAgenome (reading in sequences from ENCFF000MBM_trimmed.fq.gz_C_to_T.fastq with options -q --score-min L,0,-0.2 -p 32 --reorder --ignore-quals --nofw)\n",
      "Using Bowtie 2 index: /home/sagemaker-user/src/genome_hg19/Bisulfite_Genome/GA_conversion/BS_GA\n",
      "\n",
      "Found first alignment:\tHWUSI-EAS1758_7:1:1:996:16820:0:1:0\t4\t*\t0\t0\t*\t*\t0\t0\tNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN\tBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB\tYT:Z:UU\tYF:Z:NS\n",
      "\n",
      ">>> Writing bisulfite mapping results to ENCFF000MBM_trimmed_bismark_bt2.bam <<<\n",
      "\n",
      "\n",
      "Reading in the sequence file ENCFF000MBM_trimmed.fq.gz\n",
      "Processed 1000000 sequences so far\n",
      "Processed 2000000 sequences so far\n",
      "Processed 3000000 sequences so far\n",
      "Processed 4000000 sequences so far\n",
      "Processed 5000000 sequences so far\n",
      "Processed 6000000 sequences so far\n",
      "Processed 7000000 sequences so far\n",
      "Processed 8000000 sequences so far\n",
      "Processed 9000000 sequences so far\n",
      "Processed 10000000 sequences so far\n",
      "Processed 11000000 sequences so far\n",
      "Processed 12000000 sequences so far\n",
      "Processed 13000000 sequences so far\n",
      "Processed 14000000 sequences so far\n",
      "Processed 15000000 sequences so far\n",
      "Processed 16000000 sequences so far\n",
      "Processed 17000000 sequences so far\n",
      "Processed 18000000 sequences so far\n",
      "Processed 19000000 sequences so far\n",
      "Processed 20000000 sequences so far\n",
      "Processed 21000000 sequences so far\n",
      "Processed 22000000 sequences so far\n",
      "Processed 23000000 sequences so far\n",
      "Processed 24000000 sequences so far\n",
      "Processed 25000000 sequences so far\n",
      "Processed 26000000 sequences so far\n",
      "Processed 27000000 sequences so far\n",
      "Processed 28000000 sequences so far\n",
      "28824892 reads; of these:\n",
      "  28824892 (100.00%) were unpaired; of these:\n",
      "    13000293 (45.10%) aligned 0 times\n",
      "    7769562 (26.95%) aligned exactly 1 time\n",
      "    8055037 (27.94%) aligned >1 times\n",
      "54.90% overall alignment rate\n",
      "28824892 reads; of these:\n",
      "  28824892 (100.00%) were unpaired; of these:\n",
      "    13061833 (45.31%) aligned 0 times\n",
      "    7775924 (26.98%) aligned exactly 1 time\n",
      "    7987135 (27.71%) aligned >1 times\n",
      "54.69% overall alignment rate\n",
      "Processed 28824892 sequences in total\n",
      "\n",
      "\n",
      "Successfully deleted the temporary file ENCFF000MBM_trimmed.fq.gz_C_to_T.fastq\n",
      "\n",
      "Final Alignment report\n",
      "======================\n",
      "Sequences analysed in total:\t28824892\n",
      "Number of alignments with a unique best hit from the different alignments:\t15398242\n",
      "Mapping efficiency:\t53.4%\n",
      "\n",
      "Sequences with no alignments under any condition:\t5216803\n",
      "Sequences did not map uniquely:\t8209847\n",
      "Sequences which were discarded because genomic sequence could not be extracted:\t0\n",
      "\n",
      "Number of sequences with unique best (first) alignment came from the bowtie output:\n",
      "CT/CT:\t7709489\t((converted) top strand)\n",
      "CT/GA:\t7688753\t((converted) bottom strand)\n",
      "GA/CT:\t0\t(complementary to (converted) top strand)\n",
      "GA/GA:\t0\t(complementary to (converted) bottom strand)\n",
      "\n",
      "Number of alignments to (merely theoretical) complementary strands being rejected in total:\t0\n",
      "\n",
      "Final Cytosine Methylation Report\n",
      "=================================\n",
      "Total number of C's analysed:\t168840929\n",
      "\n",
      "Total methylated C's in CpG context:\t17046158\n",
      "Total methylated C's in CHG context:\t413004\n",
      "Total methylated C's in CHH context:\t750124\n",
      "Total methylated C's in Unknown context:\t1\n",
      "\n",
      "Total unmethylated C's in CpG context:\t37704048\n",
      "Total unmethylated C's in CHG context:\t40353163\n",
      "Total unmethylated C's in CHH context:\t72574432\n",
      "Total unmethylated C's in Unknown context:\t19\n",
      "\n",
      "C methylated in CpG context:\t31.1%\n",
      "C methylated in CHG context:\t1.0%\n",
      "C methylated in CHH context:\t1.0%\n",
      "C methylated in Unknown context (CN or CHN):\t5.0%\n",
      "\n",
      "\n",
      "Bismark completed in 0d 1h 16m 39s\n",
      "\n",
      "====================\n",
      "Bismark run complete\n",
      "====================\n",
      "\n",
      "Bowtie 2 seems to be working fine (tested command 'bowtie2 --version' [2.5.4])\n",
      "Output format is BAM (default)\n",
      "Alignments will be written out in BAM format. Samtools found here: '/home/sagemaker-user/.conda/envs/rrbs-env/bin/samtools'\n",
      "Reference genome folder provided is genome_hg19/\t(absolute path is '/home/sagemaker-user/src/genome_hg19/)'\n",
      "FastQ format assumed (by default)\n",
      "Attention: early reports suggested that high values of -p  to have diminishing returns. Please test different values using a small subset of data for your hardware setting.\n",
      "Each Bowtie 2 instance is going to be run with 32 threads. Please monitor performance closely and tune down if necessary!\n",
      "\n",
      "Input files to be analysed (in current folder '/home/sagemaker-user/src'):\n",
      "ENCFF000MBV_trimmed.fq.gz\n",
      "Library is assumed to be strand-specific (directional), alignments to strands complementary to the original top or bottom strands will be ignored (i.e. not performed!)\n",
      "Setting parallelization to single-threaded (default)\n",
      "\n",
      "Summary of all aligner options:\t-q --score-min L,0,-0.2 -p 32 --reorder --ignore-quals\n",
      "Current working directory is: /home/sagemaker-user/src\n",
      "\n",
      "Now reading in and storing sequence information of the genome specified in: /home/sagemaker-user/src/genome_hg19/\n",
      "\n",
      "chr chr1 (249250621 bp)\n",
      "chr chr2 (243199373 bp)\n",
      "chr chr3 (198022430 bp)\n",
      "chr chr4 (191154276 bp)\n",
      "chr chr5 (180915260 bp)\n",
      "chr chr6 (171115067 bp)\n",
      "chr chr7 (159138663 bp)\n",
      "chr chrX (155270560 bp)\n",
      "chr chr8 (146364022 bp)\n",
      "chr chr9 (141213431 bp)\n",
      "chr chr10 (135534747 bp)\n",
      "chr chr11 (135006516 bp)\n",
      "chr chr12 (133851895 bp)\n",
      "chr chr13 (115169878 bp)\n",
      "chr chr14 (107349540 bp)\n",
      "chr chr15 (102531392 bp)\n",
      "chr chr16 (90354753 bp)\n",
      "chr chr17 (81195210 bp)\n",
      "chr chr18 (78077248 bp)\n",
      "chr chr20 (63025520 bp)\n",
      "chr chr19 (59128983 bp)\n",
      "chr chr22 (51304566 bp)\n",
      "chr chr21 (48129895 bp)\n",
      "\n",
      "Single-core mode: setting pid to 1\n",
      "\n",
      "Single-end alignments will be performed\n",
      "=======================================\n",
      "\n",
      "Input file is in FastQ format\n",
      "Writing a C -> T converted version of the input file ENCFF000MBV_trimmed.fq.gz to ENCFF000MBV_trimmed.fq.gz_C_to_T.fastq\n"
     ]
    }
   ],
   "source": [
    "# Sample 1: ENCFF000MBM\n",
    "!/home/sagemaker-user/.conda/envs/rrbs-env/bin/bismark \\\n",
    "  --genome genome_hg19/ \\\n",
    "  -p 32 \\\n",
    "  ENCFF000MBM_trimmed.fq.gz\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4271544b-5dc9-4b1d-af66-2714e3a28d89",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-04-20T17:06:35.710857Z",
     "iopub.status.busy": "2025-04-20T17:06:35.710241Z",
     "iopub.status.idle": "2025-04-20T19:13:11.436816Z",
     "shell.execute_reply": "2025-04-20T19:13:11.436120Z",
     "shell.execute_reply.started": "2025-04-20T17:06:35.710835Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bowtie 2 seems to be working fine (tested command 'bowtie2 --version' [2.5.4])\n",
      "Output format is BAM (default)\n",
      "Alignments will be written out in BAM format. Samtools found here: '/home/sagemaker-user/.conda/envs/rrbs-env/bin/samtools'\n",
      "Reference genome folder provided is genome_hg19/\t(absolute path is '/home/sagemaker-user/src/genome_hg19/)'\n",
      "FastQ format assumed (by default)\n",
      "Attention: early reports suggested that high values of -p  to have diminishing returns. Please test different values using a small subset of data for your hardware setting.\n",
      "Each Bowtie 2 instance is going to be run with 32 threads. Please monitor performance closely and tune down if necessary!\n",
      "\n",
      "Input files to be analysed (in current folder '/home/sagemaker-user/src'):\n",
      "ENCFF000MBV_trimmed.fq.gz\n",
      "Library is assumed to be strand-specific (directional), alignments to strands complementary to the original top or bottom strands will be ignored (i.e. not performed!)\n",
      "Setting parallelization to single-threaded (default)\n",
      "\n",
      "Summary of all aligner options:\t-q --score-min L,0,-0.2 -p 32 --reorder --ignore-quals\n",
      "Current working directory is: /home/sagemaker-user/src\n",
      "\n",
      "Now reading in and storing sequence information of the genome specified in: /home/sagemaker-user/src/genome_hg19/\n",
      "\n",
      "chr chr1 (249250621 bp)\n",
      "chr chr2 (243199373 bp)\n",
      "chr chr3 (198022430 bp)\n",
      "chr chr4 (191154276 bp)\n",
      "chr chr5 (180915260 bp)\n",
      "chr chr6 (171115067 bp)\n",
      "chr chr7 (159138663 bp)\n",
      "chr chrX (155270560 bp)\n",
      "chr chr8 (146364022 bp)\n",
      "chr chr9 (141213431 bp)\n",
      "chr chr10 (135534747 bp)\n",
      "chr chr11 (135006516 bp)\n",
      "chr chr12 (133851895 bp)\n",
      "chr chr13 (115169878 bp)\n",
      "chr chr14 (107349540 bp)\n",
      "chr chr15 (102531392 bp)\n",
      "chr chr16 (90354753 bp)\n",
      "chr chr17 (81195210 bp)\n",
      "chr chr18 (78077248 bp)\n",
      "chr chr20 (63025520 bp)\n",
      "chr chr19 (59128983 bp)\n",
      "chr chr22 (51304566 bp)\n",
      "chr chr21 (48129895 bp)\n",
      "\n",
      "Single-core mode: setting pid to 1\n",
      "\n",
      "Single-end alignments will be performed\n",
      "=======================================\n",
      "\n",
      "Input file is in FastQ format\n",
      "Writing a C -> T converted version of the input file ENCFF000MBV_trimmed.fq.gz to ENCFF000MBV_trimmed.fq.gz_C_to_T.fastq\n",
      "\n",
      "Created C -> T converted version of the FastQ file ENCFF000MBV_trimmed.fq.gz (53879796 sequences in total)\n",
      "\n",
      "Input file is ENCFF000MBV_trimmed.fq.gz_C_to_T.fastq (FastQ)\n",
      "\n",
      "Now running 2 instances of Bowtie 2 against the bisulfite genome of /home/sagemaker-user/src/genome_hg19/ with the specified options: -q --score-min L,0,-0.2 -p 32 --reorder --ignore-quals\n",
      "\n",
      "Now starting the Bowtie 2 aligner for CTreadCTgenome (reading in sequences from ENCFF000MBV_trimmed.fq.gz_C_to_T.fastq with options -q --score-min L,0,-0.2 -p 32 --reorder --ignore-quals --norc)\n",
      "Using Bowtie 2 index: /home/sagemaker-user/src/genome_hg19/Bisulfite_Genome/CT_conversion/BS_CT\n",
      "\n",
      "Found first alignment:\tILLUMINA-AFD1A9_7:1:1:0:51:0:1:0\t4\t*\t0\t0\t*\t*\t0\t0\tNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN\tBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB\tYT:Z:UU\tYF:Z:NS\n",
      "Now starting the Bowtie 2 aligner for CTreadGAgenome (reading in sequences from ENCFF000MBV_trimmed.fq.gz_C_to_T.fastq with options -q --score-min L,0,-0.2 -p 32 --reorder --ignore-quals --nofw)\n",
      "Using Bowtie 2 index: /home/sagemaker-user/src/genome_hg19/Bisulfite_Genome/GA_conversion/BS_GA\n",
      "\n",
      "Found first alignment:\tILLUMINA-AFD1A9_7:1:1:0:51:0:1:0\t4\t*\t0\t0\t*\t*\t0\t0\tNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN\tBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB\tYT:Z:UU\tYF:Z:NS\n",
      "\n",
      ">>> Writing bisulfite mapping results to ENCFF000MBV_trimmed_bismark_bt2.bam <<<\n",
      "\n",
      "\n",
      "Reading in the sequence file ENCFF000MBV_trimmed.fq.gz\n",
      "Processed 1000000 sequences so far\n",
      "Processed 2000000 sequences so far\n",
      "Processed 3000000 sequences so far\n",
      "Processed 4000000 sequences so far\n",
      "Processed 5000000 sequences so far\n",
      "Processed 6000000 sequences so far\n",
      "Processed 7000000 sequences so far\n",
      "Processed 8000000 sequences so far\n",
      "Processed 9000000 sequences so far\n",
      "Processed 10000000 sequences so far\n",
      "Processed 11000000 sequences so far\n",
      "Processed 12000000 sequences so far\n",
      "Processed 13000000 sequences so far\n",
      "Processed 14000000 sequences so far\n",
      "Processed 15000000 sequences so far\n",
      "Processed 16000000 sequences so far\n",
      "Processed 17000000 sequences so far\n",
      "Processed 18000000 sequences so far\n",
      "Processed 19000000 sequences so far\n",
      "Processed 20000000 sequences so far\n",
      "Processed 21000000 sequences so far\n",
      "Processed 22000000 sequences so far\n",
      "Processed 23000000 sequences so far\n",
      "Processed 24000000 sequences so far\n",
      "Processed 25000000 sequences so far\n",
      "Processed 26000000 sequences so far\n",
      "Processed 27000000 sequences so far\n",
      "Processed 28000000 sequences so far\n",
      "Processed 29000000 sequences so far\n",
      "Processed 30000000 sequences so far\n",
      "Processed 31000000 sequences so far\n",
      "Processed 32000000 sequences so far\n",
      "Processed 33000000 sequences so far\n",
      "Processed 34000000 sequences so far\n",
      "Processed 35000000 sequences so far\n",
      "Processed 36000000 sequences so far\n",
      "Processed 37000000 sequences so far\n",
      "Processed 38000000 sequences so far\n",
      "Processed 39000000 sequences so far\n",
      "Processed 40000000 sequences so far\n",
      "Processed 41000000 sequences so far\n",
      "Processed 42000000 sequences so far\n",
      "Processed 43000000 sequences so far\n",
      "Processed 44000000 sequences so far\n",
      "Processed 45000000 sequences so far\n",
      "Processed 46000000 sequences so far\n",
      "Processed 47000000 sequences so far\n",
      "Processed 48000000 sequences so far\n",
      "Processed 49000000 sequences so far\n",
      "Processed 50000000 sequences so far\n",
      "Processed 51000000 sequences so far\n",
      "Processed 52000000 sequences so far\n",
      "Processed 53000000 sequences so far\n",
      "53879796 reads; of these:\n",
      "  53879796 (100.00%) were unpaired; of these:\n",
      "    29433072 (54.63%) aligned 0 times\n",
      "    11921137 (22.13%) aligned exactly 1 time\n",
      "    12525587 (23.25%) aligned >1 times\n",
      "45.37% overall alignment rate\n",
      "53879796 reads; of these:\n",
      "  53879796 (100.00%) were unpaired; of these:\n",
      "    29314904 (54.41%) aligned 0 times\n",
      "    11889578 (22.07%) aligned exactly 1 time\n",
      "    12675314 (23.53%) aligned >1 times\n",
      "45.59% overall alignment rate\n",
      "Processed 53879796 sequences in total\n",
      "\n",
      "\n",
      "Successfully deleted the temporary file ENCFF000MBV_trimmed.fq.gz_C_to_T.fastq\n",
      "\n",
      "Final Alignment report\n",
      "======================\n",
      "Sequences analysed in total:\t53879796\n",
      "Number of alignments with a unique best hit from the different alignments:\t23439852\n",
      "Mapping efficiency:\t43.5%\n",
      "\n",
      "Sequences with no alignments under any condition:\t17558643\n",
      "Sequences did not map uniquely:\t12881301\n",
      "Sequences which were discarded because genomic sequence could not be extracted:\t0\n",
      "\n",
      "Number of sequences with unique best (first) alignment came from the bowtie output:\n",
      "CT/CT:\t11735792\t((converted) top strand)\n",
      "CT/GA:\t11704060\t((converted) bottom strand)\n",
      "GA/CT:\t0\t(complementary to (converted) top strand)\n",
      "GA/GA:\t0\t(complementary to (converted) bottom strand)\n",
      "\n",
      "Number of alignments to (merely theoretical) complementary strands being rejected in total:\t0\n",
      "\n",
      "Final Cytosine Methylation Report\n",
      "=================================\n",
      "Total number of C's analysed:\t254924263\n",
      "\n",
      "Total methylated C's in CpG context:\t22768173\n",
      "Total methylated C's in CHG context:\t491864\n",
      "Total methylated C's in CHH context:\t1157000\n",
      "Total methylated C's in Unknown context:\t2\n",
      "\n",
      "Total unmethylated C's in CpG context:\t47615815\n",
      "Total unmethylated C's in CHG context:\t58709101\n",
      "Total unmethylated C's in CHH context:\t124182310\n",
      "Total unmethylated C's in Unknown context:\t103\n",
      "\n",
      "C methylated in CpG context:\t32.3%\n",
      "C methylated in CHG context:\t0.8%\n",
      "C methylated in CHH context:\t0.9%\n",
      "C methylated in Unknown context (CN or CHN):\t1.9%\n",
      "\n",
      "\n",
      "Bismark completed in 0d 2h 6m 36s\n",
      "\n",
      "====================\n",
      "Bismark run complete\n",
      "====================\n",
      "\n",
      "Finished\n"
     ]
    }
   ],
   "source": [
    "# Sample 2: ENCFF000MBV\n",
    "!/home/sagemaker-user/.conda/envs/rrbs-env/bin/bismark \\\n",
    "  --genome genome_hg19/ \\\n",
    "  -p 32 \\\n",
    "  ENCFF000MBV_trimmed.fq.gz\n",
    "!echo \"Finished\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f4dbb362-b834-47ce-adbb-212471d1a710",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-04-21T07:03:04.944980Z",
     "iopub.status.busy": "2025-04-21T07:03:04.944637Z",
     "iopub.status.idle": "2025-04-21T07:03:05.100216Z",
     "shell.execute_reply": "2025-04-21T07:03:05.099498Z",
     "shell.execute_reply.started": "2025-04-21T07:03:04.944957Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENCFF000MBM_trimmed_bismark_bt2.bam  ENCFF000MBV_trimmed_bismark_bt2.bam\n"
     ]
    }
   ],
   "source": [
    "!ls *.bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "04999c2c-ef16-4503-966f-8b659191ce83",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-04-21T07:22:36.189100Z",
     "iopub.status.busy": "2025-04-21T07:22:36.188764Z",
     "iopub.status.idle": "2025-04-21T07:23:16.618793Z",
     "shell.execute_reply": "2025-04-21T07:23:16.618121Z",
     "shell.execute_reply.started": "2025-04-21T07:22:36.189075Z"
    }
   },
   "outputs": [],
   "source": [
    "!/home/sagemaker-user/.conda/envs/rrbs-env/bin/samtools merge -@ 34 merged.bam ENCFF000MBM_trimmed_bismark_bt2.bam ENCFF000MBV_trimmed_bismark_bt2.bam\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "f799acbd-80c4-43fa-b400-f5b98e8e7104",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-04-21T08:21:19.427590Z",
     "iopub.status.busy": "2025-04-21T08:21:19.426970Z",
     "iopub.status.idle": "2025-04-21T08:21:41.761800Z",
     "shell.execute_reply": "2025-04-21T08:21:41.761074Z",
     "shell.execute_reply.started": "2025-04-21T08:21:19.427554Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[bam_sort_core] merging from 0 files and 34 in-memory blocks...\n"
     ]
    }
   ],
   "source": [
    "!/home/sagemaker-user/.conda/envs/rrbs-env/bin/samtools sort -@ 34 -o merged.sorted.bam merged.bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "dc134fe4-fdb2-42d4-aca7-13ce116842f4",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-04-21T08:22:53.272310Z",
     "iopub.status.busy": "2025-04-21T08:22:53.271954Z",
     "iopub.status.idle": "2025-04-21T08:23:03.483614Z",
     "shell.execute_reply": "2025-04-21T08:23:03.482818Z",
     "shell.execute_reply.started": "2025-04-21T08:22:53.272288Z"
    }
   },
   "outputs": [],
   "source": [
    "!/home/sagemaker-user/.conda/envs/rrbs-env/bin/samtools index merged.sorted.bam\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "89a17f80-6d0f-4d81-afb8-312bbd556ac6",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-04-24T14:24:29.525428Z",
     "iopub.status.busy": "2025-04-24T14:24:29.525081Z",
     "iopub.status.idle": "2025-04-24T14:34:40.711305Z",
     "shell.execute_reply": "2025-04-24T14:34:40.710223Z",
     "shell.execute_reply.started": "2025-04-24T14:24:29.525406Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " *** Bismark methylation extractor version v0.24.2 ***\n",
      "\n",
      "Trying to determine the type of mapping from the SAM header line of file merged.sorted.bam\n",
      "Treating file(s) as single-end data (as extracted from @PG line)\n",
      "\n",
      "Treating file(s) as single-end data (as extracted from @PG line)\n",
      "\n",
      "Core usage currently set to more than 20 threads. Let's see how this goes... (set value: 34)\n",
      "\n",
      "Summarising Bismark methylation extractor parameters:\n",
      "===============================================================\n",
      "Bismark single-end SAM format specified (default)\n",
      "Number of cores to be used: 34\n",
      "Output will be written to the current directory ('/home/sagemaker-user/src')\n",
      "\n",
      "\n",
      "Summarising bedGraph parameters:\n",
      "===============================================================\n",
      "Generating additional output in bedGraph and coverage format\n",
      "bedGraph format:\t<Chromosome> <Start Position> <End Position> <Methylation Percentage>\n",
      "coverage format:\t<Chromosome> <Start Position> <End Position> <Methylation Percentage> <count methylated> <count non-methylated>\n",
      "\n",
      "Using a cutoff of 1 read(s) to report cytosine positions\n",
      "Reporting and sorting cytosine methylation information in CpG context only (default)\n",
      "The bedGraph UNIX sort command will use the following memory setting:\t'2G'. Temporary directory used for sorting is the output directory\n",
      "\n",
      "Checking file >>merged.sorted.bam<< for signs of file truncation...\n",
      "\n",
      "Writing result file containing methylation information for C in CpG context from the original top strand to CpG_OT_merged.sorted.txt.gz\n",
      "Writing result file containing methylation information for C in CpG context from the complementary to original top strand to CpG_CTOT_merged.sorted.txt.gz\n",
      "Writing result file containing methylation information for C in CpG context from the complementary to original bottom strand to CpG_CTOB_merged.sorted.txt.gz\n",
      "Writing result file containing methylation information for C in CpG context from the original bottom strand to CpG_OB_merged.sorted.txt.gz\n",
      "\n",
      "Writing result file containing methylation information for C in CHG context from the original top strand to CHG_OT_merged.sorted.txt.gz\n",
      "Writing result file containing methylation information for C in CHG context from the complementary to original top strand to CHG_CTOT_merged.sorted.txt.gz\n",
      "Writing result file containing methylation information for C in CHG context from the complementary to original bottom strand to CHG_CTOB_merged.sorted.txt.gz\n",
      "Writing result file containing methylation information for C in CHG context from the original bottom strand to CHG_OB_merged.sorted.txt.gz\n",
      "\n",
      "Writing result file containing methylation information for C in CHH context from the original top strand to CHH_OT_merged.sorted.txt.gz\n",
      "Writing result file containing methylation information for C in CHH context from the complementary to original top strand to CHH_CTOT_merged.sorted.txt.gz\n",
      "Writing result file containing methylation information for C in CHH context from the complementary to original bottom strand to CHH_CTOB_merged.sorted.txt.gz\n",
      "Writing result file containing methylation information for C in CHH context from the original bottom strand to CHH_OB_merged.sorted.txt.gz\n",
      "\n",
      "\n",
      "Now reading in Bismark result file merged.sorted.bam\n",
      "\n",
      "\n",
      "Now reading in Bismark result file merged.sorted.bam\n",
      "\n",
      "\n",
      "Now reading in Bismark result file merged.sorted.bam\n",
      "\n",
      "\n",
      "Now reading in Bismark result file merged.sorted.bam\n",
      "\n",
      "\n",
      "Now reading in Bismark result file merged.sorted.bam\n",
      "\n",
      "\n",
      "Now reading in Bismark result file merged.sorted.bam\n",
      "\n",
      "\n",
      "Now reading in Bismark result file merged.sorted.bam\n",
      "\n",
      "\n",
      "Now reading in Bismark result file merged.sorted.bam\n",
      "\n",
      "\n",
      "Now reading in Bismark result file merged.sorted.bam\n",
      "\n",
      "\n",
      "Now reading in Bismark result file merged.sorted.bam\n",
      "\n",
      "\n",
      "Now reading in Bismark result file merged.sorted.bam\n",
      "\n",
      "\n",
      "Now reading in Bismark result file merged.sorted.bam\n",
      "\n",
      "\n",
      "Now reading in Bismark result file merged.sorted.bam\n",
      "\n",
      "\n",
      "Now reading in Bismark result file merged.sorted.bam\n",
      "\n",
      "\n",
      "Now reading in Bismark result file merged.sorted.bam\n",
      "\n",
      "\n",
      "Now reading in Bismark result file merged.sorted.bam\n",
      "\n",
      "\n",
      "Now reading in Bismark result file merged.sorted.bam\n",
      "\n",
      "\n",
      "Now reading in Bismark result file merged.sorted.bam\n",
      "\n",
      "\n",
      "Now reading in Bismark result file merged.sorted.bam\n",
      "\n",
      "\n",
      "Now reading in Bismark result file merged.sorted.bam\n",
      "\n",
      "\n",
      "Now reading in Bismark result file merged.sorted.bam\n",
      "\n",
      "\n",
      "Now reading in Bismark result file merged.sorted.bam\n",
      "\n",
      "\n",
      "Now reading in Bismark result file merged.sorted.bam\n",
      "\n",
      "\n",
      "Now reading in Bismark result file merged.sorted.bam\n",
      "\n",
      "\n",
      "Now reading in Bismark result file merged.sorted.bam\n",
      "\n",
      "\n",
      "Now reading in Bismark result file merged.sorted.bam\n",
      "\n",
      "\n",
      "Now reading in Bismark result file merged.sorted.bam\n",
      "\n",
      "\n",
      "Now reading in Bismark result file merged.sorted.bam\n",
      "\n",
      "\n",
      "Now reading in Bismark result file merged.sorted.bam\n",
      "\n",
      "\n",
      "Now reading in Bismark result file merged.sorted.bam\n",
      "\n",
      "Now reading in Bismark result file merged.sorted.bam\n",
      "\n",
      "Now reading in Bismark result file merged.sorted.bam\n",
      "\n",
      "\n",
      "Now reading in Bismark result file merged.sorted.bam\n",
      "\n",
      "\n",
      "Now reading in Bismark result file merged.sorted.bam\n",
      "\n",
      "skipping SAM header line:\t@HD\tVN:1.0\tSO:coordinate\n",
      "skipping SAM header line:\t@SQ\tSN:chr1\tLN:249250621\n",
      "skipping SAM header line:\t@SQ\tSN:chr2\tLN:243199373\n",
      "skipping SAM header line:\t@SQ\tSN:chr3\tLN:198022430\n",
      "skipping SAM header line:\t@SQ\tSN:chr4\tLN:191154276\n",
      "skipping SAM header line:\t@SQ\tSN:chr5\tLN:180915260\n",
      "skipping SAM header line:\t@SQ\tSN:chr6\tLN:171115067\n",
      "skipping SAM header line:\t@SQ\tSN:chr7\tLN:159138663\n",
      "skipping SAM header line:\t@SQ\tSN:chrX\tLN:155270560\n",
      "skipping SAM header line:\t@SQ\tSN:chr8\tLN:146364022\n",
      "skipping SAM header line:\t@SQ\tSN:chr9\tLN:141213431\n",
      "skipping SAM header line:\t@SQ\tSN:chr10\tLN:135534747\n",
      "skipping SAM header line:\t@SQ\tSN:chr11\tLN:135006516\n",
      "skipping SAM header line:\t@SQ\tSN:chr12\tLN:133851895\n",
      "skipping SAM header line:\t@SQ\tSN:chr13\tLN:115169878\n",
      "skipping SAM header line:\t@SQ\tSN:chr14\tLN:107349540\n",
      "skipping SAM header line:\t@SQ\tSN:chr15\tLN:102531392\n",
      "skipping SAM header line:\t@SQ\tSN:chr16\tLN:90354753\n",
      "skipping SAM header line:\t@SQ\tSN:chr17\tLN:81195210\n",
      "skipping SAM header line:\t@SQ\tSN:chr18\tLN:78077248\n",
      "skipping SAM header line:\t@SQ\tSN:chr20\tLN:63025520\n",
      "skipping SAM header line:\t@SQ\tSN:chr19\tLN:59128983\n",
      "skipping SAM header line:\t@SQ\tSN:chr22\tLN:51304566\n",
      "skipping SAM header line:\t@SQ\tSN:chr21\tLN:48129895\n",
      "skipping SAM header line:\t@PG\tID:Bismark\tVN:v0.24.2\tCL:\"bismark --genome genome_hg19/ -p 32 ENCFF000MBM_trimmed.fq.gz\"\n",
      "skipping SAM header line:\t@PG\tID:samtools\tPN:samtools\tPP:Bismark\tVN:1.21\tCL:/home/sagemaker-user/.conda/envs/rrbs-env/bin/samtools view -bSh -\n",
      "skipping SAM header line:\t@PG\tID:Bismark-129AB734\tVN:v0.24.2\tCL:\"bismark --genome genome_hg19/ -p 32 ENCFF000MBV_trimmed.fq.gz\"\n",
      "skipping SAM header line:\t@PG\tID:samtools-5630230E\tPN:samtools\tPP:Bismark-129AB734\tVN:1.21\tCL:/home/sagemaker-user/.conda/envs/rrbs-env/bin/samtools view -bSh -\n",
      "skipping SAM header line:\t@PG\tID:samtools.1\tPN:samtools\tPP:samtools\tVN:1.21\tCL:/home/sagemaker-user/.conda/envs/rrbs-env/bin/samtools merge -@ 34 merged.bam ENCFF000MBM_trimmed_bismark_bt2.bam ENCFF000MBV_trimmed_bismark_bt2.bam\n",
      "skipping SAM header line:\t@PG\tID:samtools.2\tPN:samtools\tPP:samtools-5630230E\tVN:1.21\tCL:/home/sagemaker-user/.conda/envs/rrbs-env/bin/samtools merge -@ 34 merged.bam ENCFF000MBM_trimmed_bismark_bt2.bam ENCFF000MBV_trimmed_bismark_bt2.bam\n",
      "skipping SAM header line:\t@PG\tID:samtools.3\tPN:samtools\tPP:samtools.1\tVN:1.21\tCL:/home/sagemaker-user/.conda/envs/rrbs-env/bin/samtools sort -@ 34 -o merged.sorted.bam merged.bam\n",
      "skipping SAM header line:\t@PG\tID:samtools.4\tPN:samtools\tPP:samtools.2\tVN:1.21\tCL:/home/sagemaker-user/.conda/envs/rrbs-env/bin/samtools sort -@ 34 -o merged.sorted.bam merged.bam\n",
      "skipping SAM header line:\t@PG\tID:samtools.5\tPN:samtools\tPP:samtools.3\tVN:1.21\tCL:/home/sagemaker-user/.conda/envs/rrbs-env/bin/samtools view -h merged.sorted.bam\n",
      "skipping SAM header line:\t@PG\tID:samtools.6\tPN:samtools\tPP:samtools.4\tVN:1.21\tCL:/home/sagemaker-user/.conda/envs/rrbs-env/bin/samtools view -h merged.sorted.bam\n",
      "Processed lines: 500000\n",
      "Processed lines: 500000\n",
      "Processed lines: 500000\n",
      "Processed lines: 500000\n",
      "Processed lines: 500000\n",
      "Processed lines: 500000\n",
      "Processed lines: 500000\n",
      "Processed lines: 500000\n",
      "Processed lines: 500000\n",
      "Processed lines: 500000\n",
      "Processed lines: 500000\n",
      "Processed lines: 500000\n",
      "Processed lines: 500000\n",
      "Processed lines: 500000\n",
      "Processed lines: 500000\n",
      "Processed lines: 500000\n",
      "Processed lines: 500000\n",
      "Processed lines: 500000\n",
      "Processed lines: 500000\n",
      "Processed lines: 500000\n",
      "Processed lines: 500000\n",
      "Processed lines: 500000\n",
      "Processed lines: 500000\n",
      "Processed lines: 500000\n",
      "Processed lines: 500000\n",
      "Processed lines: 500000\n",
      "Processed lines: 500000\n",
      "Processed lines: 500000\n",
      "Processed lines: 500000\n",
      "Processed lines: 500000\n",
      "Processed lines: 500000\n",
      "Processed lines: 500000\n",
      "Processed lines: 500000\n",
      "Processed lines: 500000\n",
      "Processed lines: 1000000\n",
      "Processed lines: 1000000\n",
      "Processed lines: 1000000\n",
      "Processed lines: 1000000\n",
      "Processed lines: 1000000\n",
      "Processed lines: 1000000\n",
      "Processed lines: 1000000\n",
      "Processed lines: 1000000\n",
      "Processed lines: 1000000\n",
      "Processed lines: 1000000\n",
      "Processed lines: 1000000\n",
      "Processed lines: 1000000\n",
      "Processed lines: 1000000\n",
      "Processed lines: 1000000\n",
      "Processed lines: 1000000\n",
      "Processed lines: 1000000\n",
      "Processed lines: 1000000\n",
      "Processed lines: 1000000\n",
      "Processed lines: 1000000\n",
      "Processed lines: 1000000\n",
      "Processed lines: 1000000\n",
      "Processed lines: 1000000\n",
      "Processed lines: 1000000\n",
      "Processed lines: 1000000\n",
      "Processed lines: 1000000\n",
      "Processed lines: 1000000\n",
      "Processed lines: 1000000\n",
      "Processed lines: 1000000\n",
      "Processed lines: 1000000\n",
      "Processed lines: 1000000\n",
      "Processed lines: 1000000\n",
      "Processed lines: 1000000\n",
      "Processed lines: 1000000\n",
      "Processed lines: 1000000\n",
      "Processed lines: 1500000\n",
      "Processed lines: 1500000\n",
      "Processed lines: 1500000\n",
      "Processed lines: 1500000\n",
      "Processed lines: 1500000\n",
      "Processed lines: 1500000\n",
      "Processed lines: 1500000\n",
      "Processed lines: 1500000\n",
      "Processed lines: 1500000\n",
      "Processed lines: 1500000\n",
      "Processed lines: 1500000\n",
      "Processed lines: 1500000\n",
      "Processed lines: 1500000\n",
      "Processed lines: 1500000\n",
      "Processed lines: 1500000\n",
      "Processed lines: 1500000\n",
      "Processed lines: 1500000\n",
      "Processed lines: 1500000\n",
      "Processed lines: 1500000\n",
      "Processed lines: 1500000\n",
      "Processed lines: 1500000\n",
      "Processed lines: 1500000\n",
      "Processed lines: 1500000\n",
      "Processed lines: 1500000\n",
      "Processed lines: 1500000\n",
      "Processed lines: 1500000\n",
      "Processed lines: 1500000\n",
      "Processed lines: 1500000\n",
      "Processed lines: 1500000\n",
      "Processed lines: 1500000\n",
      "Processed lines: 1500000\n",
      "Processed lines: 1500000\n",
      "Processed lines: 1500000\n",
      "Processed lines: 1500000\n",
      "Processed lines: 2000000\n",
      "Processed lines: 2000000\n",
      "Processed lines: 2000000\n",
      "Processed lines: 2000000\n",
      "Processed lines: 2000000\n",
      "Processed lines: 2000000\n",
      "Processed lines: 2000000\n",
      "Processed lines: 2000000\n",
      "Processed lines: 2000000\n",
      "Processed lines: 2000000\n",
      "Processed lines: 2000000\n",
      "Processed lines: 2000000\n",
      "Processed lines: 2000000\n",
      "Processed lines: 2000000\n",
      "Processed lines: 2000000\n",
      "Processed lines: 2000000\n",
      "Processed lines: 2000000\n",
      "Processed lines: 2000000\n",
      "Processed lines: 2000000\n",
      "Processed lines: 2000000\n",
      "Processed lines: 2000000\n",
      "Processed lines: 2000000\n",
      "Processed lines: 2000000\n",
      "Processed lines: 2000000\n",
      "Processed lines: 2000000\n",
      "Processed lines: 2000000\n",
      "Processed lines: 2000000\n",
      "Processed lines: 2000000\n",
      "Processed lines: 2000000\n",
      "Processed lines: 2000000\n",
      "Processed lines: 2000000\n",
      "Processed lines: 2000000\n",
      "Processed lines: 2000000\n",
      "Processed lines: 2000000\n",
      "Processed lines: 2500000\n",
      "Processed lines: 2500000\n",
      "Processed lines: 2500000\n",
      "Processed lines: 2500000\n",
      "Processed lines: 2500000\n",
      "Processed lines: 2500000\n",
      "Processed lines: 2500000\n",
      "Processed lines: 2500000\n",
      "Processed lines: 2500000\n",
      "Processed lines: 2500000\n",
      "Processed lines: 2500000\n",
      "Processed lines: 2500000\n",
      "Processed lines: 2500000\n",
      "Processed lines: 2500000\n",
      "Processed lines: 2500000\n",
      "Processed lines: 2500000\n",
      "Processed lines: 2500000\n",
      "Processed lines: 2500000\n",
      "Processed lines: 2500000\n",
      "Processed lines: 2500000\n",
      "Processed lines: 2500000\n",
      "Processed lines: 2500000\n",
      "Processed lines: 2500000\n",
      "Processed lines: 2500000\n",
      "Processed lines: 2500000\n",
      "Processed lines: 2500000\n",
      "Processed lines: 2500000\n",
      "Processed lines: 2500000\n",
      "Processed lines: 2500000\n",
      "Processed lines: 2500000\n",
      "Processed lines: 2500000\n",
      "Processed lines: 2500000\n",
      "Processed lines: 2500000\n",
      "Processed lines: 2500000\n",
      "Processed lines: 3000000\n",
      "Processed lines: 3000000\n",
      "Processed lines: 3000000\n",
      "Processed lines: 3000000\n",
      "Processed lines: 3000000\n",
      "Processed lines: 3000000\n",
      "Processed lines: 3000000\n",
      "Processed lines: 3000000\n",
      "Processed lines: 3000000\n",
      "Processed lines: 3000000\n",
      "Processed lines: 3000000\n",
      "Processed lines: 3000000\n",
      "Processed lines: 3000000\n",
      "Processed lines: 3000000\n",
      "Processed lines: 3000000\n",
      "Processed lines: 3000000\n",
      "Processed lines: 3000000\n",
      "Processed lines: 3000000\n",
      "Processed lines: 3000000\n",
      "Processed lines: 3000000\n",
      "Processed lines: 3000000\n",
      "Processed lines: 3000000\n",
      "Processed lines: 3000000\n",
      "Processed lines: 3000000\n",
      "Processed lines: 3000000\n",
      "Processed lines: 3000000\n",
      "Processed lines: 3000000\n",
      "Processed lines: 3000000\n",
      "Processed lines: 3000000\n",
      "Processed lines: 3000000\n",
      "Processed lines: 3000000\n",
      "Processed lines: 3000000\n",
      "Processed lines: 3000000\n",
      "Processed lines: 3000000\n",
      "Processed lines: 3500000\n",
      "Processed lines: 3500000\n",
      "Processed lines: 3500000\n",
      "Processed lines: 3500000\n",
      "Processed lines: 3500000\n",
      "Processed lines: 3500000\n",
      "Processed lines: 3500000\n",
      "Processed lines: 3500000\n",
      "Processed lines: 3500000\n",
      "Processed lines: 3500000\n",
      "Processed lines: 3500000\n",
      "Processed lines: 3500000\n",
      "Processed lines: 3500000\n",
      "Processed lines: 3500000\n",
      "Processed lines: 3500000\n",
      "Processed lines: 3500000\n",
      "Processed lines: 3500000\n",
      "Processed lines: 3500000\n",
      "Processed lines: 3500000\n",
      "Processed lines: 3500000\n",
      "Processed lines: 3500000\n",
      "Processed lines: 3500000\n",
      "Processed lines: 3500000\n",
      "Processed lines: 3500000\n",
      "Processed lines: 3500000\n",
      "Processed lines: 3500000\n",
      "Processed lines: 3500000\n",
      "Processed lines: 3500000\n",
      "Processed lines: 3500000\n",
      "Processed lines: 3500000\n",
      "Processed lines: 3500000\n",
      "Processed lines: 3500000\n",
      "Processed lines: 3500000\n",
      "Processed lines: 3500000\n",
      "Processed lines: 4000000\n",
      "Processed lines: 4000000\n",
      "Processed lines: 4000000\n",
      "Processed lines: 4000000\n",
      "Processed lines: 4000000\n",
      "Processed lines: 4000000\n",
      "Processed lines: 4000000\n",
      "Processed lines: 4000000\n",
      "Processed lines: 4000000\n",
      "Processed lines: 4000000\n",
      "Processed lines: 4000000\n",
      "Processed lines: 4000000\n",
      "Processed lines: 4000000\n",
      "Processed lines: 4000000\n",
      "Processed lines: 4000000\n",
      "Processed lines: 4000000\n",
      "Processed lines: 4000000\n",
      "Processed lines: 4000000\n",
      "Processed lines: 4000000\n",
      "Processed lines: 4000000\n",
      "Processed lines: 4000000\n",
      "Processed lines: 4000000\n",
      "Processed lines: 4000000\n",
      "Processed lines: 4000000\n",
      "Processed lines: 4000000\n",
      "Processed lines: 4000000\n",
      "Processed lines: 4000000\n",
      "Processed lines: 4000000\n",
      "Processed lines: 4000000\n",
      "Processed lines: 4000000\n",
      "Processed lines: 4000000\n",
      "Processed lines: 4000000\n",
      "Processed lines: 4000000\n",
      "Processed lines: 4000000\n",
      "Processed lines: 4500000\n",
      "Processed lines: 4500000\n",
      "Processed lines: 4500000\n",
      "Processed lines: 4500000\n",
      "Processed lines: 4500000\n",
      "Processed lines: 4500000\n",
      "Processed lines: 4500000\n",
      "Processed lines: 4500000\n",
      "Processed lines: 4500000\n",
      "Processed lines: 4500000\n",
      "Processed lines: 4500000\n",
      "Processed lines: 4500000\n",
      "Processed lines: 4500000\n",
      "Processed lines: 4500000\n",
      "Processed lines: 4500000\n",
      "Processed lines: 4500000\n",
      "Processed lines: 4500000\n",
      "Processed lines: 4500000\n",
      "Processed lines: 4500000\n",
      "Processed lines: 4500000\n",
      "Processed lines: 4500000\n",
      "Processed lines: 4500000\n",
      "Processed lines: 4500000\n",
      "Processed lines: 4500000\n",
      "Processed lines: 4500000\n",
      "Processed lines: 4500000\n",
      "Processed lines: 4500000\n",
      "Processed lines: 4500000\n",
      "Processed lines: 4500000\n",
      "Processed lines: 4500000\n",
      "Processed lines: 4500000\n",
      "Processed lines: 4500000\n",
      "Processed lines: 4500000\n",
      "Processed lines: 4500000\n",
      "Processed lines: 5000000\n",
      "Processed lines: 5000000\n",
      "Processed lines: 5000000\n",
      "Processed lines: 5000000\n",
      "Processed lines: 5000000\n",
      "Processed lines: 5000000\n",
      "Processed lines: 5000000\n",
      "Processed lines: 5000000\n",
      "Processed lines: 5000000\n",
      "Processed lines: 5000000\n",
      "Processed lines: 5000000\n",
      "Processed lines: 5000000\n",
      "Processed lines: 5000000\n",
      "Processed lines: 5000000\n",
      "Processed lines: 5000000\n",
      "Processed lines: 5000000\n",
      "Processed lines: 5000000\n",
      "Processed lines: 5000000\n",
      "Processed lines: 5000000\n",
      "Processed lines: 5000000\n",
      "Processed lines: 5000000\n",
      "Processed lines: 5000000\n",
      "Processed lines: 5000000\n",
      "Processed lines: 5000000\n",
      "Processed lines: 5000000\n",
      "Processed lines: 5000000\n",
      "Processed lines: 5000000\n",
      "Processed lines: 5000000\n",
      "Processed lines: 5000000\n",
      "Processed lines: 5000000\n",
      "Processed lines: 5000000\n",
      "Processed lines: 5000000\n",
      "Processed lines: 5000000\n",
      "Processed lines: 5000000\n",
      "Processed lines: 5500000\n",
      "Processed lines: 5500000\n",
      "Processed lines: 5500000\n",
      "Processed lines: 5500000\n",
      "Processed lines: 5500000\n",
      "Processed lines: 5500000\n",
      "Processed lines: 5500000\n",
      "Processed lines: 5500000\n",
      "Processed lines: 5500000\n",
      "Processed lines: 5500000\n",
      "Processed lines: 5500000\n",
      "Processed lines: 5500000\n",
      "Processed lines: 5500000\n",
      "Processed lines: 5500000\n",
      "Processed lines: 5500000\n",
      "Processed lines: 5500000\n",
      "Processed lines: 5500000\n",
      "Processed lines: 5500000\n",
      "Processed lines: 5500000\n",
      "Processed lines: 5500000\n",
      "Processed lines: 5500000\n",
      "Processed lines: 5500000\n",
      "Processed lines: 5500000\n",
      "Processed lines: 5500000\n",
      "Processed lines: 5500000\n",
      "Processed lines: 5500000\n",
      "Processed lines: 5500000\n",
      "Processed lines: 5500000\n",
      "Processed lines: 5500000\n",
      "Processed lines: 5500000\n",
      "Processed lines: 5500000\n",
      "Processed lines: 5500000\n",
      "Processed lines: 5500000\n",
      "Processed lines: 5500000\n",
      "Processed lines: 6000000\n",
      "Processed lines: 6000000\n",
      "Processed lines: 6000000\n",
      "Processed lines: 6000000\n",
      "Processed lines: 6000000\n",
      "Processed lines: 6000000\n",
      "Processed lines: 6000000\n",
      "Processed lines: 6000000\n",
      "Processed lines: 6000000\n",
      "Processed lines: 6000000\n",
      "Processed lines: 6000000\n",
      "Processed lines: 6000000\n",
      "Processed lines: 6000000\n",
      "Processed lines: 6000000\n",
      "Processed lines: 6000000\n",
      "Processed lines: 6000000\n",
      "Processed lines: 6000000\n",
      "Processed lines: 6000000\n",
      "Processed lines: 6000000\n",
      "Processed lines: 6000000\n",
      "Processed lines: 6000000\n",
      "Processed lines: 6000000\n",
      "Processed lines: 6000000\n",
      "Processed lines: 6000000\n",
      "Processed lines: 6000000\n",
      "Processed lines: 6000000\n",
      "Processed lines: 6000000\n",
      "Processed lines: 6000000\n",
      "Processed lines: 6000000\n",
      "Processed lines: 6000000\n",
      "Processed lines: 6000000\n",
      "Processed lines: 6000000\n",
      "Processed lines: 6000000\n",
      "Processed lines: 6000000\n",
      "Processed lines: 6500000\n",
      "Processed lines: 6500000\n",
      "Processed lines: 6500000\n",
      "Processed lines: 6500000\n",
      "Processed lines: 6500000\n",
      "Processed lines: 6500000\n",
      "Processed lines: 6500000\n",
      "Processed lines: 6500000\n",
      "Processed lines: 6500000\n",
      "Processed lines: 6500000\n",
      "Processed lines: 6500000\n",
      "Processed lines: 6500000\n",
      "Processed lines: 6500000\n",
      "Processed lines: 6500000\n",
      "Processed lines: 6500000\n",
      "Processed lines: 6500000\n",
      "Processed lines: 6500000\n",
      "Processed lines: 6500000\n",
      "Processed lines: 6500000\n",
      "Processed lines: 6500000\n",
      "Processed lines: 6500000\n",
      "Processed lines: 6500000\n",
      "Processed lines: 6500000\n",
      "Processed lines: 6500000\n",
      "Processed lines: 6500000\n",
      "Processed lines: 6500000\n",
      "Processed lines: 6500000\n",
      "Processed lines: 6500000\n",
      "Processed lines: 6500000\n",
      "Processed lines: 6500000\n",
      "Processed lines: 6500000\n",
      "Processed lines: 6500000\n",
      "Processed lines: 6500000\n",
      "Processed lines: 6500000\n",
      "Processed lines: 7000000\n",
      "Processed lines: 7000000\n",
      "Processed lines: 7000000\n",
      "Processed lines: 7000000\n",
      "Processed lines: 7000000\n",
      "Processed lines: 7000000\n",
      "Processed lines: 7000000\n",
      "Processed lines: 7000000\n",
      "Processed lines: 7000000\n",
      "Processed lines: 7000000\n",
      "Processed lines: 7000000\n",
      "Processed lines: 7000000\n",
      "Processed lines: 7000000\n",
      "Processed lines: 7000000\n",
      "Processed lines: 7000000\n",
      "Processed lines: 7000000\n",
      "Processed lines: 7000000\n",
      "Processed lines: 7000000\n",
      "Processed lines: 7000000\n",
      "Processed lines: 7000000\n",
      "Processed lines: 7000000\n",
      "Processed lines: 7000000\n",
      "Processed lines: 7000000\n",
      "Processed lines: 7000000\n",
      "Processed lines: 7000000\n",
      "Processed lines: 7000000\n",
      "Processed lines: 7000000\n",
      "Processed lines: 7000000\n",
      "Processed lines: 7000000\n",
      "Processed lines: 7000000\n",
      "Processed lines: 7000000\n",
      "Processed lines: 7000000\n",
      "Processed lines: 7000000\n",
      "Processed lines: 7000000\n",
      "Processed lines: 7500000\n",
      "Processed lines: 7500000\n",
      "Processed lines: 7500000\n",
      "Processed lines: 7500000\n",
      "Processed lines: 7500000\n",
      "Processed lines: 7500000\n",
      "Processed lines: 7500000\n",
      "Processed lines: 7500000\n",
      "Processed lines: 7500000\n",
      "Processed lines: 7500000\n",
      "Processed lines: 7500000\n",
      "Processed lines: 7500000\n",
      "Processed lines: 7500000\n",
      "Processed lines: 7500000\n",
      "Processed lines: 7500000\n",
      "Processed lines: 7500000\n",
      "Processed lines: 7500000\n",
      "Processed lines: 7500000\n",
      "Processed lines: 7500000\n",
      "Processed lines: 7500000\n",
      "Processed lines: 7500000\n",
      "Processed lines: 7500000\n",
      "Processed lines: 7500000\n",
      "Processed lines: 7500000\n",
      "Processed lines: 7500000\n",
      "Processed lines: 7500000\n",
      "Processed lines: 7500000\n",
      "Processed lines: 7500000\n",
      "Processed lines: 7500000\n",
      "Processed lines: 7500000\n",
      "Processed lines: 7500000\n",
      "Processed lines: 7500000\n",
      "Processed lines: 7500000\n",
      "Processed lines: 7500000\n",
      "Processed lines: 8000000\n",
      "Processed lines: 8000000\n",
      "Processed lines: 8000000\n",
      "Processed lines: 8000000\n",
      "Processed lines: 8000000\n",
      "Processed lines: 8000000\n",
      "Processed lines: 8000000\n",
      "Processed lines: 8000000\n",
      "Processed lines: 8000000\n",
      "Processed lines: 8000000\n",
      "Processed lines: 8000000\n",
      "Processed lines: 8000000\n",
      "Processed lines: 8000000\n",
      "Processed lines: 8000000\n",
      "Processed lines: 8000000\n",
      "Processed lines: 8000000\n",
      "Processed lines: 8000000\n",
      "Processed lines: 8000000\n",
      "Processed lines: 8000000\n",
      "Processed lines: 8000000\n",
      "Processed lines: 8000000\n",
      "Processed lines: 8000000\n",
      "Processed lines: 8000000\n",
      "Processed lines: 8000000\n",
      "Processed lines: 8000000\n",
      "Processed lines: 8000000\n",
      "Processed lines: 8000000\n",
      "Processed lines: 8000000\n",
      "Processed lines: 8000000\n",
      "Processed lines: 8000000\n",
      "Processed lines: 8000000\n",
      "Processed lines: 8000000\n",
      "Processed lines: 8000000\n",
      "Processed lines: 8000000\n",
      "Processed lines: 8500000\n",
      "Processed lines: 8500000\n",
      "Processed lines: 8500000\n",
      "Processed lines: 8500000\n",
      "Processed lines: 8500000\n",
      "Processed lines: 8500000\n",
      "Processed lines: 8500000\n",
      "Processed lines: 8500000\n",
      "Processed lines: 8500000\n",
      "Processed lines: 8500000\n",
      "Processed lines: 8500000\n",
      "Processed lines: 8500000\n",
      "Processed lines: 8500000\n",
      "Processed lines: 8500000\n",
      "Processed lines: 8500000\n",
      "Processed lines: 8500000\n",
      "Processed lines: 8500000\n",
      "Processed lines: 8500000\n",
      "Processed lines: 8500000\n",
      "Processed lines: 8500000\n",
      "Processed lines: 8500000\n",
      "Processed lines: 8500000\n",
      "Processed lines: 8500000\n",
      "Processed lines: 8500000\n",
      "Processed lines: 8500000\n",
      "Processed lines: 8500000\n",
      "Processed lines: 8500000\n",
      "Processed lines: 8500000\n",
      "Processed lines: 8500000\n",
      "Processed lines: 8500000\n",
      "Processed lines: 8500000\n",
      "Processed lines: 8500000\n",
      "Processed lines: 8500000\n",
      "Processed lines: 8500000\n",
      "Processed lines: 9000000\n",
      "Processed lines: 9000000\n",
      "Processed lines: 9000000\n",
      "Processed lines: 9000000\n",
      "Processed lines: 9000000\n",
      "Processed lines: 9000000\n",
      "Processed lines: 9000000\n",
      "Processed lines: 9000000\n",
      "Processed lines: 9000000\n",
      "Processed lines: 9000000\n",
      "Processed lines: 9000000\n",
      "Processed lines: 9000000\n",
      "Processed lines: 9000000\n",
      "Processed lines: 9000000\n",
      "Processed lines: 9000000\n",
      "Processed lines: 9000000\n",
      "Processed lines: 9000000\n",
      "Processed lines: 9000000\n",
      "Processed lines: 9000000\n",
      "Processed lines: 9000000\n",
      "Processed lines: 9000000\n",
      "Processed lines: 9000000\n",
      "Processed lines: 9000000\n",
      "Processed lines: 9000000\n",
      "Processed lines: 9000000\n",
      "Processed lines: 9000000\n",
      "Processed lines: 9000000\n",
      "Processed lines: 9000000\n",
      "Processed lines: 9000000\n",
      "Processed lines: 9000000\n",
      "Processed lines: 9000000\n",
      "Processed lines: 9000000\n",
      "Processed lines: 9000000\n",
      "Processed lines: 9000000\n",
      "Processed lines: 9500000\n",
      "Processed lines: 9500000\n",
      "Processed lines: 9500000\n",
      "Processed lines: 9500000\n",
      "Processed lines: 9500000\n",
      "Processed lines: 9500000\n",
      "Processed lines: 9500000\n",
      "Processed lines: 9500000\n",
      "Processed lines: 9500000\n",
      "Processed lines: 9500000\n",
      "Processed lines: 9500000\n",
      "Processed lines: 9500000\n",
      "Processed lines: 9500000\n",
      "Processed lines: 9500000\n",
      "Processed lines: 9500000\n",
      "Processed lines: 9500000\n",
      "Processed lines: 9500000\n",
      "Processed lines: 9500000\n",
      "Processed lines: 9500000\n",
      "Processed lines: 9500000\n",
      "Processed lines: 9500000\n",
      "Processed lines: 9500000\n",
      "Processed lines: 9500000\n",
      "Processed lines: 9500000\n",
      "Processed lines: 9500000\n",
      "Processed lines: 9500000\n",
      "Processed lines: 9500000\n",
      "Processed lines: 9500000\n",
      "Processed lines: 9500000\n",
      "Processed lines: 9500000\n",
      "Processed lines: 9500000\n",
      "Processed lines: 9500000\n",
      "Processed lines: 9500000\n",
      "Processed lines: 9500000\n",
      "Processed lines: 10000000\n",
      "Processed lines: 10000000\n",
      "Processed lines: 10000000\n",
      "Processed lines: 10000000\n",
      "Processed lines: 10000000\n",
      "Processed lines: 10000000\n",
      "Processed lines: 10000000\n",
      "Processed lines: 10000000\n",
      "Processed lines: 10000000\n",
      "Processed lines: 10000000\n",
      "Processed lines: 10000000\n",
      "Processed lines: 10000000\n",
      "Processed lines: 10000000\n",
      "Processed lines: 10000000\n",
      "Processed lines: 10000000\n",
      "Processed lines: 10000000\n",
      "Processed lines: 10000000\n",
      "Processed lines: 10000000\n",
      "Processed lines: 10000000\n",
      "Processed lines: 10000000\n",
      "Processed lines: 10000000\n",
      "Processed lines: 10000000\n",
      "Processed lines: 10000000\n",
      "Processed lines: 10000000\n",
      "Processed lines: 10000000\n",
      "Processed lines: 10000000\n",
      "Processed lines: 10000000\n",
      "Processed lines: 10000000\n",
      "Processed lines: 10000000\n",
      "Processed lines: 10000000\n",
      "Processed lines: 10000000\n",
      "Processed lines: 10000000\n",
      "Processed lines: 10000000\n",
      "Processed lines: 10000000\n",
      "Processed lines: 10500000\n",
      "Processed lines: 10500000\n",
      "Processed lines: 10500000\n",
      "Processed lines: 10500000\n",
      "Processed lines: 10500000\n",
      "Processed lines: 10500000\n",
      "Processed lines: 10500000\n",
      "Processed lines: 10500000\n",
      "Processed lines: 10500000\n",
      "Processed lines: 10500000\n",
      "Processed lines: 10500000\n",
      "Processed lines: 10500000\n",
      "Processed lines: 10500000\n",
      "Processed lines: 10500000\n",
      "Processed lines: 10500000\n",
      "Processed lines: 10500000\n",
      "Processed lines: 10500000\n",
      "Processed lines: 10500000\n",
      "Processed lines: 10500000\n",
      "Processed lines: 10500000\n",
      "Processed lines: 10500000\n",
      "Processed lines: 10500000\n",
      "Processed lines: 10500000\n",
      "Processed lines: 10500000\n",
      "Processed lines: 10500000\n",
      "Processed lines: 10500000\n",
      "Processed lines: 10500000\n",
      "Processed lines: 10500000\n",
      "Processed lines: 10500000\n",
      "Processed lines: 10500000\n",
      "Processed lines: 10500000\n",
      "Processed lines: 10500000\n",
      "Processed lines: 10500000\n",
      "Processed lines: 10500000\n",
      "Processed lines: 11000000\n",
      "Processed lines: 11000000\n",
      "Processed lines: 11000000\n",
      "Processed lines: 11000000\n",
      "Processed lines: 11000000\n",
      "Processed lines: 11000000\n",
      "Processed lines: 11000000\n",
      "Processed lines: 11000000\n",
      "Processed lines: 11000000\n",
      "Processed lines: 11000000\n",
      "Processed lines: 11000000\n",
      "Processed lines: 11000000\n",
      "Processed lines: 11000000\n",
      "Processed lines: 11000000\n",
      "Processed lines: 11000000\n",
      "Processed lines: 11000000\n",
      "Processed lines: 11000000\n",
      "Processed lines: 11000000\n",
      "Processed lines: 11000000\n",
      "Processed lines: 11000000\n",
      "Processed lines: 11000000\n",
      "Processed lines: 11000000\n",
      "Processed lines: 11000000\n",
      "Processed lines: 11000000\n",
      "Processed lines: 11000000\n",
      "Processed lines: 11000000\n",
      "Processed lines: 11000000\n",
      "Processed lines: 11000000\n",
      "Processed lines: 11000000\n",
      "Processed lines: 11000000\n",
      "Processed lines: 11000000\n",
      "Processed lines: 11000000\n",
      "Processed lines: 11000000\n",
      "Processed lines: 11000000\n",
      "Processed lines: 11500000\n",
      "Processed lines: 11500000\n",
      "Processed lines: 11500000\n",
      "Processed lines: 11500000\n",
      "Processed lines: 11500000\n",
      "Processed lines: 11500000\n",
      "Processed lines: 11500000\n",
      "Processed lines: 11500000\n",
      "Processed lines: 11500000\n",
      "Processed lines: 11500000\n",
      "Processed lines: 11500000\n",
      "Processed lines: 11500000\n",
      "Processed lines: 11500000\n",
      "Processed lines: 11500000\n",
      "Processed lines: 11500000\n",
      "Processed lines: 11500000\n",
      "Processed lines: 11500000\n",
      "Processed lines: 11500000\n",
      "Processed lines: 11500000\n",
      "Processed lines: 11500000\n",
      "Processed lines: 11500000\n",
      "Processed lines: 11500000\n",
      "Processed lines: 11500000\n",
      "Processed lines: 11500000\n",
      "Processed lines: 11500000\n",
      "Processed lines: 11500000\n",
      "Processed lines: 11500000\n",
      "Processed lines: 11500000\n",
      "Processed lines: 11500000\n",
      "Processed lines: 11500000\n",
      "Processed lines: 11500000\n",
      "Processed lines: 11500000\n",
      "Processed lines: 11500000\n",
      "Processed lines: 11500000\n",
      "Processed lines: 12000000\n",
      "Processed lines: 12000000\n",
      "Processed lines: 12000000\n",
      "Processed lines: 12000000\n",
      "Processed lines: 12000000\n",
      "Processed lines: 12000000\n",
      "Processed lines: 12000000\n",
      "Processed lines: 12000000\n",
      "Processed lines: 12000000\n",
      "Processed lines: 12000000\n",
      "Processed lines: 12000000\n",
      "Processed lines: 12000000\n",
      "Processed lines: 12000000\n",
      "Processed lines: 12000000\n",
      "Processed lines: 12000000\n",
      "Processed lines: 12000000\n",
      "Processed lines: 12000000\n",
      "Processed lines: 12000000\n",
      "Processed lines: 12000000\n",
      "Processed lines: 12000000\n",
      "Processed lines: 12000000\n",
      "Processed lines: 12000000\n",
      "Processed lines: 12000000\n",
      "Processed lines: 12000000\n",
      "Processed lines: 12000000\n",
      "Processed lines: 12000000\n",
      "Processed lines: 12000000\n",
      "Processed lines: 12000000\n",
      "Processed lines: 12000000\n",
      "Processed lines: 12000000\n",
      "Processed lines: 12000000\n",
      "Processed lines: 12000000\n",
      "Processed lines: 12000000\n",
      "Processed lines: 12000000\n",
      "Processed lines: 12500000\n",
      "Processed lines: 12500000\n",
      "Processed lines: 12500000\n",
      "Processed lines: 12500000\n",
      "Processed lines: 12500000\n",
      "Processed lines: 12500000\n",
      "Processed lines: 12500000\n",
      "Processed lines: 12500000\n",
      "Processed lines: 12500000\n",
      "Processed lines: 12500000\n",
      "Processed lines: 12500000\n",
      "Processed lines: 12500000\n",
      "Processed lines: 12500000\n",
      "Processed lines: 12500000\n",
      "Processed lines: 12500000\n",
      "Processed lines: 12500000\n",
      "Processed lines: 12500000\n",
      "Processed lines: 12500000\n",
      "Processed lines: 12500000\n",
      "Processed lines: 12500000\n",
      "Processed lines: 12500000\n",
      "Processed lines: 12500000\n",
      "Processed lines: 12500000\n",
      "Processed lines: 12500000\n",
      "Processed lines: 12500000\n",
      "Processed lines: 12500000\n",
      "Processed lines: 12500000\n",
      "Processed lines: 12500000\n",
      "Processed lines: 12500000\n",
      "Processed lines: 12500000\n",
      "Processed lines: 12500000\n",
      "Processed lines: 12500000\n",
      "Processed lines: 12500000\n",
      "Processed lines: 12500000\n",
      "Processed lines: 13000000\n",
      "Processed lines: 13000000\n",
      "Processed lines: 13000000\n",
      "Processed lines: 13000000\n",
      "Processed lines: 13000000\n",
      "Processed lines: 13000000\n",
      "Processed lines: 13000000\n",
      "Processed lines: 13000000\n",
      "Processed lines: 13000000\n",
      "Processed lines: 13000000\n",
      "Processed lines: 13000000\n",
      "Processed lines: 13000000\n",
      "Processed lines: 13000000\n",
      "Processed lines: 13000000\n",
      "Processed lines: 13000000\n",
      "Processed lines: 13000000\n",
      "Processed lines: 13000000\n",
      "Processed lines: 13000000\n",
      "Processed lines: 13000000\n",
      "Processed lines: 13000000\n",
      "Processed lines: 13000000\n",
      "Processed lines: 13000000\n",
      "Processed lines: 13000000\n",
      "Processed lines: 13000000\n",
      "Processed lines: 13000000\n",
      "Processed lines: 13000000\n",
      "Processed lines: 13000000\n",
      "Processed lines: 13000000\n",
      "Processed lines: 13000000\n",
      "Processed lines: 13000000\n",
      "Processed lines: 13000000\n",
      "Processed lines: 13000000\n",
      "Processed lines: 13000000\n",
      "Processed lines: 13000000\n",
      "Processed lines: 13500000\n",
      "Processed lines: 13500000\n",
      "Processed lines: 13500000\n",
      "Processed lines: 13500000\n",
      "Processed lines: 13500000\n",
      "Processed lines: 13500000\n",
      "Processed lines: 13500000\n",
      "Processed lines: 13500000\n",
      "Processed lines: 13500000\n",
      "Processed lines: 13500000\n",
      "Processed lines: 13500000\n",
      "Processed lines: 13500000\n",
      "Processed lines: 13500000\n",
      "Processed lines: 13500000\n",
      "Processed lines: 13500000\n",
      "Processed lines: 13500000\n",
      "Processed lines: 13500000\n",
      "Processed lines: 13500000\n",
      "Processed lines: 13500000\n",
      "Processed lines: 13500000\n",
      "Processed lines: 13500000\n",
      "Processed lines: 13500000\n",
      "Processed lines: 13500000\n",
      "Processed lines: 13500000\n",
      "Processed lines: 13500000\n",
      "Processed lines: 13500000\n",
      "Processed lines: 13500000\n",
      "Processed lines: 13500000\n",
      "Processed lines: 13500000\n",
      "Processed lines: 13500000\n",
      "Processed lines: 13500000\n",
      "Processed lines: 13500000\n",
      "Processed lines: 13500000\n",
      "Processed lines: 13500000\n",
      "Processed lines: 14000000\n",
      "Processed lines: 14000000\n",
      "Processed lines: 14000000\n",
      "Processed lines: 14000000\n",
      "Processed lines: 14000000\n",
      "Processed lines: 14000000\n",
      "Processed lines: 14000000\n",
      "Processed lines: 14000000\n",
      "Processed lines: 14000000\n",
      "Processed lines: 14000000\n",
      "Processed lines: 14000000\n",
      "Processed lines: 14000000\n",
      "Processed lines: 14000000\n",
      "Processed lines: 14000000\n",
      "Processed lines: 14000000\n",
      "Processed lines: 14000000\n",
      "Processed lines: 14000000\n",
      "Processed lines: 14000000\n",
      "Processed lines: 14000000\n",
      "Processed lines: 14000000\n",
      "Processed lines: 14000000\n",
      "Processed lines: 14000000\n",
      "Processed lines: 14000000\n",
      "Processed lines: 14000000\n",
      "Processed lines: 14000000\n",
      "Processed lines: 14000000\n",
      "Processed lines: 14000000\n",
      "Processed lines: 14000000\n",
      "Processed lines: 14000000\n",
      "Processed lines: 14000000\n",
      "Processed lines: 14000000\n",
      "Processed lines: 14000000\n",
      "Processed lines: 14000000\n",
      "Processed lines: 14000000\n",
      "Processed lines: 14500000\n",
      "Processed lines: 14500000\n",
      "Processed lines: 14500000\n",
      "Processed lines: 14500000\n",
      "Processed lines: 14500000\n",
      "Processed lines: 14500000\n",
      "Processed lines: 14500000\n",
      "Processed lines: 14500000\n",
      "Processed lines: 14500000\n",
      "Processed lines: 14500000\n",
      "Processed lines: 14500000\n",
      "Processed lines: 14500000\n",
      "Processed lines: 14500000\n",
      "Processed lines: 14500000\n",
      "Processed lines: 14500000\n",
      "Processed lines: 14500000\n",
      "Processed lines: 14500000\n",
      "Processed lines: 14500000\n",
      "Processed lines: 14500000\n",
      "Processed lines: 14500000\n",
      "Processed lines: 14500000\n",
      "Processed lines: 14500000\n",
      "Processed lines: 14500000\n",
      "Processed lines: 14500000\n",
      "Processed lines: 14500000\n",
      "Processed lines: 14500000\n",
      "Processed lines: 14500000\n",
      "Processed lines: 14500000\n",
      "Processed lines: 14500000\n",
      "Processed lines: 14500000\n",
      "Processed lines: 14500000\n",
      "Processed lines: 14500000\n",
      "Processed lines: 14500000\n",
      "Processed lines: 14500000\n",
      "Processed lines: 15000000\n",
      "Processed lines: 15000000\n",
      "Processed lines: 15000000\n",
      "Processed lines: 15000000\n",
      "Processed lines: 15000000\n",
      "Processed lines: 15000000\n",
      "Processed lines: 15000000\n",
      "Processed lines: 15000000\n",
      "Processed lines: 15000000\n",
      "Processed lines: 15000000\n",
      "Processed lines: 15000000\n",
      "Processed lines: 15000000\n",
      "Processed lines: 15000000\n",
      "Processed lines: 15000000\n",
      "Processed lines: 15000000\n",
      "Processed lines: 15000000\n",
      "Processed lines: 15000000\n",
      "Processed lines: 15000000\n",
      "Processed lines: 15000000\n",
      "Processed lines: 15000000\n",
      "Processed lines: 15000000\n",
      "Processed lines: 15000000\n",
      "Processed lines: 15000000\n",
      "Processed lines: 15000000\n",
      "Processed lines: 15000000\n",
      "Processed lines: 15000000\n",
      "Processed lines: 15000000\n",
      "Processed lines: 15000000\n",
      "Processed lines: 15000000\n",
      "Processed lines: 15000000\n",
      "Processed lines: 15000000\n",
      "Processed lines: 15000000\n",
      "Processed lines: 15000000\n",
      "Processed lines: 15000000\n",
      "Processed lines: 15500000\n",
      "Processed lines: 15500000\n",
      "Processed lines: 15500000\n",
      "Processed lines: 15500000\n",
      "Processed lines: 15500000\n",
      "Processed lines: 15500000\n",
      "Processed lines: 15500000\n",
      "Processed lines: 15500000\n",
      "Processed lines: 15500000\n",
      "Processed lines: 15500000\n",
      "Processed lines: 15500000\n",
      "Processed lines: 15500000\n",
      "Processed lines: 15500000\n",
      "Processed lines: 15500000\n",
      "Processed lines: 15500000\n",
      "Processed lines: 15500000\n",
      "Processed lines: 15500000\n",
      "Processed lines: 15500000\n",
      "Processed lines: 15500000\n",
      "Processed lines: 15500000\n",
      "Processed lines: 15500000\n",
      "Processed lines: 15500000\n",
      "Processed lines: 15500000\n",
      "Processed lines: 15500000\n",
      "Processed lines: 15500000\n",
      "Processed lines: 15500000\n",
      "Processed lines: 15500000\n",
      "Processed lines: 15500000\n",
      "Processed lines: 15500000\n",
      "Processed lines: 15500000\n",
      "Processed lines: 15500000\n",
      "Processed lines: 15500000\n",
      "Processed lines: 15500000\n",
      "Processed lines: 15500000\n",
      "Processed lines: 16000000\n",
      "Processed lines: 16000000\n",
      "Processed lines: 16000000\n",
      "Processed lines: 16000000\n",
      "Processed lines: 16000000\n",
      "Processed lines: 16000000\n",
      "Processed lines: 16000000\n",
      "Processed lines: 16000000\n",
      "Processed lines: 16000000\n",
      "Processed lines: 16000000\n",
      "Processed lines: 16000000\n",
      "Processed lines: 16000000\n",
      "Processed lines: 16000000\n",
      "Processed lines: 16000000\n",
      "Processed lines: 16000000\n",
      "Processed lines: 16000000\n",
      "Processed lines: 16000000\n",
      "Processed lines: 16000000\n",
      "Processed lines: 16000000\n",
      "Processed lines: 16000000\n",
      "Processed lines: 16000000\n",
      "Processed lines: 16000000\n",
      "Processed lines: 16000000\n",
      "Processed lines: 16000000\n",
      "Processed lines: 16000000\n",
      "Processed lines: 16000000\n",
      "Processed lines: 16000000\n",
      "Processed lines: 16000000\n",
      "Processed lines: 16000000\n",
      "Processed lines: 16000000\n",
      "Processed lines: 16000000\n",
      "Processed lines: 16000000\n",
      "Processed lines: 16000000\n",
      "Processed lines: 16000000\n",
      "Processed lines: 16500000\n",
      "Processed lines: 16500000\n",
      "Processed lines: 16500000\n",
      "Processed lines: 16500000\n",
      "Processed lines: 16500000\n",
      "Processed lines: 16500000\n",
      "Processed lines: 16500000\n",
      "Processed lines: 16500000\n",
      "Processed lines: 16500000\n",
      "Processed lines: 16500000\n",
      "Processed lines: 16500000\n",
      "Processed lines: 16500000\n",
      "Processed lines: 16500000\n",
      "Processed lines: 16500000\n",
      "Processed lines: 16500000\n",
      "Processed lines: 16500000\n",
      "Processed lines: 16500000\n",
      "Processed lines: 16500000\n",
      "Processed lines: 16500000\n",
      "Processed lines: 16500000\n",
      "Processed lines: 16500000\n",
      "Processed lines: 16500000\n",
      "Processed lines: 16500000\n",
      "Processed lines: 16500000\n",
      "Processed lines: 16500000\n",
      "Processed lines: 16500000\n",
      "Processed lines: 16500000\n",
      "Processed lines: 16500000\n",
      "Processed lines: 16500000\n",
      "Processed lines: 16500000\n",
      "Processed lines: 16500000\n",
      "Processed lines: 16500000\n",
      "Processed lines: 16500000\n",
      "Processed lines: 16500000\n",
      "Processed lines: 17000000\n",
      "Processed lines: 17000000\n",
      "Processed lines: 17000000\n",
      "Processed lines: 17000000\n",
      "Processed lines: 17000000\n",
      "Processed lines: 17000000\n",
      "Processed lines: 17000000\n",
      "Processed lines: 17000000\n",
      "Processed lines: 17000000\n",
      "Processed lines: 17000000\n",
      "Processed lines: 17000000\n",
      "Processed lines: 17000000\n",
      "Processed lines: 17000000\n",
      "Processed lines: 17000000\n",
      "Processed lines: 17000000\n",
      "Processed lines: 17000000\n",
      "Processed lines: 17000000\n",
      "Processed lines: 17000000\n",
      "Processed lines: 17000000\n",
      "Processed lines: 17000000\n",
      "Processed lines: 17000000\n",
      "Processed lines: 17000000\n",
      "Processed lines: 17000000\n",
      "Processed lines: 17000000\n",
      "Processed lines: 17000000\n",
      "Processed lines: 17000000\n",
      "Processed lines: 17000000\n",
      "Processed lines: 17000000\n",
      "Processed lines: 17000000\n",
      "Processed lines: 17000000\n",
      "Processed lines: 17000000\n",
      "Processed lines: 17000000\n",
      "Processed lines: 17000000\n",
      "Processed lines: 17000000\n",
      "Processed lines: 17500000\n",
      "Processed lines: 17500000\n",
      "Processed lines: 17500000\n",
      "Processed lines: 17500000\n",
      "Processed lines: 17500000\n",
      "Processed lines: 17500000\n",
      "Processed lines: 17500000\n",
      "Processed lines: 17500000\n",
      "Processed lines: 17500000\n",
      "Processed lines: 17500000\n",
      "Processed lines: 17500000\n",
      "Processed lines: 17500000\n",
      "Processed lines: 17500000\n",
      "Processed lines: 17500000\n",
      "Processed lines: 17500000\n",
      "Processed lines: 17500000\n",
      "Processed lines: 17500000\n",
      "Processed lines: 17500000\n",
      "Processed lines: 17500000\n",
      "Processed lines: 17500000\n",
      "Processed lines: 17500000\n",
      "Processed lines: 17500000\n",
      "Processed lines: 17500000\n",
      "Processed lines: 17500000\n",
      "Processed lines: 17500000\n",
      "Processed lines: 17500000\n",
      "Processed lines: 17500000\n",
      "Processed lines: 17500000\n",
      "Processed lines: 17500000\n",
      "Processed lines: 17500000\n",
      "Processed lines: 17500000\n",
      "Processed lines: 17500000\n",
      "Processed lines: 17500000\n",
      "Processed lines: 17500000\n",
      "Processed lines: 18000000\n",
      "Processed lines: 18000000\n",
      "Processed lines: 18000000\n",
      "Processed lines: 18000000\n",
      "Processed lines: 18000000\n",
      "Processed lines: 18000000\n",
      "Processed lines: 18000000\n",
      "Processed lines: 18000000\n",
      "Processed lines: 18000000\n",
      "Processed lines: 18000000\n",
      "Processed lines: 18000000\n",
      "Processed lines: 18000000\n",
      "Processed lines: 18000000\n",
      "Processed lines: 18000000\n",
      "Processed lines: 18000000\n",
      "Processed lines: 18000000\n",
      "Processed lines: 18000000\n",
      "Processed lines: 18000000\n",
      "Processed lines: 18000000\n",
      "Processed lines: 18000000\n",
      "Processed lines: 18000000\n",
      "Processed lines: 18000000\n",
      "Processed lines: 18000000\n",
      "Processed lines: 18000000\n",
      "Processed lines: 18000000\n",
      "Processed lines: 18000000\n",
      "Processed lines: 18000000\n",
      "Processed lines: 18000000\n",
      "Processed lines: 18000000\n",
      "Processed lines: 18000000\n",
      "Processed lines: 18000000\n",
      "Processed lines: 18000000\n",
      "Processed lines: 18000000\n",
      "Processed lines: 18000000\n",
      "Processed lines: 18500000\n",
      "Processed lines: 18500000\n",
      "Processed lines: 18500000\n",
      "Processed lines: 18500000\n",
      "Processed lines: 18500000\n",
      "Processed lines: 18500000\n",
      "Processed lines: 18500000\n",
      "Processed lines: 18500000\n",
      "Processed lines: 18500000\n",
      "Processed lines: 18500000\n",
      "Processed lines: 18500000\n",
      "Processed lines: 18500000\n",
      "Processed lines: 18500000\n",
      "Processed lines: 18500000\n",
      "Processed lines: 18500000\n",
      "Processed lines: 18500000\n",
      "Processed lines: 18500000\n",
      "Processed lines: 18500000\n",
      "Processed lines: 18500000\n",
      "Processed lines: 18500000\n",
      "Processed lines: 18500000\n",
      "Processed lines: 18500000\n",
      "Processed lines: 18500000\n",
      "Processed lines: 18500000\n",
      "Processed lines: 18500000\n",
      "Processed lines: 18500000\n",
      "Processed lines: 18500000\n",
      "Processed lines: 18500000\n",
      "Processed lines: 18500000\n",
      "Processed lines: 18500000\n",
      "Processed lines: 18500000\n",
      "Processed lines: 18500000\n",
      "Processed lines: 18500000\n",
      "Processed lines: 18500000\n",
      "Processed lines: 19000000\n",
      "Processed lines: 19000000\n",
      "Processed lines: 19000000\n",
      "Processed lines: 19000000\n",
      "Processed lines: 19000000\n",
      "Processed lines: 19000000\n",
      "Processed lines: 19000000\n",
      "Processed lines: 19000000\n",
      "Processed lines: 19000000\n",
      "Processed lines: 19000000\n",
      "Processed lines: 19000000\n",
      "Processed lines: 19000000\n",
      "Processed lines: 19000000\n",
      "Processed lines: 19000000\n",
      "Processed lines: 19000000\n",
      "Processed lines: 19000000\n",
      "Processed lines: 19000000\n",
      "Processed lines: 19000000\n",
      "Processed lines: 19000000\n",
      "Processed lines: 19000000\n",
      "Processed lines: 19000000\n",
      "Processed lines: 19000000\n",
      "Processed lines: 19000000\n",
      "Processed lines: 19000000\n",
      "Processed lines: 19000000\n",
      "Processed lines: 19000000\n",
      "Processed lines: 19000000\n",
      "Processed lines: 19000000\n",
      "Processed lines: 19000000\n",
      "Processed lines: 19000000\n",
      "Processed lines: 19000000\n",
      "Processed lines: 19000000\n",
      "Processed lines: 19000000\n",
      "Processed lines: 19000000\n",
      "Processed lines: 19500000\n",
      "Processed lines: 19500000\n",
      "Processed lines: 19500000\n",
      "Processed lines: 19500000\n",
      "Processed lines: 19500000\n",
      "Processed lines: 19500000\n",
      "Processed lines: 19500000\n",
      "Processed lines: 19500000\n",
      "Processed lines: 19500000\n",
      "Processed lines: 19500000\n",
      "Processed lines: 19500000\n",
      "Processed lines: 19500000\n",
      "Processed lines: 19500000\n",
      "Processed lines: 19500000\n",
      "Processed lines: 19500000\n",
      "Processed lines: 19500000\n",
      "Processed lines: 19500000\n",
      "Processed lines: 19500000\n",
      "Processed lines: 19500000\n",
      "Processed lines: 19500000\n",
      "Processed lines: 19500000\n",
      "Processed lines: 19500000\n",
      "Processed lines: 19500000\n",
      "Processed lines: 19500000\n",
      "Processed lines: 19500000\n",
      "Processed lines: 19500000\n",
      "Processed lines: 19500000\n",
      "Processed lines: 19500000\n",
      "Processed lines: 19500000\n",
      "Processed lines: 19500000\n",
      "Processed lines: 19500000\n",
      "Processed lines: 19500000\n",
      "Processed lines: 19500000\n",
      "Processed lines: 19500000\n",
      "Processed lines: 20000000\n",
      "Processed lines: 20000000\n",
      "Processed lines: 20000000\n",
      "Processed lines: 20000000\n",
      "Processed lines: 20000000\n",
      "Processed lines: 20000000\n",
      "Processed lines: 20000000\n",
      "Processed lines: 20000000\n",
      "Processed lines: 20000000\n",
      "Processed lines: 20000000\n",
      "Processed lines: 20000000\n",
      "Processed lines: 20000000\n",
      "Processed lines: 20000000\n",
      "Processed lines: 20000000\n",
      "Processed lines: 20000000\n",
      "Processed lines: 20000000\n",
      "Processed lines: 20000000\n",
      "Processed lines: 20000000\n",
      "Processed lines: 20000000\n",
      "Processed lines: 20000000\n",
      "Processed lines: 20000000\n",
      "Processed lines: 20000000\n",
      "Processed lines: 20000000\n",
      "Processed lines: 20000000\n",
      "Processed lines: 20000000\n",
      "Processed lines: 20000000\n",
      "Processed lines: 20000000\n",
      "Processed lines: 20000000\n",
      "Processed lines: 20000000\n",
      "Processed lines: 20000000\n",
      "Processed lines: 20000000\n",
      "Processed lines: 20000000\n",
      "Processed lines: 20000000\n",
      "Processed lines: 20500000\n",
      "Processed lines: 20500000\n",
      "Processed lines: 20000000\n",
      "Processed lines: 20500000\n",
      "Processed lines: 20500000\n",
      "Processed lines: 20500000\n",
      "Processed lines: 20500000\n",
      "Processed lines: 20500000\n",
      "Processed lines: 20500000\n",
      "Processed lines: 20500000\n",
      "Processed lines: 20500000\n",
      "Processed lines: 20500000\n",
      "Processed lines: 20500000\n",
      "Processed lines: 20500000\n",
      "Processed lines: 20500000\n",
      "Processed lines: 20500000\n",
      "Processed lines: 20500000\n",
      "Processed lines: 20500000\n",
      "Processed lines: 20500000\n",
      "Processed lines: 20500000\n",
      "Processed lines: 20500000\n",
      "Processed lines: 20500000\n",
      "Processed lines: 20500000\n",
      "Processed lines: 20500000\n",
      "Processed lines: 20500000\n",
      "Processed lines: 20500000\n",
      "Processed lines: 20500000\n",
      "Processed lines: 20500000\n",
      "Processed lines: 20500000\n",
      "Processed lines: 20500000\n",
      "Processed lines: 20500000\n",
      "Processed lines: 20500000\n",
      "Processed lines: 20500000\n",
      "Processed lines: 20500000\n",
      "Processed lines: 20500000\n",
      "Processed lines: 21000000\n",
      "Processed lines: 21000000\n",
      "Processed lines: 21000000\n",
      "Processed lines: 21000000\n",
      "Processed lines: 21000000\n",
      "Processed lines: 21000000\n",
      "Processed lines: 21000000\n",
      "Processed lines: 21000000\n",
      "Processed lines: 21000000\n",
      "Processed lines: 21000000\n",
      "Processed lines: 21000000\n",
      "Processed lines: 21000000\n",
      "Processed lines: 21000000\n",
      "Processed lines: 21000000\n",
      "Processed lines: 21000000\n",
      "Processed lines: 21000000\n",
      "Processed lines: 21000000\n",
      "Processed lines: 21000000\n",
      "Processed lines: 21000000\n",
      "Processed lines: 21000000\n",
      "Processed lines: 21000000\n",
      "Processed lines: 21000000\n",
      "Processed lines: 21000000\n",
      "Processed lines: 21000000\n",
      "Processed lines: 21000000\n",
      "Processed lines: 21000000\n",
      "Processed lines: 21000000\n",
      "Processed lines: 21000000\n",
      "Processed lines: 21000000\n",
      "Processed lines: 21000000\n",
      "Processed lines: 21000000\n",
      "Processed lines: 21000000\n",
      "Processed lines: 21000000\n",
      "Processed lines: 21000000\n",
      "Processed lines: 21500000\n",
      "Processed lines: 21500000\n",
      "Processed lines: 21500000\n",
      "Processed lines: 21500000\n",
      "Processed lines: 21500000\n",
      "Processed lines: 21500000\n",
      "Processed lines: 21500000\n",
      "Processed lines: 21500000\n",
      "Processed lines: 21500000\n",
      "Processed lines: 21500000\n",
      "Processed lines: 21500000\n",
      "Processed lines: 21500000\n",
      "Processed lines: 21500000\n",
      "Processed lines: 21500000\n",
      "Processed lines: 21500000\n",
      "Processed lines: 21500000\n",
      "Processed lines: 21500000\n",
      "Processed lines: 21500000\n",
      "Processed lines: 21500000\n",
      "Processed lines: 21500000\n",
      "Processed lines: 21500000\n",
      "Processed lines: 21500000\n",
      "Processed lines: 21500000\n",
      "Processed lines: 21500000\n",
      "Processed lines: 21500000\n",
      "Processed lines: 21500000\n",
      "Processed lines: 21500000\n",
      "Processed lines: 21500000\n",
      "Processed lines: 21500000\n",
      "Processed lines: 21500000\n",
      "Processed lines: 21500000\n",
      "Processed lines: 21500000\n",
      "Processed lines: 21500000\n",
      "Processed lines: 22000000\n",
      "Processed lines: 22000000\n",
      "Processed lines: 21500000\n",
      "Processed lines: 22000000\n",
      "Processed lines: 22000000\n",
      "Processed lines: 22000000\n",
      "Processed lines: 22000000\n",
      "Processed lines: 22000000\n",
      "Processed lines: 22000000\n",
      "Processed lines: 22000000\n",
      "Processed lines: 22000000\n",
      "Processed lines: 22000000\n",
      "Processed lines: 22000000\n",
      "Processed lines: 22000000\n",
      "Processed lines: 22000000\n",
      "Processed lines: 22000000\n",
      "Processed lines: 22000000\n",
      "Processed lines: 22000000\n",
      "Processed lines: 22000000\n",
      "Processed lines: 22000000\n",
      "Processed lines: 22000000\n",
      "Processed lines: 22000000\n",
      "Processed lines: 22000000\n",
      "Processed lines: 22000000\n",
      "Processed lines: 22000000\n",
      "Processed lines: 22000000\n",
      "Processed lines: 22000000\n",
      "Processed lines: 22000000\n",
      "Processed lines: 22000000\n",
      "Processed lines: 22000000\n",
      "Processed lines: 22000000\n",
      "Processed lines: 22000000\n",
      "Processed lines: 22000000\n",
      "Processed lines: 22000000\n",
      "Processed lines: 22500000\n",
      "Processed lines: 22000000\n",
      "Processed lines: 22500000\n",
      "Processed lines: 22500000\n",
      "Processed lines: 22500000\n",
      "Processed lines: 22500000\n",
      "Processed lines: 22500000\n",
      "Processed lines: 22500000\n",
      "Processed lines: 22500000\n",
      "Processed lines: 22500000\n",
      "Processed lines: 22500000\n",
      "Processed lines: 22500000\n",
      "Processed lines: 22500000\n",
      "Processed lines: 22500000\n",
      "Processed lines: 22500000\n",
      "Processed lines: 22500000\n",
      "Processed lines: 22500000\n",
      "Processed lines: 22500000\n",
      "Processed lines: 22500000\n",
      "Processed lines: 22500000\n",
      "Processed lines: 22500000\n",
      "Processed lines: 22500000\n",
      "Processed lines: 22500000\n",
      "Processed lines: 22500000\n",
      "Processed lines: 22500000\n",
      "Processed lines: 22500000\n",
      "Processed lines: 22500000\n",
      "Processed lines: 22500000\n",
      "Processed lines: 22500000\n",
      "Processed lines: 22500000\n",
      "Processed lines: 22500000\n",
      "Processed lines: 22500000\n",
      "Processed lines: 22500000\n",
      "Processed lines: 22500000\n",
      "Processed lines: 23000000\n",
      "Processed lines: 23000000\n",
      "Processed lines: 22500000\n",
      "Processed lines: 23000000\n",
      "Processed lines: 23000000\n",
      "Processed lines: 23000000\n",
      "Processed lines: 23000000\n",
      "Processed lines: 23000000\n",
      "Processed lines: 23000000\n",
      "Processed lines: 23000000\n",
      "Processed lines: 23000000\n",
      "Processed lines: 23000000\n",
      "Processed lines: 23000000\n",
      "Processed lines: 23000000\n",
      "Processed lines: 23000000\n",
      "Processed lines: 23000000\n",
      "Processed lines: 23000000\n",
      "Processed lines: 23000000\n",
      "Processed lines: 23000000\n",
      "Processed lines: 23000000\n",
      "Processed lines: 23000000\n",
      "Processed lines: 23000000\n",
      "Processed lines: 23000000\n",
      "Processed lines: 23000000\n",
      "Processed lines: 23000000\n",
      "Processed lines: 23000000\n",
      "Processed lines: 23000000\n",
      "Processed lines: 23000000\n",
      "Processed lines: 23000000\n",
      "Processed lines: 23000000\n",
      "Processed lines: 23000000\n",
      "Processed lines: 23000000\n",
      "Processed lines: 23000000\n",
      "Processed lines: 23000000\n",
      "Processed lines: 23000000\n",
      "Processed lines: 23500000\n",
      "Processed lines: 23500000\n",
      "Processed lines: 23500000\n",
      "Processed lines: 23500000\n",
      "Processed lines: 23500000\n",
      "Processed lines: 23500000\n",
      "Processed lines: 23500000\n",
      "Processed lines: 23500000\n",
      "Processed lines: 23500000\n",
      "Processed lines: 23500000\n",
      "Processed lines: 23500000\n",
      "Processed lines: 23500000\n",
      "Processed lines: 23500000\n",
      "Processed lines: 23500000\n",
      "Processed lines: 23500000\n",
      "Processed lines: 23500000\n",
      "Processed lines: 23500000\n",
      "Processed lines: 23500000\n",
      "Processed lines: 23500000\n",
      "Processed lines: 23500000\n",
      "Processed lines: 23500000\n",
      "Processed lines: 23500000\n",
      "Processed lines: 23500000\n",
      "Processed lines: 23500000\n",
      "Processed lines: 23500000\n",
      "Processed lines: 23500000\n",
      "Processed lines: 23500000\n",
      "Processed lines: 23500000\n",
      "Processed lines: 23500000\n",
      "Processed lines: 23500000\n",
      "Processed lines: 23500000\n",
      "Processed lines: 23500000\n",
      "Processed lines: 23500000\n",
      "Processed lines: 23500000\n",
      "Processed lines: 24000000\n",
      "Processed lines: 24000000\n",
      "Processed lines: 24000000\n",
      "Processed lines: 24000000\n",
      "Processed lines: 24000000\n",
      "Processed lines: 24000000\n",
      "Processed lines: 24000000\n",
      "Processed lines: 24000000\n",
      "Processed lines: 24000000\n",
      "Processed lines: 24000000\n",
      "Processed lines: 24000000\n",
      "Processed lines: 24000000\n",
      "Processed lines: 24000000\n",
      "Processed lines: 24000000\n",
      "Processed lines: 24000000\n",
      "Processed lines: 24000000\n",
      "Processed lines: 24000000\n",
      "Processed lines: 24000000\n",
      "Processed lines: 24000000\n",
      "Processed lines: 24000000\n",
      "Processed lines: 24000000\n",
      "Processed lines: 24000000\n",
      "Processed lines: 24000000\n",
      "Processed lines: 24000000\n",
      "Processed lines: 24000000\n",
      "Processed lines: 24000000\n",
      "Processed lines: 24000000\n",
      "Processed lines: 24000000\n",
      "Processed lines: 24000000\n",
      "Processed lines: 24000000\n",
      "Processed lines: 24000000\n",
      "Processed lines: 24000000\n",
      "Processed lines: 24000000\n",
      "Processed lines: 24000000\n",
      "Processed lines: 24500000\n",
      "Processed lines: 24500000\n",
      "Processed lines: 24500000\n",
      "Processed lines: 24500000\n",
      "Processed lines: 24500000\n",
      "Processed lines: 24500000\n",
      "Processed lines: 24500000\n",
      "Processed lines: 24500000\n",
      "Processed lines: 24500000\n",
      "Processed lines: 24500000\n",
      "Processed lines: 24500000\n",
      "Processed lines: 24500000\n",
      "Processed lines: 24500000\n",
      "Processed lines: 24500000\n",
      "Processed lines: 24500000\n",
      "Processed lines: 24500000\n",
      "Processed lines: 24500000\n",
      "Processed lines: 24500000\n",
      "Processed lines: 24500000\n",
      "Processed lines: 24500000\n",
      "Processed lines: 24500000\n",
      "Processed lines: 24500000\n",
      "Processed lines: 24500000\n",
      "Processed lines: 24500000\n",
      "Processed lines: 24500000\n",
      "Processed lines: 24500000\n",
      "Processed lines: 24500000\n",
      "Processed lines: 24500000\n",
      "Processed lines: 24500000\n",
      "Processed lines: 24500000\n",
      "Processed lines: 24500000\n",
      "Processed lines: 24500000\n",
      "Processed lines: 24500000\n",
      "Processed lines: 24500000\n",
      "Processed lines: 25000000\n",
      "Processed lines: 25000000\n",
      "Processed lines: 25000000\n",
      "Processed lines: 25000000\n",
      "Processed lines: 25000000\n",
      "Processed lines: 25000000\n",
      "Processed lines: 25000000\n",
      "Processed lines: 25000000\n",
      "Processed lines: 25000000\n",
      "Processed lines: 25000000\n",
      "Processed lines: 25000000\n",
      "Processed lines: 25000000\n",
      "Processed lines: 25000000\n",
      "Processed lines: 25000000\n",
      "Processed lines: 25000000\n",
      "Processed lines: 25000000\n",
      "Processed lines: 25000000\n",
      "Processed lines: 25000000\n",
      "Processed lines: 25000000\n",
      "Processed lines: 25000000\n",
      "Processed lines: 25000000\n",
      "Processed lines: 25000000\n",
      "Processed lines: 25000000\n",
      "Processed lines: 25000000\n",
      "Processed lines: 25000000\n",
      "Processed lines: 25000000\n",
      "Processed lines: 25000000\n",
      "Processed lines: 25000000\n",
      "Processed lines: 25000000\n",
      "Processed lines: 25000000\n",
      "Processed lines: 25000000\n",
      "Processed lines: 25000000\n",
      "Processed lines: 25000000\n",
      "Processed lines: 25000000\n",
      "Processed lines: 25500000\n",
      "Processed lines: 25500000\n",
      "Processed lines: 25500000\n",
      "Processed lines: 25500000\n",
      "Processed lines: 25500000\n",
      "Processed lines: 25500000\n",
      "Processed lines: 25500000\n",
      "Processed lines: 25500000\n",
      "Processed lines: 25500000\n",
      "Processed lines: 25500000\n",
      "Processed lines: 25500000\n",
      "Processed lines: 25500000\n",
      "Processed lines: 25500000\n",
      "Processed lines: 25500000\n",
      "Processed lines: 25500000\n",
      "Processed lines: 25500000\n",
      "Processed lines: 25500000\n",
      "Processed lines: 25500000\n",
      "Processed lines: 25500000\n",
      "Processed lines: 25500000\n",
      "Processed lines: 25500000\n",
      "Processed lines: 25500000\n",
      "Processed lines: 25500000\n",
      "Processed lines: 25500000\n",
      "Processed lines: 25500000\n",
      "Processed lines: 25500000\n",
      "Processed lines: 25500000\n",
      "Processed lines: 25500000\n",
      "Processed lines: 25500000\n",
      "Processed lines: 25500000\n",
      "Processed lines: 25500000\n",
      "Processed lines: 25500000\n",
      "Processed lines: 25500000\n",
      "Processed lines: 25500000\n",
      "Processed lines: 26000000\n",
      "Processed lines: 26000000\n",
      "Processed lines: 26000000\n",
      "Processed lines: 26000000\n",
      "Processed lines: 26000000\n",
      "Processed lines: 26000000\n",
      "Processed lines: 26000000\n",
      "Processed lines: 26000000\n",
      "Processed lines: 26000000\n",
      "Processed lines: 26000000\n",
      "Processed lines: 26000000\n",
      "Processed lines: 26000000\n",
      "Processed lines: 26000000\n",
      "Processed lines: 26000000\n",
      "Processed lines: 26000000\n",
      "Processed lines: 26000000\n",
      "Processed lines: 26000000\n",
      "Processed lines: 26000000\n",
      "Processed lines: 26000000\n",
      "Processed lines: 26000000\n",
      "Processed lines: 26000000\n",
      "Processed lines: 26000000\n",
      "Processed lines: 26000000\n",
      "Processed lines: 26000000\n",
      "Processed lines: 26000000\n",
      "Processed lines: 26000000\n",
      "Processed lines: 26000000\n",
      "Processed lines: 26000000\n",
      "Processed lines: 26000000\n",
      "Processed lines: 26000000\n",
      "Processed lines: 26000000\n",
      "Processed lines: 26000000\n",
      "Processed lines: 26000000\n",
      "Processed lines: 26000000\n",
      "Processed lines: 26500000\n",
      "Processed lines: 26500000\n",
      "Processed lines: 26500000\n",
      "Processed lines: 26500000\n",
      "Processed lines: 26500000\n",
      "Processed lines: 26500000\n",
      "Processed lines: 26500000\n",
      "Processed lines: 26500000\n",
      "Processed lines: 26500000\n",
      "Processed lines: 26500000\n",
      "Processed lines: 26500000\n",
      "Processed lines: 26500000\n",
      "Processed lines: 26500000\n",
      "Processed lines: 26500000\n",
      "Processed lines: 26500000\n",
      "Processed lines: 26500000\n",
      "Processed lines: 26500000\n",
      "Processed lines: 26500000\n",
      "Processed lines: 26500000\n",
      "Processed lines: 26500000\n",
      "Processed lines: 26500000\n",
      "Processed lines: 26500000\n",
      "Processed lines: 26500000\n",
      "Processed lines: 26500000\n",
      "Processed lines: 26500000\n",
      "Processed lines: 26500000\n",
      "Processed lines: 26500000\n",
      "Processed lines: 26500000\n",
      "Processed lines: 26500000\n",
      "Processed lines: 26500000\n",
      "Processed lines: 26500000\n",
      "Processed lines: 26500000\n",
      "Processed lines: 26500000\n",
      "Processed lines: 26500000\n",
      "Processed lines: 27000000\n",
      "Processed lines: 27000000\n",
      "Processed lines: 27000000\n",
      "Processed lines: 27000000\n",
      "Processed lines: 27000000\n",
      "Processed lines: 27000000\n",
      "Processed lines: 27000000\n",
      "Processed lines: 27000000\n",
      "Processed lines: 27000000\n",
      "Processed lines: 27000000\n",
      "Processed lines: 27000000\n",
      "Processed lines: 27000000\n",
      "Processed lines: 27000000\n",
      "Processed lines: 27000000\n",
      "Processed lines: 27000000\n",
      "Processed lines: 27000000\n",
      "Processed lines: 27000000\n",
      "Processed lines: 27000000\n",
      "Processed lines: 27000000\n",
      "Processed lines: 27000000\n",
      "Processed lines: 27000000\n",
      "Processed lines: 27000000\n",
      "Processed lines: 27000000\n",
      "Processed lines: 27000000\n",
      "Processed lines: 27000000\n",
      "Processed lines: 27000000\n",
      "Processed lines: 27000000\n",
      "Processed lines: 27000000\n",
      "Processed lines: 27000000\n",
      "Processed lines: 27000000\n",
      "Processed lines: 27000000\n",
      "Processed lines: 27000000\n",
      "Processed lines: 27000000\n",
      "Processed lines: 27000000\n",
      "Processed lines: 27500000\n",
      "Processed lines: 27500000\n",
      "Processed lines: 27500000\n",
      "Processed lines: 27500000\n",
      "Processed lines: 27500000\n",
      "Processed lines: 27500000\n",
      "Processed lines: 27500000\n",
      "Processed lines: 27500000\n",
      "Processed lines: 27500000\n",
      "Processed lines: 27500000\n",
      "Processed lines: 27500000\n",
      "Processed lines: 27500000\n",
      "Processed lines: 27500000\n",
      "Processed lines: 27500000\n",
      "Processed lines: 27500000\n",
      "Processed lines: 27500000\n",
      "Processed lines: 27500000\n",
      "Processed lines: 27500000\n",
      "Processed lines: 27500000\n",
      "Processed lines: 27500000\n",
      "Processed lines: 27500000\n",
      "Processed lines: 27500000\n",
      "Processed lines: 27500000\n",
      "Processed lines: 27500000\n",
      "Processed lines: 27500000\n",
      "Processed lines: 27500000\n",
      "Processed lines: 27500000\n",
      "Processed lines: 27500000\n",
      "Processed lines: 27500000\n",
      "Processed lines: 27500000\n",
      "Processed lines: 27500000\n",
      "Processed lines: 27500000\n",
      "Processed lines: 27500000\n",
      "Processed lines: 27500000\n",
      "Processed lines: 28000000\n",
      "Processed lines: 28000000\n",
      "Processed lines: 28000000\n",
      "Processed lines: 28000000\n",
      "Processed lines: 28000000\n",
      "Processed lines: 28000000\n",
      "Processed lines: 28000000\n",
      "Processed lines: 28000000\n",
      "Processed lines: 28000000\n",
      "Processed lines: 28000000\n",
      "Processed lines: 28000000\n",
      "Processed lines: 28000000\n",
      "Processed lines: 28000000\n",
      "Processed lines: 28000000\n",
      "Processed lines: 28000000\n",
      "Processed lines: 28000000\n",
      "Processed lines: 28000000\n",
      "Processed lines: 28000000\n",
      "Processed lines: 28000000\n",
      "Processed lines: 28000000\n",
      "Processed lines: 28000000\n",
      "Processed lines: 28000000\n",
      "Processed lines: 28000000\n",
      "Processed lines: 28000000\n",
      "Processed lines: 28000000\n",
      "Processed lines: 28000000\n",
      "Processed lines: 28000000\n",
      "Processed lines: 28000000\n",
      "Processed lines: 28000000\n",
      "Processed lines: 28000000\n",
      "Processed lines: 28000000\n",
      "Processed lines: 28000000\n",
      "Processed lines: 28000000\n",
      "Processed lines: 28000000\n",
      "Processed lines: 28500000\n",
      "Processed lines: 28500000\n",
      "Processed lines: 28500000\n",
      "Processed lines: 28500000\n",
      "Processed lines: 28500000\n",
      "Processed lines: 28500000\n",
      "Processed lines: 28500000\n",
      "Processed lines: 28500000\n",
      "Processed lines: 28500000\n",
      "Processed lines: 28500000\n",
      "Processed lines: 28500000\n",
      "Processed lines: 28500000\n",
      "Processed lines: 28500000\n",
      "Processed lines: 28500000\n",
      "Processed lines: 28500000\n",
      "Processed lines: 28500000\n",
      "Processed lines: 28500000\n",
      "Processed lines: 28500000\n",
      "Processed lines: 28500000\n",
      "Processed lines: 28500000\n",
      "Processed lines: 28500000\n",
      "Processed lines: 28500000\n",
      "Processed lines: 28500000\n",
      "Processed lines: 28500000\n",
      "Processed lines: 28500000\n",
      "Processed lines: 28500000\n",
      "Processed lines: 28500000\n",
      "Processed lines: 28500000\n",
      "Processed lines: 28500000\n",
      "Processed lines: 28500000\n",
      "Processed lines: 28500000\n",
      "Processed lines: 28500000\n",
      "Processed lines: 28500000\n",
      "Processed lines: 28500000\n",
      "Processed lines: 29000000\n",
      "Processed lines: 29000000\n",
      "Processed lines: 29000000\n",
      "Processed lines: 29000000\n",
      "Processed lines: 29000000\n",
      "Processed lines: 29000000\n",
      "Processed lines: 29000000\n",
      "Processed lines: 29000000\n",
      "Processed lines: 29000000\n",
      "Processed lines: 29000000\n",
      "Processed lines: 29000000\n",
      "Processed lines: 29000000\n",
      "Processed lines: 29000000\n",
      "Processed lines: 29000000\n",
      "Processed lines: 29000000\n",
      "Processed lines: 29000000\n",
      "Processed lines: 29000000\n",
      "Processed lines: 29000000\n",
      "Processed lines: 29000000\n",
      "Processed lines: 29000000\n",
      "Processed lines: 29000000\n",
      "Processed lines: 29000000\n",
      "Processed lines: 29000000\n",
      "Processed lines: 29000000\n",
      "Processed lines: 29000000\n",
      "Processed lines: 29000000\n",
      "Processed lines: 29000000\n",
      "Processed lines: 29000000\n",
      "Processed lines: 29000000\n",
      "Processed lines: 29000000\n",
      "Processed lines: 29000000\n",
      "Processed lines: 29000000\n",
      "Processed lines: 29000000\n",
      "Processed lines: 29000000\n",
      "Processed lines: 29500000\n",
      "Processed lines: 29500000\n",
      "Processed lines: 29500000\n",
      "Processed lines: 29500000\n",
      "Processed lines: 29500000\n",
      "Processed lines: 29500000\n",
      "Processed lines: 29500000\n",
      "Processed lines: 29500000\n",
      "Processed lines: 29500000\n",
      "Processed lines: 29500000\n",
      "Processed lines: 29500000\n",
      "Processed lines: 29500000\n",
      "Processed lines: 29500000\n",
      "Processed lines: 29500000\n",
      "Processed lines: 29500000\n",
      "Processed lines: 29500000\n",
      "Processed lines: 29500000\n",
      "Processed lines: 29500000\n",
      "Processed lines: 29500000\n",
      "Processed lines: 29500000\n",
      "Processed lines: 29500000\n",
      "Processed lines: 29500000\n",
      "Processed lines: 29500000\n",
      "Processed lines: 29500000\n",
      "Processed lines: 29500000\n",
      "Processed lines: 29500000\n",
      "Processed lines: 29500000\n",
      "Processed lines: 29500000\n",
      "Processed lines: 29500000\n",
      "Processed lines: 29500000\n",
      "Processed lines: 29500000\n",
      "Processed lines: 29500000\n",
      "Processed lines: 29500000\n",
      "Processed lines: 29500000\n",
      "Processed lines: 30000000\n",
      "Processed lines: 30000000\n",
      "Processed lines: 30000000\n",
      "Processed lines: 30000000\n",
      "Processed lines: 30000000\n",
      "Processed lines: 30000000\n",
      "Processed lines: 30000000\n",
      "Processed lines: 30000000\n",
      "Processed lines: 30000000\n",
      "Processed lines: 30000000\n",
      "Processed lines: 30000000\n",
      "Processed lines: 30000000\n",
      "Processed lines: 30000000\n",
      "Processed lines: 30000000\n",
      "Processed lines: 30000000\n",
      "Processed lines: 30000000\n",
      "Processed lines: 30000000\n",
      "Processed lines: 30000000\n",
      "Processed lines: 30000000\n",
      "Processed lines: 30000000\n",
      "Processed lines: 30000000\n",
      "Processed lines: 30000000\n",
      "Processed lines: 30000000\n",
      "Processed lines: 30000000\n",
      "Processed lines: 30000000\n",
      "Processed lines: 30000000\n",
      "Processed lines: 30000000\n",
      "Processed lines: 30000000\n",
      "Processed lines: 30000000\n",
      "Processed lines: 30000000\n",
      "Processed lines: 30000000\n",
      "Processed lines: 30000000\n",
      "Processed lines: 30000000\n",
      "Processed lines: 30000000\n",
      "Processed lines: 30500000\n",
      "Processed lines: 30500000\n",
      "Processed lines: 30500000\n",
      "Processed lines: 30500000\n",
      "Processed lines: 30500000\n",
      "Processed lines: 30500000\n",
      "Processed lines: 30500000\n",
      "Processed lines: 30500000\n",
      "Processed lines: 30500000\n",
      "Processed lines: 30500000\n",
      "Processed lines: 30500000\n",
      "Processed lines: 30500000\n",
      "Processed lines: 30500000\n",
      "Processed lines: 30500000\n",
      "Processed lines: 30500000\n",
      "Processed lines: 30500000\n",
      "Processed lines: 30500000\n",
      "Processed lines: 30500000\n",
      "Processed lines: 30500000\n",
      "Processed lines: 30500000\n",
      "Processed lines: 30500000\n",
      "Processed lines: 30500000\n",
      "Processed lines: 30500000\n",
      "Processed lines: 30500000\n",
      "Processed lines: 30500000\n",
      "Processed lines: 30500000\n",
      "Processed lines: 30500000\n",
      "Processed lines: 30500000\n",
      "Processed lines: 30500000\n",
      "Processed lines: 30500000\n",
      "Processed lines: 30500000\n",
      "Processed lines: 30500000\n",
      "Processed lines: 30500000\n",
      "Processed lines: 31000000\n",
      "Processed lines: 30500000\n",
      "Processed lines: 31000000\n",
      "Processed lines: 31000000\n",
      "Processed lines: 31000000\n",
      "Processed lines: 31000000\n",
      "Processed lines: 31000000\n",
      "Processed lines: 31000000\n",
      "Processed lines: 31000000\n",
      "Processed lines: 31000000\n",
      "Processed lines: 31000000\n",
      "Processed lines: 31000000\n",
      "Processed lines: 31000000\n",
      "Processed lines: 31000000\n",
      "Processed lines: 31000000\n",
      "Processed lines: 31000000\n",
      "Processed lines: 31000000\n",
      "Processed lines: 31000000\n",
      "Processed lines: 31000000\n",
      "Processed lines: 31000000\n",
      "Processed lines: 31000000\n",
      "Processed lines: 31000000\n",
      "Processed lines: 31000000\n",
      "Processed lines: 31000000\n",
      "Processed lines: 31000000\n",
      "Processed lines: 31000000\n",
      "Processed lines: 31000000\n",
      "Processed lines: 31000000\n",
      "Processed lines: 31000000\n",
      "Processed lines: 31000000\n",
      "Processed lines: 31000000\n",
      "Processed lines: 31000000\n",
      "Processed lines: 31000000\n",
      "Processed lines: 31000000\n",
      "Processed lines: 31500000\n",
      "Processed lines: 31000000\n",
      "Processed lines: 31500000\n",
      "Processed lines: 31500000\n",
      "Processed lines: 31500000\n",
      "Processed lines: 31500000\n",
      "Processed lines: 31500000\n",
      "Processed lines: 31500000\n",
      "Processed lines: 31500000\n",
      "Processed lines: 31500000\n",
      "Processed lines: 31500000\n",
      "Processed lines: 31500000\n",
      "Processed lines: 31500000\n",
      "Processed lines: 31500000\n",
      "Processed lines: 31500000\n",
      "Processed lines: 31500000\n",
      "Processed lines: 31500000\n",
      "Processed lines: 31500000\n",
      "Processed lines: 31500000\n",
      "Processed lines: 31500000\n",
      "Processed lines: 31500000\n",
      "Processed lines: 31500000\n",
      "Processed lines: 31500000\n",
      "Processed lines: 31500000\n",
      "Processed lines: 31500000\n",
      "Processed lines: 31500000\n",
      "Processed lines: 31500000\n",
      "Processed lines: 31500000\n",
      "Processed lines: 31500000\n",
      "Processed lines: 31500000\n",
      "Processed lines: 31500000\n",
      "Processed lines: 31500000\n",
      "Processed lines: 31500000\n",
      "Processed lines: 31500000\n",
      "Processed lines: 32000000\n",
      "Processed lines: 31500000\n",
      "Processed lines: 32000000\n",
      "Processed lines: 32000000\n",
      "Processed lines: 32000000\n",
      "Processed lines: 32000000\n",
      "Processed lines: 32000000\n",
      "Processed lines: 32000000\n",
      "Processed lines: 32000000\n",
      "Processed lines: 32000000\n",
      "Processed lines: 32000000\n",
      "Processed lines: 32000000\n",
      "Processed lines: 32000000\n",
      "Processed lines: 32000000\n",
      "Processed lines: 32000000\n",
      "Processed lines: 32000000\n",
      "Processed lines: 32000000\n",
      "Processed lines: 32000000\n",
      "Processed lines: 32000000\n",
      "Processed lines: 32000000\n",
      "Processed lines: 32000000\n",
      "Processed lines: 32000000\n",
      "Processed lines: 32000000\n",
      "Processed lines: 32000000\n",
      "Processed lines: 32000000\n",
      "Processed lines: 32000000\n",
      "Processed lines: 32000000\n",
      "Processed lines: 32000000\n",
      "Processed lines: 32000000\n",
      "Processed lines: 32000000\n",
      "Processed lines: 32000000\n",
      "Processed lines: 32000000\n",
      "Processed lines: 32500000\n",
      "Processed lines: 32000000\n",
      "Processed lines: 32000000\n",
      "Processed lines: 32000000\n",
      "Processed lines: 32500000\n",
      "Processed lines: 32500000\n",
      "Processed lines: 32500000\n",
      "Processed lines: 32500000\n",
      "Processed lines: 32500000\n",
      "Processed lines: 32500000\n",
      "Processed lines: 32500000\n",
      "Processed lines: 32500000\n",
      "Processed lines: 32500000\n",
      "Processed lines: 32500000\n",
      "Processed lines: 32500000\n",
      "Processed lines: 32500000\n",
      "Processed lines: 32500000\n",
      "Processed lines: 32500000\n",
      "Processed lines: 32500000\n",
      "Processed lines: 32500000\n",
      "Processed lines: 32500000\n",
      "Processed lines: 32500000\n",
      "Processed lines: 32500000\n",
      "Processed lines: 32500000\n",
      "Processed lines: 32500000\n",
      "Processed lines: 32500000\n",
      "Processed lines: 32500000\n",
      "Processed lines: 32500000\n",
      "Processed lines: 32500000\n",
      "Processed lines: 32500000\n",
      "Processed lines: 32500000\n",
      "Processed lines: 32500000\n",
      "Processed lines: 32500000\n",
      "Processed lines: 32500000\n",
      "Processed lines: 33000000\n",
      "Processed lines: 32500000\n",
      "Processed lines: 32500000\n",
      "Processed lines: 32500000\n",
      "Processed lines: 33000000\n",
      "Processed lines: 33000000\n",
      "Processed lines: 33000000\n",
      "Processed lines: 33000000\n",
      "Processed lines: 33000000\n",
      "Processed lines: 33000000\n",
      "Processed lines: 33000000\n",
      "Processed lines: 33000000\n",
      "Processed lines: 33000000\n",
      "Processed lines: 33000000\n",
      "Processed lines: 33000000\n",
      "Processed lines: 33000000\n",
      "Processed lines: 33000000\n",
      "Processed lines: 33000000\n",
      "Processed lines: 33000000\n",
      "Processed lines: 33000000\n",
      "Processed lines: 33000000\n",
      "Processed lines: 33000000\n",
      "Processed lines: 33000000\n",
      "Processed lines: 33000000\n",
      "Processed lines: 33000000\n",
      "Processed lines: 33000000\n",
      "Processed lines: 33000000\n",
      "Processed lines: 33000000\n",
      "Processed lines: 33000000\n",
      "Processed lines: 33000000\n",
      "Processed lines: 33000000\n",
      "Processed lines: 33000000\n",
      "Processed lines: 33000000\n",
      "Processed lines: 33000000\n",
      "Processed lines: 33000000\n",
      "Processed lines: 33500000\n",
      "Processed lines: 33000000\n",
      "Processed lines: 33500000\n",
      "Processed lines: 33000000\n",
      "Processed lines: 33500000\n",
      "Processed lines: 33500000\n",
      "Processed lines: 33500000\n",
      "Processed lines: 33500000\n",
      "Processed lines: 33500000\n",
      "Processed lines: 33500000\n",
      "Processed lines: 33500000\n",
      "Processed lines: 33500000\n",
      "Processed lines: 33500000\n",
      "Processed lines: 33500000\n",
      "Processed lines: 33500000\n",
      "Processed lines: 33500000\n",
      "Processed lines: 33500000\n",
      "Processed lines: 33500000\n",
      "Processed lines: 33500000\n",
      "Processed lines: 33500000\n",
      "Processed lines: 33500000\n",
      "Processed lines: 33500000\n",
      "Processed lines: 33500000\n",
      "Processed lines: 33500000\n",
      "Processed lines: 33500000\n",
      "Processed lines: 33500000\n",
      "Processed lines: 33500000\n",
      "Processed lines: 33500000\n",
      "Processed lines: 33500000\n",
      "Processed lines: 33500000\n",
      "Processed lines: 33500000\n",
      "Processed lines: 33500000\n",
      "Processed lines: 33500000\n",
      "Processed lines: 33500000\n",
      "Processed lines: 34000000\n",
      "Processed lines: 33500000\n",
      "Processed lines: 33500000\n",
      "Processed lines: 34000000\n",
      "Processed lines: 34000000\n",
      "Processed lines: 34000000\n",
      "Processed lines: 34000000\n",
      "Processed lines: 34000000\n",
      "Processed lines: 34000000\n",
      "Processed lines: 34000000\n",
      "Processed lines: 34000000\n",
      "Processed lines: 34000000\n",
      "Processed lines: 34000000\n",
      "Processed lines: 34000000\n",
      "Processed lines: 34000000\n",
      "Processed lines: 34000000\n",
      "Processed lines: 34000000\n",
      "Processed lines: 34000000\n",
      "Processed lines: 34000000\n",
      "Processed lines: 34000000\n",
      "Processed lines: 34000000\n",
      "Processed lines: 34000000\n",
      "Processed lines: 34000000\n",
      "Processed lines: 34000000\n",
      "Processed lines: 34000000\n",
      "Processed lines: 34000000\n",
      "Processed lines: 34000000\n",
      "Processed lines: 34000000\n",
      "Processed lines: 34000000\n",
      "Processed lines: 34000000\n",
      "Processed lines: 34000000\n",
      "Processed lines: 34000000\n",
      "Processed lines: 34000000\n",
      "Processed lines: 34000000\n",
      "Processed lines: 34500000\n",
      "Processed lines: 34000000\n",
      "Processed lines: 34000000\n",
      "Processed lines: 34500000\n",
      "Processed lines: 34500000\n",
      "Processed lines: 34500000\n",
      "Processed lines: 34500000\n",
      "Processed lines: 34500000\n",
      "Processed lines: 34500000\n",
      "Processed lines: 34500000\n",
      "Processed lines: 34500000\n",
      "Processed lines: 34500000\n",
      "Processed lines: 34500000\n",
      "Processed lines: 34500000\n",
      "Processed lines: 34500000\n",
      "Processed lines: 34500000\n",
      "Processed lines: 34500000\n",
      "Processed lines: 34500000\n",
      "Processed lines: 34500000\n",
      "Processed lines: 34500000\n",
      "Processed lines: 34500000\n",
      "Processed lines: 34500000\n",
      "Processed lines: 34500000\n",
      "Processed lines: 34500000\n",
      "Processed lines: 34500000\n",
      "Processed lines: 34500000\n",
      "Processed lines: 34500000\n",
      "Processed lines: 34500000\n",
      "Processed lines: 34500000\n",
      "Processed lines: 34500000\n",
      "Processed lines: 34500000\n",
      "Processed lines: 34500000\n",
      "Processed lines: 34500000\n",
      "Processed lines: 34500000\n",
      "Processed lines: 35000000\n",
      "Processed lines: 34500000\n",
      "Processed lines: 34500000\n",
      "Processed lines: 35000000\n",
      "Processed lines: 35000000\n",
      "Processed lines: 35000000\n",
      "Processed lines: 35000000\n",
      "Processed lines: 35000000\n",
      "Processed lines: 35000000\n",
      "Processed lines: 35000000\n",
      "Processed lines: 35000000\n",
      "Processed lines: 35000000\n",
      "Processed lines: 35000000\n",
      "Processed lines: 35000000\n",
      "Processed lines: 35000000\n",
      "Processed lines: 35000000\n",
      "Processed lines: 35000000\n",
      "Processed lines: 35000000\n",
      "Processed lines: 35000000\n",
      "Processed lines: 35000000\n",
      "Processed lines: 35000000\n",
      "Processed lines: 35000000\n",
      "Processed lines: 35000000\n",
      "Processed lines: 35000000\n",
      "Processed lines: 35000000\n",
      "Processed lines: 35000000\n",
      "Processed lines: 35000000\n",
      "Processed lines: 35000000\n",
      "Processed lines: 35000000\n",
      "Processed lines: 35000000\n",
      "Processed lines: 35000000\n",
      "Processed lines: 35000000\n",
      "Processed lines: 35000000\n",
      "Processed lines: 35000000\n",
      "Processed lines: 35500000\n",
      "Processed lines: 35000000\n",
      "Processed lines: 35500000\n",
      "Processed lines: 35500000\n",
      "Processed lines: 35000000\n",
      "Processed lines: 35500000\n",
      "Processed lines: 35500000\n",
      "Processed lines: 35500000\n",
      "Processed lines: 35500000\n",
      "Processed lines: 35500000\n",
      "Processed lines: 35500000\n",
      "Processed lines: 35500000\n",
      "Processed lines: 35500000\n",
      "Processed lines: 35500000\n",
      "Processed lines: 35500000\n",
      "Processed lines: 35500000\n",
      "Processed lines: 35500000\n",
      "Processed lines: 35500000\n",
      "Processed lines: 35500000\n",
      "Processed lines: 35500000\n",
      "Processed lines: 35500000\n",
      "Processed lines: 35500000\n",
      "Processed lines: 35500000\n",
      "Processed lines: 35500000\n",
      "Processed lines: 35500000\n",
      "Processed lines: 35500000\n",
      "Processed lines: 35500000\n",
      "Processed lines: 35500000\n",
      "Processed lines: 35500000\n",
      "Processed lines: 35500000\n",
      "Processed lines: 35500000\n",
      "Processed lines: 35500000\n",
      "Processed lines: 35500000\n",
      "Processed lines: 35500000\n",
      "Processed lines: 36000000\n",
      "Processed lines: 35500000\n",
      "Processed lines: 36000000\n",
      "Processed lines: 36000000\n",
      "Processed lines: 35500000\n",
      "Processed lines: 36000000\n",
      "Processed lines: 36000000\n",
      "Processed lines: 36000000\n",
      "Processed lines: 36000000\n",
      "Processed lines: 36000000\n",
      "Processed lines: 36000000\n",
      "Processed lines: 36000000\n",
      "Processed lines: 36000000\n",
      "Processed lines: 36000000\n",
      "Processed lines: 36000000\n",
      "Processed lines: 36000000\n",
      "Processed lines: 36000000\n",
      "Processed lines: 36000000\n",
      "Processed lines: 36000000\n",
      "Processed lines: 36000000\n",
      "Processed lines: 36000000\n",
      "Processed lines: 36000000\n",
      "Processed lines: 36000000\n",
      "Processed lines: 36000000\n",
      "Processed lines: 36000000\n",
      "Processed lines: 36000000\n",
      "Processed lines: 36000000\n",
      "Processed lines: 36000000\n",
      "Processed lines: 36000000\n",
      "Processed lines: 36000000\n",
      "Processed lines: 36000000\n",
      "Processed lines: 36000000\n",
      "Processed lines: 36000000\n",
      "Processed lines: 36000000\n",
      "Processed lines: 36500000\n",
      "Processed lines: 36000000\n",
      "Processed lines: 36500000\n",
      "Processed lines: 36500000\n",
      "Processed lines: 36500000\n",
      "Processed lines: 36000000\n",
      "Processed lines: 36500000\n",
      "Processed lines: 36500000\n",
      "Processed lines: 36500000\n",
      "Processed lines: 36500000\n",
      "Processed lines: 36500000\n",
      "Processed lines: 36500000\n",
      "Processed lines: 36500000\n",
      "Processed lines: 36500000\n",
      "Processed lines: 36500000\n",
      "Processed lines: 36500000\n",
      "Processed lines: 36500000\n",
      "Processed lines: 36500000\n",
      "Processed lines: 36500000\n",
      "Processed lines: 36500000\n",
      "Processed lines: 36500000\n",
      "Processed lines: 36500000\n",
      "Processed lines: 36500000\n",
      "Processed lines: 36500000\n",
      "Processed lines: 36500000\n",
      "Processed lines: 36500000\n",
      "Processed lines: 36500000\n",
      "Processed lines: 36500000\n",
      "Processed lines: 36500000\n",
      "Processed lines: 36500000\n",
      "Processed lines: 36500000\n",
      "Processed lines: 36500000\n",
      "Processed lines: 36500000\n",
      "Processed lines: 37000000\n",
      "Processed lines: 36500000\n",
      "Processed lines: 36500000\n",
      "Processed lines: 37000000\n",
      "Processed lines: 37000000\n",
      "Processed lines: 37000000\n",
      "Processed lines: 36500000\n",
      "Processed lines: 37000000\n",
      "Processed lines: 37000000\n",
      "Processed lines: 37000000\n",
      "Processed lines: 37000000\n",
      "Processed lines: 37000000\n",
      "Processed lines: 37000000\n",
      "Processed lines: 37000000\n",
      "Processed lines: 37000000\n",
      "Processed lines: 37000000\n",
      "Processed lines: 37000000\n",
      "Processed lines: 37000000\n",
      "Processed lines: 37000000\n",
      "Processed lines: 37000000\n",
      "Processed lines: 37000000\n",
      "Processed lines: 37000000\n",
      "Processed lines: 37000000\n",
      "Processed lines: 37000000\n",
      "Processed lines: 37000000\n",
      "Processed lines: 37000000\n",
      "Processed lines: 37000000\n",
      "Processed lines: 37000000\n",
      "Processed lines: 37000000\n",
      "Processed lines: 37000000\n",
      "Processed lines: 37000000\n",
      "Processed lines: 37000000\n",
      "Processed lines: 37000000\n",
      "Processed lines: 37000000\n",
      "Processed lines: 37000000\n",
      "Processed lines: 37500000\n",
      "Processed lines: 37000000\n",
      "Processed lines: 37500000\n",
      "Processed lines: 37500000\n",
      "Processed lines: 37500000\n",
      "Processed lines: 37000000\n",
      "Processed lines: 37500000\n",
      "Processed lines: 37500000\n",
      "Processed lines: 37500000\n",
      "Processed lines: 37500000\n",
      "Processed lines: 37500000\n",
      "Processed lines: 37500000\n",
      "Processed lines: 37500000\n",
      "Processed lines: 37500000\n",
      "Processed lines: 37500000\n",
      "Processed lines: 37500000\n",
      "Processed lines: 37500000\n",
      "Processed lines: 37500000\n",
      "Processed lines: 37500000\n",
      "Processed lines: 37500000\n",
      "Processed lines: 37500000\n",
      "Processed lines: 38000000\n",
      "Processed lines: 37500000\n",
      "Processed lines: 37500000\n",
      "Processed lines: 37500000\n",
      "Processed lines: 37500000\n",
      "Processed lines: 37500000\n",
      "Processed lines: 37500000\n",
      "Processed lines: 37500000\n",
      "Processed lines: 37500000\n",
      "Processed lines: 38000000\n",
      "Processed lines: 37500000\n",
      "Processed lines: 37500000\n",
      "Processed lines: 38000000\n",
      "Processed lines: 37500000\n",
      "Processed lines: 37500000\n",
      "Processed lines: 37500000\n",
      "Processed lines: 38000000\n",
      "Processed lines: 38000000\n",
      "Processed lines: 37500000\n",
      "Processed lines: 38000000\n",
      "Processed lines: 38000000\n",
      "Processed lines: 38000000\n",
      "Processed lines: 38000000\n",
      "Processed lines: 38000000\n",
      "Processed lines: 38000000\n",
      "Processed lines: 38000000\n",
      "Processed lines: 37500000\n",
      "Processed lines: 38000000\n",
      "Processed lines: 38000000\n",
      "Processed lines: 38000000\n",
      "Processed lines: 38000000\n",
      "Processed lines: 38000000\n",
      "Processed lines: 38000000\n",
      "Processed lines: 38000000\n",
      "Processed lines: 38000000\n",
      "Processed lines: 38000000\n",
      "Processed lines: 38500000\n",
      "Processed lines: 38000000\n",
      "Processed lines: 38000000\n",
      "Processed lines: 38000000\n",
      "Processed lines: 38000000\n",
      "Processed lines: 38000000\n",
      "Processed lines: 38500000\n",
      "Processed lines: 38000000\n",
      "Processed lines: 38000000\n",
      "Processed lines: 38000000\n",
      "Processed lines: 38500000\n",
      "Processed lines: 38000000\n",
      "Processed lines: 38000000\n",
      "Processed lines: 38000000\n",
      "Processed lines: 38000000\n",
      "Processed lines: 38500000\n",
      "Processed lines: 38500000\n",
      "Processed lines: 38500000\n",
      "Processed lines: 38000000\n",
      "Processed lines: 38500000\n",
      "Processed lines: 38500000\n",
      "Processed lines: 38500000\n",
      "Processed lines: 38500000\n",
      "Processed lines: 38500000\n",
      "Processed lines: 38500000\n",
      "Processed lines: 38500000\n",
      "Finished processing child process. Exiting..\n",
      "Processed lines: 38500000\n",
      "Processed lines: 38500000\n",
      "Finished processing child process. Exiting..\n",
      "Processed lines: 38500000\n",
      "Processed lines: 38500000\n",
      "Finished processing child process. Exiting..\n",
      "Processed lines: 38500000\n",
      "Processed lines: 38500000\n",
      "Processed lines: 38500000\n",
      "Processed lines: 38500000\n",
      "Processed lines: 38500000\n",
      "Processed lines: 38500000\n",
      "Processed lines: 38500000\n",
      "Processed lines: 38500000\n",
      "Processed lines: 38500000\n",
      "Finished processing child process. Exiting..\n",
      "Finished processing child process. Exiting..\n",
      "Finished processing child process. Exiting..\n",
      "Processed lines: 38500000\n",
      "Processed lines: 38500000\n",
      "Finished processing child process. Exiting..\n",
      "Finished processing child process. Exiting..\n",
      "Processed lines: 38500000\n",
      "Finished processing child process. Exiting..\n",
      "Processed lines: 38500000\n",
      "Processed lines: 38500000\n",
      "Finished processing child process. Exiting..\n",
      "Processed lines: 38500000\n",
      "Processed lines: 38500000\n",
      "Finished processing child process. Exiting..\n",
      "Finished processing child process. Exiting..\n",
      "Finished processing child process. Exiting..\n",
      "Processed lines: 38500000\n",
      "Finished processing child process. Exiting..\n",
      "Finished processing child process. Exiting..\n",
      "Finished processing child process. Exiting..\n",
      "Finished processing child process. Exiting..\n",
      "Finished processing child process. Exiting..\n",
      "Finished processing child process. Exiting..\n",
      "Finished processing child process. Exiting..\n",
      "Finished processing child process. Exiting..\n",
      "Finished processing child process. Exiting..\n",
      "Finished processing child process. Exiting..\n",
      "Finished processing child process. Exiting..\n",
      "Now waiting for all child processes to complete\n",
      "Finished processing child process. Exiting..\n",
      "Finished processing child process. Exiting..\n",
      "Finished processing child process. Exiting..\n",
      "Finished processing child process. Exiting..\n",
      "Finished processing child process. Exiting..\n",
      "Finished processing child process. Exiting..\n",
      "Finished processing child process. Exiting..\n",
      "Finished processing child process. Exiting..\n",
      "Finished processing child process. Exiting..\n",
      "All child process successfully finished.\n",
      "Merging individual splitting reports into overall report: 'merged.sorted_splitting_report.txt'\n",
      "Merging from these individual files:\n",
      "merged.sorted_splitting_report.txt.1\n",
      "merged.sorted_splitting_report.txt.2\n",
      "merged.sorted_splitting_report.txt.3\n",
      "merged.sorted_splitting_report.txt.4\n",
      "merged.sorted_splitting_report.txt.5\n",
      "merged.sorted_splitting_report.txt.6\n",
      "merged.sorted_splitting_report.txt.7\n",
      "merged.sorted_splitting_report.txt.8\n",
      "merged.sorted_splitting_report.txt.9\n",
      "merged.sorted_splitting_report.txt.10\n",
      "merged.sorted_splitting_report.txt.11\n",
      "merged.sorted_splitting_report.txt.12\n",
      "merged.sorted_splitting_report.txt.13\n",
      "merged.sorted_splitting_report.txt.14\n",
      "merged.sorted_splitting_report.txt.15\n",
      "merged.sorted_splitting_report.txt.16\n",
      "merged.sorted_splitting_report.txt.17\n",
      "merged.sorted_splitting_report.txt.18\n",
      "merged.sorted_splitting_report.txt.19\n",
      "merged.sorted_splitting_report.txt.20\n",
      "merged.sorted_splitting_report.txt.21\n",
      "merged.sorted_splitting_report.txt.22\n",
      "merged.sorted_splitting_report.txt.23\n",
      "merged.sorted_splitting_report.txt.24\n",
      "merged.sorted_splitting_report.txt.25\n",
      "merged.sorted_splitting_report.txt.26\n",
      "merged.sorted_splitting_report.txt.27\n",
      "merged.sorted_splitting_report.txt.28\n",
      "merged.sorted_splitting_report.txt.29\n",
      "merged.sorted_splitting_report.txt.30\n",
      "merged.sorted_splitting_report.txt.31\n",
      "merged.sorted_splitting_report.txt.32\n",
      "merged.sorted_splitting_report.txt.33\n",
      "merged.sorted_splitting_report.txt.34\n",
      "\n",
      "\n",
      "Processed 38838094 lines in total\n",
      "Total number of methylation call strings processed: 38838094\n",
      "\n",
      "Final Cytosine Methylation Report\n",
      "=================================\n",
      "Total number of C's analysed:\t423765192\n",
      "\n",
      "Total methylated C's in CpG context:\t39814331\n",
      "Total methylated C's in CHG context:\t904868\n",
      "Total methylated C's in CHH context:\t1907124\n",
      "\n",
      "Total C to T conversions in CpG context:\t85319863\n",
      "Total C to T conversions in CHG context:\t99062264\n",
      "Total C to T conversions in CHH context:\t196756742\n",
      "\n",
      "C methylated in CpG context:\t31.8%\n",
      "C methylated in CHG context:\t0.9%\n",
      "C methylated in CHH context:\t1.0%\n",
      "\n",
      "\n",
      "\n",
      "Merging individual M-bias reports into overall M-bias statistics from these 34 individual files:\n",
      "merged.sorted_splitting_report.txt.1.mbias\n",
      "merged.sorted_splitting_report.txt.2.mbias\n",
      "merged.sorted_splitting_report.txt.3.mbias\n",
      "merged.sorted_splitting_report.txt.4.mbias\n",
      "merged.sorted_splitting_report.txt.5.mbias\n",
      "merged.sorted_splitting_report.txt.6.mbias\n",
      "merged.sorted_splitting_report.txt.7.mbias\n",
      "merged.sorted_splitting_report.txt.8.mbias\n",
      "merged.sorted_splitting_report.txt.9.mbias\n",
      "merged.sorted_splitting_report.txt.10.mbias\n",
      "merged.sorted_splitting_report.txt.11.mbias\n",
      "merged.sorted_splitting_report.txt.12.mbias\n",
      "merged.sorted_splitting_report.txt.13.mbias\n",
      "merged.sorted_splitting_report.txt.14.mbias\n",
      "merged.sorted_splitting_report.txt.15.mbias\n",
      "merged.sorted_splitting_report.txt.16.mbias\n",
      "merged.sorted_splitting_report.txt.17.mbias\n",
      "merged.sorted_splitting_report.txt.18.mbias\n",
      "merged.sorted_splitting_report.txt.19.mbias\n",
      "merged.sorted_splitting_report.txt.20.mbias\n",
      "merged.sorted_splitting_report.txt.21.mbias\n",
      "merged.sorted_splitting_report.txt.22.mbias\n",
      "merged.sorted_splitting_report.txt.23.mbias\n",
      "merged.sorted_splitting_report.txt.24.mbias\n",
      "merged.sorted_splitting_report.txt.25.mbias\n",
      "merged.sorted_splitting_report.txt.26.mbias\n",
      "merged.sorted_splitting_report.txt.27.mbias\n",
      "merged.sorted_splitting_report.txt.28.mbias\n",
      "merged.sorted_splitting_report.txt.29.mbias\n",
      "merged.sorted_splitting_report.txt.30.mbias\n",
      "merged.sorted_splitting_report.txt.31.mbias\n",
      "merged.sorted_splitting_report.txt.32.mbias\n",
      "merged.sorted_splitting_report.txt.33.mbias\n",
      "merged.sorted_splitting_report.txt.34.mbias\n",
      "\n",
      "Determining maximum read length for M-Bias plot\n",
      "Maximum read length of Read 1: 36\n",
      "\n",
      "Perl module GD::Graph::lines is not installed, skipping drawing M-bias plots (only writing out M-bias plot table)\n",
      "Determining maximum read length for M-Bias plot\n",
      "Maximum read length of Read 1: 36\n",
      "\n",
      "Perl module GD::Graph::lines is not installed, skipping drawing M-bias plots (only writing out M-bias plot table)\n",
      "Deleting unused files ...\n",
      "\n",
      "CpG_OT_merged.sorted.txt.gz contains data ->\tkept\n",
      "\n",
      "gzip: stdout: Broken pipe\n",
      "CpG_CTOT_merged.sorted.txt.gz was empty ->\tdeleted\n",
      "CpG_CTOB_merged.sorted.txt.gz was empty ->\tdeleted\n",
      "CpG_OB_merged.sorted.txt.gz contains data ->\tkept\n",
      "\n",
      "gzip: stdout: Broken pipe\n",
      "CHG_OT_merged.sorted.txt.gz contains data ->\tkept\n",
      "\n",
      "gzip: stdout: Broken pipe\n",
      "CHG_CTOT_merged.sorted.txt.gz was empty ->\tdeleted\n",
      "CHG_CTOB_merged.sorted.txt.gz was empty ->\tdeleted\n",
      "CHG_OB_merged.sorted.txt.gz contains data ->\tkept\n",
      "\n",
      "gzip: stdout: Broken pipe\n",
      "CHH_OT_merged.sorted.txt.gz contains data ->\tkept\n",
      "\n",
      "gzip: stdout: Broken pipe\n",
      "CHH_CTOT_merged.sorted.txt.gz was empty ->\tdeleted\n",
      "CHH_CTOB_merged.sorted.txt.gz was empty ->\tdeleted\n",
      "CHH_OB_merged.sorted.txt.gz contains data ->\tkept\n",
      "\n",
      "\n",
      "Using these input files: CpG_OT_merged.sorted.txt.gz CpG_OB_merged.sorted.txt.gz CHG_OT_merged.sorted.txt.gz CHG_OB_merged.sorted.txt.gz CHH_OT_merged.sorted.txt.gz CHH_OB_merged.sorted.txt.gz\n",
      "\n",
      "Summary of parameters for bismark2bedGraph conversion:\n",
      "======================================================\n",
      "bedGraph output:\t\tmerged.sorted.bedGraph.gz\n",
      "output directory:\t\t><\n",
      "remove whitespaces:\t\tno\n",
      "CX context:\t\t\tno (CpG context only, default)\n",
      "No-header selected:\t\tno\n",
      "Sorting method:\t\t\tUnix sort-based (smaller memory footprint, but slower)\n",
      "Sort buffer size:\t\t2G\n",
      "Coverage threshold:\t\t1\n",
      "=============================================================================\n",
      "Methylation information will now be written into a bedGraph and coverage file\n",
      "=============================================================================\n",
      "\n",
      "Using the following files as Input:\n",
      "/home/sagemaker-user/src/CpG_OT_merged.sorted.txt.gz\t/home/sagemaker-user/src/CpG_OB_merged.sorted.txt.gz\n",
      "\n",
      "Writing bedGraph to file: merged.sorted.bedGraph.gz\n",
      "Also writing out a coverage file including counts methylated and unmethylated residues to file: merged.sorted.bismark.cov.gz\n",
      "\n",
      "Now writing methylation information for file >>CpG_OT_merged.sorted.txt.gz<< to individual files for each chromosome\n",
      "Finished writing out individual chromosome files for CpG_OT_merged.sorted.txt.gz\n",
      "Now writing methylation information for file >>CpG_OB_merged.sorted.txt.gz<< to individual files for each chromosome\n",
      "Finished writing out individual chromosome files for CpG_OB_merged.sorted.txt.gz\n",
      "\n",
      "Collecting temporary chromosome file information... Processing the following input file(s):\n",
      "CpG_OT_merged.sorted.txt.gz.chrchr1.methXtractor.temp\n",
      "CpG_OT_merged.sorted.txt.gz.chrchr2.methXtractor.temp\n",
      "CpG_OT_merged.sorted.txt.gz.chrchr3.methXtractor.temp\n",
      "CpG_OT_merged.sorted.txt.gz.chrchr4.methXtractor.temp\n",
      "CpG_OT_merged.sorted.txt.gz.chrchr5.methXtractor.temp\n",
      "CpG_OT_merged.sorted.txt.gz.chrchr6.methXtractor.temp\n",
      "CpG_OT_merged.sorted.txt.gz.chrchr7.methXtractor.temp\n",
      "CpG_OT_merged.sorted.txt.gz.chrchrX.methXtractor.temp\n",
      "CpG_OT_merged.sorted.txt.gz.chrchr8.methXtractor.temp\n",
      "CpG_OT_merged.sorted.txt.gz.chrchr9.methXtractor.temp\n",
      "CpG_OT_merged.sorted.txt.gz.chrchr10.methXtractor.temp\n",
      "CpG_OT_merged.sorted.txt.gz.chrchr11.methXtractor.temp\n",
      "CpG_OT_merged.sorted.txt.gz.chrchr12.methXtractor.temp\n",
      "CpG_OT_merged.sorted.txt.gz.chrchr13.methXtractor.temp\n",
      "CpG_OT_merged.sorted.txt.gz.chrchr14.methXtractor.temp\n",
      "CpG_OT_merged.sorted.txt.gz.chrchr15.methXtractor.temp\n",
      "CpG_OT_merged.sorted.txt.gz.chrchr16.methXtractor.temp\n",
      "CpG_OT_merged.sorted.txt.gz.chrchr17.methXtractor.temp\n",
      "CpG_OT_merged.sorted.txt.gz.chrchr18.methXtractor.temp\n",
      "CpG_OT_merged.sorted.txt.gz.chrchr20.methXtractor.temp\n",
      "CpG_OT_merged.sorted.txt.gz.chrchr19.methXtractor.temp\n",
      "CpG_OT_merged.sorted.txt.gz.chrchr22.methXtractor.temp\n",
      "CpG_OT_merged.sorted.txt.gz.chrchr21.methXtractor.temp\n",
      "\n",
      "Sorting input file CpG_OT_merged.sorted.txt.gz.chrchr1.methXtractor.temp by positions (using -S of 2G)\n",
      "Successfully deleted the temporary input file CpG_OT_merged.sorted.txt.gz.chrchr1.methXtractor.temp\n",
      "\n",
      "Sorting input file CpG_OT_merged.sorted.txt.gz.chrchr10.methXtractor.temp by positions (using -S of 2G)\n",
      "Successfully deleted the temporary input file CpG_OT_merged.sorted.txt.gz.chrchr10.methXtractor.temp\n",
      "\n",
      "Sorting input file CpG_OT_merged.sorted.txt.gz.chrchr11.methXtractor.temp by positions (using -S of 2G)\n",
      "Successfully deleted the temporary input file CpG_OT_merged.sorted.txt.gz.chrchr11.methXtractor.temp\n",
      "\n",
      "Sorting input file CpG_OT_merged.sorted.txt.gz.chrchr12.methXtractor.temp by positions (using -S of 2G)\n",
      "Successfully deleted the temporary input file CpG_OT_merged.sorted.txt.gz.chrchr12.methXtractor.temp\n",
      "\n",
      "Sorting input file CpG_OT_merged.sorted.txt.gz.chrchr13.methXtractor.temp by positions (using -S of 2G)\n",
      "Successfully deleted the temporary input file CpG_OT_merged.sorted.txt.gz.chrchr13.methXtractor.temp\n",
      "\n",
      "Sorting input file CpG_OT_merged.sorted.txt.gz.chrchr14.methXtractor.temp by positions (using -S of 2G)\n",
      "Successfully deleted the temporary input file CpG_OT_merged.sorted.txt.gz.chrchr14.methXtractor.temp\n",
      "\n",
      "Sorting input file CpG_OT_merged.sorted.txt.gz.chrchr15.methXtractor.temp by positions (using -S of 2G)\n",
      "Successfully deleted the temporary input file CpG_OT_merged.sorted.txt.gz.chrchr15.methXtractor.temp\n",
      "\n",
      "Sorting input file CpG_OT_merged.sorted.txt.gz.chrchr16.methXtractor.temp by positions (using -S of 2G)\n",
      "Successfully deleted the temporary input file CpG_OT_merged.sorted.txt.gz.chrchr16.methXtractor.temp\n",
      "\n",
      "Sorting input file CpG_OT_merged.sorted.txt.gz.chrchr17.methXtractor.temp by positions (using -S of 2G)\n",
      "Successfully deleted the temporary input file CpG_OT_merged.sorted.txt.gz.chrchr17.methXtractor.temp\n",
      "\n",
      "Sorting input file CpG_OT_merged.sorted.txt.gz.chrchr18.methXtractor.temp by positions (using -S of 2G)\n",
      "Successfully deleted the temporary input file CpG_OT_merged.sorted.txt.gz.chrchr18.methXtractor.temp\n",
      "\n",
      "Sorting input file CpG_OT_merged.sorted.txt.gz.chrchr19.methXtractor.temp by positions (using -S of 2G)\n",
      "Successfully deleted the temporary input file CpG_OT_merged.sorted.txt.gz.chrchr19.methXtractor.temp\n",
      "\n",
      "Sorting input file CpG_OT_merged.sorted.txt.gz.chrchr2.methXtractor.temp by positions (using -S of 2G)\n",
      "Successfully deleted the temporary input file CpG_OT_merged.sorted.txt.gz.chrchr2.methXtractor.temp\n",
      "\n",
      "Sorting input file CpG_OT_merged.sorted.txt.gz.chrchr20.methXtractor.temp by positions (using -S of 2G)\n",
      "Successfully deleted the temporary input file CpG_OT_merged.sorted.txt.gz.chrchr20.methXtractor.temp\n",
      "\n",
      "Sorting input file CpG_OT_merged.sorted.txt.gz.chrchr21.methXtractor.temp by positions (using -S of 2G)\n",
      "Successfully deleted the temporary input file CpG_OT_merged.sorted.txt.gz.chrchr21.methXtractor.temp\n",
      "\n",
      "Sorting input file CpG_OT_merged.sorted.txt.gz.chrchr22.methXtractor.temp by positions (using -S of 2G)\n",
      "Successfully deleted the temporary input file CpG_OT_merged.sorted.txt.gz.chrchr22.methXtractor.temp\n",
      "\n",
      "Sorting input file CpG_OT_merged.sorted.txt.gz.chrchr3.methXtractor.temp by positions (using -S of 2G)\n",
      "Successfully deleted the temporary input file CpG_OT_merged.sorted.txt.gz.chrchr3.methXtractor.temp\n",
      "\n",
      "Sorting input file CpG_OT_merged.sorted.txt.gz.chrchr4.methXtractor.temp by positions (using -S of 2G)\n",
      "Successfully deleted the temporary input file CpG_OT_merged.sorted.txt.gz.chrchr4.methXtractor.temp\n",
      "\n",
      "Sorting input file CpG_OT_merged.sorted.txt.gz.chrchr5.methXtractor.temp by positions (using -S of 2G)\n",
      "Successfully deleted the temporary input file CpG_OT_merged.sorted.txt.gz.chrchr5.methXtractor.temp\n",
      "\n",
      "Sorting input file CpG_OT_merged.sorted.txt.gz.chrchr6.methXtractor.temp by positions (using -S of 2G)\n",
      "Successfully deleted the temporary input file CpG_OT_merged.sorted.txt.gz.chrchr6.methXtractor.temp\n",
      "\n",
      "Sorting input file CpG_OT_merged.sorted.txt.gz.chrchr7.methXtractor.temp by positions (using -S of 2G)\n",
      "Successfully deleted the temporary input file CpG_OT_merged.sorted.txt.gz.chrchr7.methXtractor.temp\n",
      "\n",
      "Sorting input file CpG_OT_merged.sorted.txt.gz.chrchr8.methXtractor.temp by positions (using -S of 2G)\n",
      "Successfully deleted the temporary input file CpG_OT_merged.sorted.txt.gz.chrchr8.methXtractor.temp\n",
      "\n",
      "Sorting input file CpG_OT_merged.sorted.txt.gz.chrchr9.methXtractor.temp by positions (using -S of 2G)\n",
      "Successfully deleted the temporary input file CpG_OT_merged.sorted.txt.gz.chrchr9.methXtractor.temp\n",
      "\n",
      "Sorting input file CpG_OT_merged.sorted.txt.gz.chrchrX.methXtractor.temp by positions (using -S of 2G)\n",
      "Successfully deleted the temporary input file CpG_OT_merged.sorted.txt.gz.chrchrX.methXtractor.temp\n",
      "\n",
      "Finished BedGraph conversion ...\n",
      "\n",
      "samtools view: writing to standard output failed: Broken pipe\n",
      "samtools view: error reading file \"merged.sorted.bam\": Broken pipe\n",
      "samtools view: error closing standard output: -1\n",
      "\n",
      "gzip: stdout: Broken pipe\n"
     ]
    }
   ],
   "source": [
    "!/home/sagemaker-user/.conda/envs/rrbs-env/bin/bismark_methylation_extractor --samtools_path /home/sagemaker-user/.conda/envs/rrbs-env/bin/samtools --bedGraph --gzip --multicore 34 --buffer_size 2G --genome_folder genome_hg19 merged.sorted.bam\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "9e23cc0c-6ac8-4f65-ad57-1c27044133ba",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-04-24T14:22:52.211448Z",
     "iopub.status.busy": "2025-04-24T14:22:52.211083Z",
     "iopub.status.idle": "2025-04-24T14:22:52.363088Z",
     "shell.execute_reply": "2025-04-24T14:22:52.362285Z",
     "shell.execute_reply.started": "2025-04-24T14:22:52.211425Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-rw-r--r-- 1 sagemaker-user users 274M Apr 21 09:03 CHG_OB_merged.sorted.txt.gz\n",
      "-rw-r--r-- 1 sagemaker-user users 375M Apr 24 13:56 CHG_OB_merged.txt.gz\n",
      "-rw-r--r-- 1 sagemaker-user users 274M Apr 21 09:03 CHG_OT_merged.sorted.txt.gz\n",
      "-rw-r--r-- 1 sagemaker-user users 377M Apr 24 13:56 CHG_OT_merged.txt.gz\n",
      "-rw-r--r-- 1 sagemaker-user users 441M Apr 21 09:03 CHH_OB_merged.sorted.txt.gz\n",
      "-rw-r--r-- 1 sagemaker-user users 570M Apr 24 13:56 CHH_OB_merged.txt.gz\n",
      "-rw-r--r-- 1 sagemaker-user users 440M Apr 21 09:03 CHH_OT_merged.sorted.txt.gz\n",
      "-rw-r--r-- 1 sagemaker-user users 569M Apr 24 13:56 CHH_OT_merged.txt.gz\n",
      "-rw-r--r-- 1 sagemaker-user users 319M Apr 21 09:03 CpG_OB_merged.sorted.txt.gz\n",
      "-rw-r--r-- 1 sagemaker-user users 452M Apr 24 13:56 CpG_OB_merged.txt.gz\n",
      "-rw-r--r-- 1 sagemaker-user users 317M Apr 21 09:03 CpG_OT_merged.sorted.txt.gz\n",
      "-rw-r--r-- 1 sagemaker-user users 451M Apr 24 13:56 CpG_OT_merged.txt.gz\n",
      "-rw-r--r-- 1 sagemaker-user users 643M Feb  9  2019 ENCFF000MBM.fastq.gz\n",
      "-rw-r--r-- 1 sagemaker-user users 630M Apr 18 09:52 ENCFF000MBM_trimmed.fq.gz\n",
      "-rw-r--r-- 1 sagemaker-user users 1.7G Feb  9  2019 ENCFF000MBV.fastq.gz\n",
      "-rw-r--r-- 1 sagemaker-user users 1.7G Apr 18 10:03 ENCFF000MBV_trimmed.fq.gz\n",
      "-rw-r--r-- 1 sagemaker-user users  23M Apr 24 14:03 merged.bedGraph.gz\n",
      "-rw-r--r-- 1 sagemaker-user users  23M Apr 24 14:03 merged.bismark.cov.gz\n",
      "-rw-r--r-- 1 sagemaker-user users 2.0G Apr 21 09:56 merged.sorted.CX_report.txt.gz\n",
      "-rw-r--r-- 1 sagemaker-user users 128M Apr 21 09:28 merged.sorted.bedGraph.gz\n",
      "-rw-r--r-- 1 sagemaker-user users 112M Apr 21 09:28 merged.sorted.bismark.cov.gz\n"
     ]
    }
   ],
   "source": [
    "!ls -lh *.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "d1eb7829-0e2f-4c8d-a8df-fb47e9ce075c",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-04-24T14:48:42.088443Z",
     "iopub.status.busy": "2025-04-24T14:48:42.088088Z",
     "iopub.status.idle": "2025-04-24T14:48:42.835930Z",
     "shell.execute_reply": "2025-04-24T14:48:42.835329Z",
     "shell.execute_reply.started": "2025-04-24T14:48:42.088424Z"
    }
   },
   "outputs": [],
   "source": [
    "!gunzip merged.sorted.bedGraph.gz\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "1937a97e-c1fd-4a51-8691-f82625c561f3",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-04-24T14:49:14.683863Z",
     "iopub.status.busy": "2025-04-24T14:49:14.683533Z",
     "iopub.status.idle": "2025-04-24T14:49:15.563451Z",
     "shell.execute_reply": "2025-04-24T14:49:15.562683Z",
     "shell.execute_reply.started": "2025-04-24T14:49:14.683843Z"
    }
   },
   "outputs": [],
   "source": [
    "!sort -k1,1 -k2,2n merged.sorted.bedGraph > merged.sorted.sorted.bedGraph\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "e234b820-c0a9-4b1a-8525-dc9dfcc1547c",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-04-24T14:50:19.550149Z",
     "iopub.status.busy": "2025-04-24T14:50:19.549791Z",
     "iopub.status.idle": "2025-04-24T14:50:19.700218Z",
     "shell.execute_reply": "2025-04-24T14:50:19.699648Z",
     "shell.execute_reply.started": "2025-04-24T14:50:19.550126Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bisulfite_Genome  hg19_main.fa\n"
     ]
    }
   ],
   "source": [
    "!ls genome_hg19"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "fed857be-0458-4a5c-b0c6-ce2f537f9c59",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-04-24T14:51:45.445378Z",
     "iopub.status.busy": "2025-04-24T14:51:45.445017Z",
     "iopub.status.idle": "2025-04-24T14:52:07.735284Z",
     "shell.execute_reply": "2025-04-24T14:52:07.734542Z",
     "shell.execute_reply.started": "2025-04-24T14:51:45.445343Z"
    }
   },
   "outputs": [],
   "source": [
    "!/home/sagemaker-user/.conda/envs/rrbs-env/bin/samtools faidx genome_hg19/hg19_main.fa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "3cd82b53-50b9-4e41-824b-30a4018e2147",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-04-24T14:52:07.736492Z",
     "iopub.status.busy": "2025-04-24T14:52:07.736229Z",
     "iopub.status.idle": "2025-04-24T14:52:08.027052Z",
     "shell.execute_reply": "2025-04-24T14:52:08.026376Z",
     "shell.execute_reply.started": "2025-04-24T14:52:07.736472Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/bin/bash: line 1: bedGraphToBigWig: command not found\n"
     ]
    }
   ],
   "source": [
    "!cut -f1,2 genome_hg19/hg19_main.fa.fai > genome_hg19/hg19.chrom.sizes\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f87e142e-1c85-471f-b125-d1b479ea54f8",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-04-26T14:18:18.691130Z",
     "iopub.status.busy": "2025-04-26T14:18:18.690644Z",
     "iopub.status.idle": "2025-04-26T14:18:20.291551Z",
     "shell.execute_reply": "2025-04-26T14:18:20.290298Z",
     "shell.execute_reply.started": "2025-04-26T14:18:18.691099Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2025-04-26 14:18:18--  http://hgdownload.soe.ucsc.edu/admin/exe/linux.x86_64/bedGraphToBigWig\n",
      "Resolving hgdownload.soe.ucsc.edu (hgdownload.soe.ucsc.edu)... 128.114.119.163\n",
      "Connecting to hgdownload.soe.ucsc.edu (hgdownload.soe.ucsc.edu)|128.114.119.163|:80... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 8778344 (8.4M)\n",
      "Saving to: ‘bedGraphToBigWig’\n",
      "\n",
      "bedGraphToBigWig    100%[===================>]   8.37M  13.2MB/s    in 0.6s    \n",
      "\n",
      "2025-04-26 14:18:19 (13.2 MB/s) - ‘bedGraphToBigWig’ saved [8778344/8778344]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!wget http://hgdownload.soe.ucsc.edu/admin/exe/linux.x86_64/bedGraphToBigWig\n",
    "!chmod +x bedGraphToBigWig\n",
    "!sudo mv bedGraphToBigWig /usr/local/bin/\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "6fd22438-37aa-4a1b-bdce-786710b9bb04",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-04-24T14:54:43.007236Z",
     "iopub.status.busy": "2025-04-24T14:54:43.006859Z",
     "iopub.status.idle": "2025-04-24T14:54:43.559209Z",
     "shell.execute_reply": "2025-04-24T14:54:43.558559Z",
     "shell.execute_reply.started": "2025-04-24T14:54:43.007214Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Expecting at least 3 words line 3254461 of merged.sorted.sorted.bedGraph got 2\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# 4. Convert to BigWig\n",
    "!bedGraphToBigWig merged.sorted.sorted.bedGraph genome_hg19/hg19.chrom.sizes merged.sorted.bw"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9804d4c2-da19-449c-bb76-6b046fe5d58d",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-04-26T13:40:43.188697Z",
     "iopub.status.busy": "2025-04-26T13:40:43.188413Z",
     "iopub.status.idle": "2025-04-26T13:40:43.342026Z",
     "shell.execute_reply": "2025-04-26T13:40:43.341434Z",
     "shell.execute_reply.started": "2025-04-26T13:40:43.188676Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\t10469\t10470\t100\n",
      "chr1\t10471\t10472\t100\n",
      "chr1\t10484\t10485\t100\n",
      "chr1\t10489\t10490\t100\n",
      "chr1\t10493\t10494\t100\n",
      "chr1\t10496\t10497\t85.0340136054422\n",
      "chr1\t10497\t10498\t100\n",
      "chr1\t10524\t10525\t91.8367346938776\n",
      "chr1\t10525\t10526\t100\n",
      "chr1\t10541\t10542\t100\n"
     ]
    }
   ],
   "source": [
    "!head merged.sorted.sorted.bedGraph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a77d79f2-f7ef-4111-829f-b347ce007e30",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-04-26T13:43:48.051069Z",
     "iopub.status.busy": "2025-04-26T13:43:48.050766Z",
     "iopub.status.idle": "2025-04-26T13:43:48.950412Z",
     "shell.execute_reply": "2025-04-26T13:43:48.949728Z",
     "shell.execute_reply.started": "2025-04-26T13:43:48.051047Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "track type=bedGraph\n"
     ]
    }
   ],
   "source": [
    "!awk 'NF<3' merged.sorted.sorted.bedGraph\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a14cfe93-eadd-47dc-bd98-5af48e378a16",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-04-26T13:44:34.834882Z",
     "iopub.status.busy": "2025-04-26T13:44:34.834603Z",
     "iopub.status.idle": "2025-04-26T13:44:35.507501Z",
     "shell.execute_reply": "2025-04-26T13:44:35.506750Z",
     "shell.execute_reply.started": "2025-04-26T13:44:34.834861Z"
    }
   },
   "outputs": [],
   "source": [
    "!awk 'NF>=3' merged.sorted.sorted.bedGraph > fixed.bedGraph\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1c77de88-3627-4c91-ba59-59453d945f1b",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-04-26T14:18:12.579018Z",
     "iopub.status.busy": "2025-04-26T14:18:12.578592Z",
     "iopub.status.idle": "2025-04-26T14:18:12.749785Z",
     "shell.execute_reply": "2025-04-26T14:18:12.748712Z",
     "shell.execute_reply.started": "2025-04-26T14:18:12.578986Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/bin/bash: line 1: bedGraphToBigWig: command not found\n"
     ]
    }
   ],
   "source": [
    "!bedGraphToBigWig fixed.bedGraph genome_hg19/hg19.chrom.sizes merged.sorted.bw\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "0ad712a3-6672-4251-aab8-c5686303f555",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-05-09T05:31:36.872426Z",
     "iopub.status.busy": "2025-05-09T05:31:36.872064Z",
     "iopub.status.idle": "2025-05-09T05:31:37.026142Z",
     "shell.execute_reply": "2025-05-09T05:31:37.025410Z",
     "shell.execute_reply.started": "2025-05-09T05:31:36.872408Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-rw-r--r-- 1 sagemaker-user users 248M May 28  2021 ATAC.bigWig\n",
      "-rw-r--r-- 1 sagemaker-user users 1.6G May  6 14:09 CTCF.bigWig\n",
      "-rw-r--r-- 1 sagemaker-user users 3.2G May  6 14:12 DNase.bigWig\n",
      "-rw-r--r-- 1 sagemaker-user users 162M Apr 28 08:29 H3K27me3.bigWig\n",
      "-rw-r--r-- 1 sagemaker-user users 450M Apr 28 08:10 H3K36me3.bigWig\n",
      "-rw-r--r-- 1 sagemaker-user users 196M May  6 14:12 H3K4me1.bigWig\n",
      "-rw-r--r-- 1 sagemaker-user users 172M Apr 26 14:15 H3K9me3.bigWig\n",
      "-rw-r--r-- 1 sagemaker-user users 336M May  6 14:12 H3k27ac.bigWig\n",
      "-rw-r--r-- 1 sagemaker-user users 225M May  6 14:12 H3k4me3.bigWig\n",
      "-rw-r--r-- 1 sagemaker-user users  28M May  6 14:12 RRBS.bigWig\n"
     ]
    }
   ],
   "source": [
    "!ls -lh *.bigWig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aec330ee-cdb9-4561-a096-ad432293b9e5",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
